<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<title>The Bioconductor project: All in One View</title>
<meta name="viewport" content="width=device-width, initial-scale=1">
<script src="../assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="../assets/styles.css">
<script src="../assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="../favicons/incubator/apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicons/incubator/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="../favicons/incubator/favicon-16x16.png">
<link rel="manifest" href="../favicons/incubator/site.webmanifest">
<link rel="mask-icon" href="../favicons/incubator/safari-pinned-tab.svg" color="#5bbad5">
<meta name="msapplication-TileColor" content="#da532c">
<meta name="theme-color" media="(prefers-color-scheme: light)" content="white">
<meta name="theme-color" media="(prefers-color-scheme: dark)" content="black">
</head>
<body>
    <header id="top" class="navbar navbar-expand-md top-nav incubator"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="incubator-logo" alt="Carpentries Incubator" src="../assets/images/incubator-logo.svg"><span class="badge text-bg-danger">
          <abbr title="This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught.">
            <a href="https://docs.carpentries.org/resources/curriculum/lesson-life-cycle.html" class="external-link alert-link">
              <i aria-hidden="true" class="icon" data-feather="alert-octagon" style="border-radius: 5px"></i>
              Pre-Alpha
            </a>
            <span class="visually-hidden">This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught.</span>
          </abbr>
        </span>

      </div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text">
<li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul>
</li>
      </div>

      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1">
<li><button class="dropdown-item" type="button" onclick="window.location.href='../aio.html';">Learner View</button></li>
        </ul>
</div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Carpentries Incubator" src="../assets/images/incubator-logo-sm.svg">
</div>
    <div class="lesson-title-md">
      The Bioconductor project
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0">
<li class="nav-item">
          <span class="lesson-title">
            The Bioconductor project
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/instructor-notes.html">Instructor Notes</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/images.html">Extract All Images</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown">
<hr>
<li><a class="dropdown-item" href="discuss.html">Discussion</a></li>
<li><a class="dropdown-item" href="reference.html"></a></li>
          </ul>
</li>
      </ul>
</div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="../instructor/aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div>
<!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  The Bioconductor project
</div>

<aside class="col-md-12 lesson-progress"><div style="width: %" class="percentage">
    %
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: %" aria-valuenow="" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text">
<li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul>
</li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="../aio.html">Learner View</a>
                      </div>
                    </div>
                  </div>
<!--/div.accordion-item-->
                </div>
<!--/div.accordion-flush-->
              </div>
<!--div.sidenav-view-selector -->
            </div>
<!--/div.col -->

            <hr>
</div>
<!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Schedule</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="01-setup.html">1. Introduction and setup</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="02-introduction-to-bioconductor.html">2. Bioconductorの概要</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="03-installing-bioconductor.html">3. Installing Bioconductor packages</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="04-getting-help.html">4. Getting help</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="05-s4.html">5. BioconductorにおけるS4クラス</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="06-biological-sequences.html">6. Working with biological sequences</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="07-genomic-ranges.html">7. ゲノミクスの範囲での作業</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="08-annotations.html">8. Working with annotations</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush10">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading10">
        <a href="09-summarizedexperiment.html">9. SummarizedExperimentクラス</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width">
<div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul>
<li>
                        <a href="../instructor/key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="../instructor/instructor-notes.html">Instructor Notes</a>
                      </li>
                      <li>
                        <a href="../instructor/images.html">Extract All Images</a>
                      </li>
                      <hr>
<li><a class="dropdown-item" href="discuss.html">Discussion</a></li>
<li><a class="dropdown-item" href="reference.html"></a></li>
                    </ul>
</div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources">
<a href="../instructor/aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none">
<div class="d-grid gap-1">

            </div>
          </div>
<!-- /div.accordion -->
        </div>
<!-- /div.sidebar-inner -->
      </nav>
</div>
<!-- /div.sidebar -->
  </div>
<!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-extra.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <main id="main-content" class="main-content"><div class="container lesson-content">
        
        
<section id="aio-01-setup"><p>Content from <a href="01-setup.html">Introduction and setup</a></p>
<hr>
<p>Last updated on 2025-07-22 |

        <a href="https://github.com/carpentries-incubator/bioc-project/edit/main/episodes/01-setup.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> NA minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>Am I using the correct version of R for this lesson?</li>
<li>Why does my version of R matter?</li>
<li>How do I obtain the files that are used in this lesson?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Ensure that participants are using the correct version of R to
reproduce exactly the contents of this lesson.</li>
<li>Download the example files for this lesson.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="version-of-r">Version of R<a class="anchor" aria-label="anchor" href="#version-of-r"></a>
</h2>
<hr class="half-width">
<p>This lesson was developed and tested with R version 4.4.3
(2025-02-28).</p>
<p>Take a moment to launch RStudio and verify that you are using R
version 4.4.x, with <code>x</code> being any patch version,
e.g. 4.4.3.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">R.version.string</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "R version 4.4.3 (2025-02-28)"</code></pre>
</div>
<p>This is important because Bioconductor uses the version of R running
in the current session to determine the version of Bioconductor packages
that can be installed in the R library associated with the current R
session. Using a different version of R while following this lesson may
lead to unexpected results.</p>
</section><section><h2 class="section-heading" id="download-files">Download files<a class="anchor" aria-label="anchor" href="#download-files"></a>
</h2>
<hr class="half-width">
<p>Several episodes in this lesson rely on example files that
participants need to download.</p>
<p>Run the code below programmatically create a folder called
<code>data</code> in the current working directory, and download the
lesson files in that folder.</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dir.create</span><span class="op">(</span><span class="st">"data"</span>, showWarnings <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu">download.file</span><span class="op">(</span></span>
<span>    url <span class="op">=</span> <span class="st">"https://raw.githubusercontent.com/Bioconductor/bioconductor-teaching/master/data/TrimmomaticAdapters/TruSeq3-PE-2.fa"</span>, </span>
<span>    destfile <span class="op">=</span> <span class="st">"data/TruSeq3-PE-2.fa"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu">download.file</span><span class="op">(</span></span>
<span>    url <span class="op">=</span> <span class="st">"https://raw.githubusercontent.com/Bioconductor/bioconductor-teaching/master/data/ActbGtf/actb.gtf"</span>, </span>
<span>    destfile <span class="op">=</span> <span class="st">"data/actb.gtf"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu">download.file</span><span class="op">(</span></span>
<span>    url <span class="op">=</span> <span class="st">"https://raw.githubusercontent.com/Bioconductor/bioconductor-teaching/master/data/ActbOrf/actb_orfs.fasta"</span>, </span>
<span>    destfile <span class="op">=</span> <span class="st">"data/actb_orfs.fasta"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu">download.file</span><span class="op">(</span></span>
<span>  url <span class="op">=</span> <span class="st">"https://raw.githubusercontent.com/Bioconductor/bioconductor-teaching/devel/data/SummarizedExperiment/counts.csv"</span>,</span>
<span>  destfile <span class="op">=</span> <span class="st">"data/counts.csv"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu">download.file</span><span class="op">(</span></span>
<span>  url <span class="op">=</span> <span class="st">"https://raw.githubusercontent.com/Bioconductor/bioconductor-teaching/devel/data/SummarizedExperiment/gene_metadata.csv"</span>,</span>
<span>  destfile <span class="op">=</span> <span class="st">"data/gene_metadata.csv"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu">download.file</span><span class="op">(</span></span>
<span>  url <span class="op">=</span> <span class="st">"https://raw.githubusercontent.com/Bioconductor/bioconductor-teaching/devel/data/SummarizedExperiment/sample_metadata.csv"</span>,</span>
<span>  destfile <span class="op">=</span> <span class="st">"data/sample_metadata.csv"</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<div id="note" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="note" class="callout-inner">
<h3 class="callout-title">Note</h3>
<div class="callout-content">
<p>Ideally, participants might want to create a new <a href="https://support.rstudio.com/hc/en-us/articles/200526207-Using-Projects" class="external-link">RStudio
project</a> and download the lesson files in a sub-directory of that
project.</p>
<p>Using an RStudio project sets the working directory to the root
directory of that project. As a consequence, code is executed relative
to that root directory, often avoiding the need for using absolute file
paths to import/export data from/to files.</p>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>Participants will only be able to install the version of
Bioconductor packages described in this lesson and reproduce their exact
outputs if they use the correct version of R.</li>
<li>The files used in this lesson should be downloaded in a local path
that is easily accessible from an R session.</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-02-introduction-to-bioconductor"><p>Content from <a href="02-introduction-to-bioconductor.html">Bioconductorの概要</a></p>
<hr>
<p>Last updated on 2025-07-22 |

        <a href="https://github.com/carpentries-incubator/bioc-project/edit/main/episodes/02-introduction-to-bioconductor.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> NA minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>Bioconductorプロジェクトには何が含まれていますか？</li>
<li>BioconductorプロジェクトはCRANリポジトリとどのように関連していますか？</li>
<li>Bioconductorのパッケージを効果的に使用するにはどうすればいいですか？</li>
<li>Bioconductorコミュニティに参加し、コミュニケーションをとるにはどうすればいいですか？</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Bioconductorプロジェクトについて全体を説明する。</li>
<li>RエコシステムにおけるBioconductorプロジェクトの全体像を把握する。</li>
<li>Bioconductorプロジェクトの今後の更新情報を監視するための情報源を特定する。</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="bioconductorとは何ですか">Bioconductorとは何ですか？<a class="anchor" aria-label="anchor" href="#bioconductor%E3%81%A8%E3%81%AF%E4%BD%95%E3%81%A7%E3%81%99%E3%81%8B"></a>
</h2>
<hr class="half-width">
<div class="section level3">
<h3 id="bioconductorの簡単な歴史">Bioconductorの簡単な歴史<a class="anchor" aria-label="anchor" href="#bioconductor%E3%81%AE%E7%B0%A1%E5%8D%98%E3%81%AA%E6%AD%B4%E5%8F%B2"></a>
</h3>
<p>Bioconductorプロジェクトは2001年秋に始まり、計算生物学とバイオインフォマティクスのための拡張可能なソフトウェアの共同制作のためのイニシアティブとして立ち上げられました
(Gentleman, Carey, Bates, Bolstad, Dettling, Dudoit, Ellis, Gautier, Ge,
Gentry, Hornik, Hothorn, Huber, Iacus, Irizarry, Leisch, Li, Maechler,
Rossini, Sawitzki, Smith, Smyth, Tierney, Yang, and Zhang, 2004)。
プロジェクトの初めから、記載された使命は、大規模データセットと厳密かつ堅牢に設計された実験の技術的成果物のための統計分析と理解のためのツールを開発することでした。
統計分析を超えて、統計結果の解釈は、生物学的文脈、視覚化、および再現性を提供するパッケージによってサポートされています。</p>
<p>年月が経つにつれて、Bioconductorプロジェクトに貢献したソフトウェアパッケージは、マイクロアレイから単一細胞ゲノミクスまで、さまざまな高スループット技術の進化と出現を反映してきました。シーケンシング実験の多くのバリエーション（例：RNA-seq、ChIP-seq、DNA-seq）、分析（例：塩基配列のばらつき、コピー数の変動、単一ヌクレオチド多型）、およびデータモダリティ（例：フローサイトメトリー、プロテオミクス、顕微鏡および画像分析）を通じて。</p>
<p>重要なのは、プロジェクトは新しい統計テストと方法論を実装するソフトウェアパッケージをリリースするだけでなく、分子注釈と実験データセットのデータベースへのアクセスを提供する多様なパッケージタイプを生産してきたことです。</p>
<p>Bioconductorプロジェクトは、夏に北アメリカで開催される年次会議でクライマックスに達し、地域会議はいずれもヨーロッパ、アジア、北アメリカでのネットワーキングに優れた機会を提供します。
このプロジェクトは多様で包括的なコミュニティの促進に取り組んでおり、<a href="https://www.bioconductor.org/about/code-of-conduct/" class="external-link">行動規範</a>を含む、行動規範委員会によって施行されています。</p>
<figure><img src="../fig/bioc-timeline.svg" alt="Major Bioconductor milestonesと技術の進歩のタイムライン。" class="figure mx-auto d-block"></figure><p><strong>Major Bioconductor
milestonesと技術の進歩のタイムライン。</strong>
タイムラインの上部には、主要イベントの初めの発生が示されています。
タイムライン内では、コアインフラを提供するパッケージの名前がリリース日を示しています。
タイムラインの下部では、主要な技術的進歩がBioconductorプロジェクトの進化を時系列で文脈化します。</p>
</div>
<div class="section level3">
<h3 id="科学プロジェクト">科学プロジェクト<a class="anchor" aria-label="anchor" href="#%E7%A7%91%E5%AD%A6%E3%83%97%E3%83%AD%E3%82%B8%E3%82%A7%E3%82%AF%E3%83%88"></a>
</h3>
<p>元の出版物では、プロジェクトの設立時の目的と方法が説明されています
Gentleman, Carey, Bates et al. (2004)。</p>
<p>Huber, Carey, Gentleman, Anders, Carlson, Carvalho, Bravo, Davis,
Gatto, Girke, Gottardo, Hahne, Hansen, Irizarry, Lawrence, Love,
MacDonald, Obenchain, Oles, Pages, Reyes, Shannon, Smyth, Tenenbaum,
Waldron, and Morgan
(2015)は、コアインフラとケーススタディの説明を含むプロジェクトの進展を、ユーザーと開発者の両方の視点から説明しています。</p>
<p>Amezquita, Lun, Becht, Carey, Carpp, Geistlinger, Marini,
Rue-Albrecht, Risso, Soneson, Waldron, Pages, Smith, Huber, Morgan,
Gottardo, and Hicks
(2020)は、単一細胞ゲノミクス技術の目覚めに続くプロジェクトのさらなる発展をレビューしています。</p>
<p>多くの出版物や書籍の章がBioconductorプロジェクトを引用しており、最近の例は<a href="https://www.bioconductor.org/help/publications/" class="external-link">Bioconductorウェブサイト</a>にリストされています。</p>
</div>
<div class="section level3">
<h3 id="パッケージリポジトリ">パッケージリポジトリ<a class="anchor" aria-label="anchor" href="#%E3%83%91%E3%83%83%E3%82%B1%E3%83%BC%E3%82%B8%E3%83%AA%E3%83%9D%E3%82%B8%E3%83%88%E3%83%AA"></a>
</h3>
<div class="section level4">
<h4 id="概要およびcranとの関係">概要およびCRANとの関係<a class="anchor" aria-label="anchor" href="#%E6%A6%82%E8%A6%81%E3%81%8A%E3%82%88%E3%81%B3cran%E3%81%A8%E3%81%AE%E9%96%A2%E4%BF%82"></a>
</h4>
<p>間違いなく、ソフトウェアパッケージはBioconductorプロジェクトの最も知られている側面です。
2001年に設立されて以来、リポジトリは時間とともに成長し、数千のパッケージをホストするようになりました。</p>
<p>Bioconductorプロジェクトは、Rパッケージが主にバイオインフォマティクスと計算生物学の分析を対象にしている既存のCRANリポジトリを拡張しました。</p>
<div id="さらに進む" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="さらに進む" class="callout-inner">
<h3 class="callout-title">さらに進む</h3>
<div class="callout-content">
<p>このレッスンの<a href="discuss.html#bioconductor-and-CRAN">ディスカッション</a>記事には、BioconductorとCRANの関係についてさらに詳しく説明しているセクションが含まれています。</p>
</div>
</div>
</div>
</div>
<div class="section level4">
<h4 id="bioconductorリリースサイクル">Bioconductorリリースサイクル<a class="anchor" aria-label="anchor" href="#bioconductor%E3%83%AA%E3%83%AA%E3%83%BC%E3%82%B9%E3%82%B5%E3%82%A4%E3%82%AF%E3%83%AB"></a>
</h4>
<p>Bioconductorプロジェクトは、CRANリポジトリのパッケージ管理インフラを拡張して、ユーザーのレベルでのパッケージの展開と管理をより良くサポートしました
(Gentleman, Carey, Bates et al., 2004)。
特に、Bioconductorプロジェクトは、6か月のリリースサイクル（通常は4月と10月頃）を特徴としており、Bioconductorリポジトリ内のすべてのパッケージの現在のバージョンのスナップショットが特定のRバージョン用に指定されています。
R自体は年に1回リリースされ（通常は4月頃）、Rの各リリースごとに2つのBioconductorパッケージの互換性のあるリリースが利用可能になります。</p>
<p>そのため、Bioconductorパッケージ開発者は、次のBioconductorプロジェクトのリリースに関連付けられるRのバージョンを常に使用する必要があります。
これは、10月から4月の間はRの開発バージョンを使用し、4月から10月の間はRのリリースバージョンを使用することを意味します。</p>
<p>重要なのは、厳格なBioconductorリリースサイクルが、ユーザーが互いに非常に遠く離れたバージョンのパッケージをインストールできないようにし、テストを受ける可能性が低いパッケージをインストールできなくすることです。
この慣行は、CRANとBioconductorの両方のパッケージの開発サイクルを反映しており、同時のパッケージが自動化されたシステムによって定期的にテストされ、パッケージ依存関係内の最新のソフトウェア更新が下流のパッケージに壊れることがないことを保証します。これにより、パッケージのメンテナが自分のソフトウェアを更新することを促します。</p>
<p>各Bioconductorリリースの前に、必要な自動テストスイートを通過しないパッケージは廃止され、その後リポジトリから削除されます。
これにより、各Bioconductorリリースは、相互に互換性があり、追跡可能で、関連するRのバージョンで機能するパッケージのスイートを提供することが保証されます。</p>
<figure><img src="../fig/bioc-release-cycle.svg" alt="選択されたBioconductorおよびRのバージョンのリリース日タイムライン。" class="figure mx-auto d-block"></figure><p><strong>選択されたBioconductorおよびRのバージョンのリリース日タイムライン。</strong>
タイムラインの上部には、Rプロジェクトのバージョンとおおよそのリリース日が示されています。
タイムラインの下部には、Bioconductorプロジェクトのバージョンとリリース日が示されています。
出典: <a href="https://bioconductor.org/about/release-announcements/" class="external-link">Bioconductor</a>。</p>
</div>
<div class="section level4">
<h4 id="パッケージタイプ">パッケージタイプ<a class="anchor" aria-label="anchor" href="#%E3%83%91%E3%83%83%E3%82%B1%E3%83%BC%E3%82%B8%E3%82%BF%E3%82%A4%E3%83%97"></a>
</h4>
<p>パッケージは、主に4つの主要カテゴリに分けられます。</p>
<ul>
<li>ソフトウェア</li>
<li>注釈データ</li>
<li>実験データ</li>
<li>ワークフロー</li>
</ul>
<p><a href="reference.html#software-package">ソフトウェアパッケージ</a>自体は、データを保存してアクセスするためのインフラ（すなわち、クラス）を提供するパッケージと、そのデータ構造に保存されたデータを処理するための方法論的ツールを提供するパッケージに再分割できます。
この構造と分析の分離はBioconductorプロジェクトの中心にあり、新しい方法論的ソフトウェアパッケージの開発者が可能な限り既存のデータコンテナを再利用することを奨励し、異なるデータ構造間の学習と変換の必要なく、ユーザーが代替ワークフローをより容易に試すことができるようにします。</p>
<p><a href="reference.html#annotationdata-package">注釈データパッケージ</a>は、多様なゲノム注釈の自己完結型データベース（例：遺伝子識別子、生物学的経路）を提供します。
さまざまなコレクションの注釈パッケージはBioconductorプロジェクトに存在します。
それらは、それぞれの命名パターンや含まれている情報によって識別できます。
たとえば、いわゆる<code>OrgDb</code>パッケージ（例として<em><a href="https://bioconductor.org/packages/3.19/org.Hs.eg.db" class="external-link">org.Hs.eg.db</a></em>パッケージ）は、異なるタイプの遺伝子識別子および経路データベースをマッピングする情報を提供します。
いわゆる<code>EnsDb</code>（例として<em><a href="https://bioconductor.org/packages/3.19/EnsDb.Hsapiens.v86" class="external-link">EnsDb.Hsapiens.v86</a></em>）パッケージは、BioconductorパッケージのEnsembl注釈の個別バージョンをカプセル化します。いわゆる<code>TxDb</code>パッケージ（例として<em><a href="https://bioconductor.org/packages/3.19/TxDb.Hsapiens.UCSC.hg38.knownGene" class="external-link">TxDb.Hsapiens.UCSC.hg38.knownGene</a></em>）は、個別バージョンのUCSC遺伝子注釈テーブルをカプセル化します。</p>
<p><a href="reference.html#experimentdata-package">実験データパッケージ</a>は、ソフトウェアパッケージの開発者が、<a href="reference.html#vignette">パッケージのビネット</a>において自分のパッケージの使用を示すためによく使用される自己完結型データセットを提供します。</p>
<p>最後に、<a href="reference.html#workflow-package">ワークフローパッケージ</a>は、他のパッケージの結合使用を示すビネットのコレクションを提供しますが、新しいソースコードや機能は提供しません。</p>
<div id="チャレンジ-bioconductorウェブサイト" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="チャレンジ-bioconductorウェブサイト" class="callout-inner">
<h3 class="callout-title">チャレンジ: Bioconductorウェブサイト</h3>
<div class="callout-content">
<p>Bioconductorウェブサイトは<a href="https://bioconductor.org/" class="external-link">https://bioconductor.org/</a>からアクセス可能です。</p>
<p>ウェブサイトをブラウズして、次の質問に答える情報を見つけてください：</p>
<ol style="list-style-type: decimal">
<li>現在のBioconductorプロジェクトのリリースにはいくつのパッケージが含まれていますか？</li>
<li>この数には各タイプのパッケージがいくつ含まれていますか？</li>
</ol>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  解決策
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<p>以下の解決策には、執筆時点で有効だった数字（Bioconductorリリース3.13）が含まれています；
Bioconductorプロジェクトの将来のリリースでは、数字が異なることは避けられません。</p>
<ol style="list-style-type: decimal">
<li>
<a href="https://bioconductor.org/" class="external-link">https://bioconductor.org/</a>のページで、「インストール」というセクションを読むことができます。</li>
</ol>
<blockquote>
<p>Bioconductorリリース3.13で利用可能な2042ソフトウェアパッケージを発見する。</p>
</blockquote>
<ol start="2" style="list-style-type: decimal">
<li>
<a href="https://bioconductor.org/" class="external-link">https://bioconductor.org/</a>の「ニュース」セクションで、「Bioconductor
Bioc
<code>X.Y</code>リリース」というリンクをクリックしてください（<code>X.Y</code>は、この演習を通じて行くときの現在のBioconductorリリースのバージョン）。
リンクされたページでは、以下の内容を読むことができます：</li>
</ol>
<blockquote>
<p>Bioconductor
3.13を発表できることを嬉しく思います。これは2042ソフトウェアパッケージ、406の実験データパッケージ、965の注釈パッケージ、29のワークフローから構成されています。</p>
<p>新しいソフトウェアパッケージは133、新しいデータ実験パッケージは22、新しい注釈パッケージは7、新しいワークフローは1、新しい書籍はなく、多くの更新と既存パッケージの改善があります。Bioconductor
3.13はR
4.1.0と互換性があり、Linux、32ビットおよび64ビットのWindows、macOS
10.14.6 Mojave以上でサポートされています。
このリリースには更新されたBioconductor Dockerコンテナが含まれます。</p>
</blockquote>
</div>
</div>
</div>
</div>
</div>
<div class="section level4">
<h4 id="biocviewsを使用したパッケージ分類">biocViewsを使用したパッケージ分類<a class="anchor" aria-label="anchor" href="#biocviews%E3%82%92%E4%BD%BF%E7%94%A8%E3%81%97%E3%81%9F%E3%83%91%E3%83%83%E3%82%B1%E3%83%BC%E3%82%B8%E5%88%86%E9%A1%9E"></a>
</h4>
<p>Bioconductorプロジェクトは、Bioconductorパッケージを分類し、<a href="https://www.bioconductor.org/packages/release/BiocViews.html" class="external-link">Bioconductorウェブサイト</a>でのテーマ別検索を介してその発見を容易にするために、制御語彙からの用語の集合である<a href="reference.html#biocviews">biocViews</a>を使用します。</p>
<p>各Bioconductorパッケージは、パッケージのタイプと機能を説明するために選ばれた制御語彙からの小さな用語セットでタグ付けされます。
用語は最初にパッケージの著者によって選択され、次にパッケージレビューや制御語彙の更新中に精査されます。</p>
<div id="チャレンジ" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="チャレンジ" class="callout-inner">
<h3 class="callout-title">チャレンジ</h3>
<div class="callout-content">
<p>BioconductorのすべてのパッケージのリストをBioconductor <a href="https://www.bioconductor.org/packages/release/BiocViews.html" class="external-link">biocViews</a>のウェブページで訪れてください。
左上の「Autocomplete biocViews
search」ボックスを使用して、カテゴリ別にパッケージをフィルタリングし、個々の用語を展開および収縮させてソフトウェアパッケージのグラフを探ります。</p>
<ol style="list-style-type: decimal">
<li>RNAシーケンシング解析のためにタグ付けされたパッケージを識別するために使用できるbiocView用語は何ですか？
ChIP-seq？ エピジェネティクス？ バリアント注釈？ プロテオミクス？
単一細胞ゲノミクス？</li>
<li>
<code>RNASeq</code>カテゴリでは、非常に人気のある2つのパッケージがあり、<em><a href="https://bioconductor.org/packages/3.19/DESeq2" class="external-link">DESeq2</a></em>と、<em><a href="https://bioconductor.org/packages/3.19/edgeR" class="external-link">edgeR</a></em>です。
ダウンロード統計（すなわち、低いランク）という点で、どちらがより人気がありますか？</li>
</ol>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">
  解決策
  </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<ol style="list-style-type: decimal">
<li><p><code>RNAseq</code>, <code>ChIPSeq</code>,
<code>Epigenetics</code>, <code>VariantAnnotation</code>,
<code>Proteomics</code>, <code>SingleCell</code>.</p></li>
<li><p>Bioconductorリリース<code>3.14</code>では、<em><a href="https://bioconductor.org/packages/3.19/DESeq2" class="external-link">DESeq2</a></em>と<em><a href="https://bioconductor.org/packages/3.19/edgeR" class="external-link">edgeR</a></em>がそれぞれランク23と28にリストされています。
言い換えれば、2つのパッケージはBioconductorプロジェクトで最も頻繁にダウンロードされるパッケージの中にあり、この場合、<em><a href="https://bioconductor.org/packages/3.19/edgeR" class="external-link">edgeR</a></em>に小さな利点があります。</p></li>
</ol>
</div>
</div>
</div>
</div>
<div id="さらに進む-1" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="さらに進む-1" class="callout-inner">
<h3 class="callout-title">さらに進む</h3>
<div class="callout-content">
<p>Bioconductorパッケージ<em><a href="https://bioconductor.org/packages/3.19/biocViews" class="external-link">biocViews</a></em>は、制御語彙のインフラをサポートおよび管理するために使用されます。
また、用語のリストをプログラム的に検査し、グラフとしてその関係によってサブセットを取得するためにも使用されます。</p>
<p>さらに、<em><a href="https://bioconductor.org/packages/3.19/BiocPkgTools" class="external-link">BiocPkgTools</a></em>パッケージを使用して、異なる<a href="reference.html#biocviews">biocViews</a>のもとにあるパッケージをブラウズすることができます
(Su, Carey, Shepherd, Ritchie, Morgan, and Davis, 2019)。</p>
</div>
</div>
</div>
</div>
<div class="section level4">
<h4 id="パッケージの相互運用性">パッケージの相互運用性<a class="anchor" aria-label="anchor" href="#%E3%83%91%E3%83%83%E3%82%B1%E3%83%BC%E3%82%B8%E3%81%AE%E7%9B%B8%E4%BA%92%E9%81%8B%E7%94%A8%E6%80%A7"></a>
</h4>
<p>Bioconductorの哲学の中心には、相互運用性の概念があります。
つまり、パッケージが同じデータ構造で動作する能力です。
重要なのは、相互運用性はユーザーと開発者の両方に利益をもたらすことです。</p>
<p>ユーザーは、複数のパッケージを組み合わせた任意の複雑なワークフローを書くことがより簡単になります。
同じデータ構造に基づくパッケージがあるため、ユーザーはワークフローの実際のステップに最大限注意を払い、各パッケージに特有の異なるデータ構造間で行う複雑かつエラーが発生しやすい変換に費やす時間を最小限に抑えることができます。
同様に、新しいパッケージの開発者は、情報の保存、検証、およびインデックスを行うために堅牢で信頼されるインフラを提供する既存のデータ構造を借用し、革新的な機能の実装に焦点を当てることができます。</p>
<p>最終的に、下図は、さまざまなBioconductorパッケージとベースRパッケージがどのように組み合わされて、シーケンシングデータをRセッションにインポートし、データと結果の注釈、統合、視覚化を実行するさまざまな分析を実行できるかを示しています。</p>
<figure><img src="../fig/bioc-sequencing-ecosystem.svg" alt="シーケンシングエコシステム。" class="figure mx-auto d-block"></figure><p><strong>シーケンシングエコシステム</strong>
主要データ処理ステップ（青）と関連ソフトウェアパッケージ（ピンク）が、さまざまなタイプのゲノム分析の典型的なワークフローの文脈で一覧表示されます。</p>
<p>ワークフローのステップとソフトウェアパッケージの連続的な関係は、完全なエンドツーエンドのワークフローを構成するためのソフトウェアパッケージ間の相互運用性の重要性を示しています。
ワークフローのステップとソフトウェアパッケージの逐次関係は、完全なエンドツーエンドのワークフローを構成するために、ソフトウェアパッケージ間の相互運用性の重要性を示しています。</p>
</div>
</div>
<div class="section level3">
<h3 id="会議コースワークショップ">会議、コース、ワークショップ<a class="anchor" aria-label="anchor" href="#%E4%BC%9A%E8%AD%B0%E3%82%B3%E3%83%BC%E3%82%B9%E3%83%AF%E3%83%BC%E3%82%AF%E3%82%B7%E3%83%A7%E3%83%83%E3%83%97"></a>
</h3>
<p>Bioconductorコミュニティは、年中世界中の数多くのイベントを定期的に開催しています。
例えば：</p>
<ul>
<li>北アメリカでの年次BioCサマー会議</li>
<li>冬の地域会議（例：BioC Europe、BioC Asia）</li>
<li>サマー学校（例：CSAMA）</li>
<li>すべてのコミュニティメンバーが参加できるオンラインミーティング（例：Bioconductor
Developers Forum）</li>
</ul>
<p>コース教材は、これらのイベントの各後に定期的に<a href="https://bioconductor.org/help/course-materials/" class="external-link">Bioconductorウェブサイト</a>にアップロードされます。
特に、<a href="https://www.bioconductor.org/books/release/" class="external-link">オンライン書籍</a>はコミュニティメンバーによって開発および維持されています。</p>
<p>Bioconductorの<a href="https://www.youtube.com/user/bioconductor" class="external-link">YouTubeチャンネル</a>は、会議の発表のビデオ録画（講演やワークショップを含む）を公開するために使用され、通常のBioconductor開発者フォーラムの版（リンクが必要）も含まれています。</p>
<div id="貢献しよう" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="貢献しよう" class="callout-inner">
<h3 class="callout-title">貢献しよう！</h3>
<div class="callout-content">
<p>年を通じての会議の典型的なサイクルを示すことができれば素晴らしいでしょう。例えば、</p>
<ul>
<li>7月末頃の北アメリカでのBioC会議</li>
<li>12月頃のヨーロッパでのEuroBioC会議</li>
<li>11月頃のアジアでのBioCAsia会議</li>
</ul>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="オンラインコミュニケーションチャネル">オンラインコミュニケーションチャネル<a class="anchor" aria-label="anchor" href="#%E3%82%AA%E3%83%B3%E3%83%A9%E3%82%A4%E3%83%B3%E3%82%B3%E3%83%9F%E3%83%A5%E3%83%8B%E3%82%B1%E3%83%BC%E3%82%B7%E3%83%A7%E3%83%B3%E3%83%81%E3%83%A3%E3%83%8D%E3%83%AB"></a>
</h3>
<div class="section level4">
<h4 id="サポートサイト">サポートサイト<a class="anchor" aria-label="anchor" href="#%E3%82%B5%E3%83%9D%E3%83%BC%E3%83%88%E3%82%B5%E3%82%A4%E3%83%88"></a>
</h4>
<p>Bioconductorの<a href="https://support.bioconductor.org/" class="external-link">サポートサイト</a>は、ユーザーと開発者が自由にコミュニケーションをとるためのプラットフォームを提供し（Bioconductorの<a href="https://www.bioconductor.org/about/code-of-conduct/" class="external-link">行動規範</a>に従って）、パッケージに関する問題やベストプラクティスに関する概念的な質問について議論します。</p>
</div>
<div class="section level4">
<h4 id="slackワークスペース">Slackワークスペース<a class="anchor" aria-label="anchor" href="#slack%E3%83%AF%E3%83%BC%E3%82%AF%E3%82%B9%E3%83%9A%E3%83%BC%E3%82%B9"></a>
</h4>
<p>Bioconductorの<a href="https://bioc-community.herokuapp.com/" class="external-link">Slackワークスペース</a>は、すべてのコミュニティメンバーが参加（無料）して、迅速な相互作用を行うことができるオープンスペースです。
現在、「Pro」プランの価格設定は、コア資金によって支援されています。</p>
<ul>
<li>無制限のメッセージアーカイブ</li>
<li>無制限のアプリ</li>
<li>画面共有付きのグループビデオ通話</li>
<li>Slack Connectを使用して他の組織と安全に作業する</li>
</ul>
<p>さまざまなトピックについて議論するために多くのチャンネルが作成されており、コミュニティメンバーはそのチャンネルに自由に参加することができ、また新しいチャンネルを作成して新しいトピックについて議論することもできます。</p>
<p>重要な発表は<code>#general</code>チャンネルに投稿されます。</p>
<div id="注意" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="注意" class="callout-inner">
<h3 class="callout-title">注意</h3>
<div class="callout-content">
<p>ユーザーは、Bioconductorの<a href="https://support.bioconductor.org/" class="external-link">サポートサイト</a>を使用して、より広いコミュニティに関連する問題を提起することを奨励されています。
Slack
ワークスペースは、ライブディスカッションに最も便利で、広く参加しているチャンネル（例：<code>#general</code>）は適度に使用する必要があります。</p>
</div>
</div>
</div>
</div>
<div class="section level4">
<h4 id="開発者メーリングリスト">開発者メーリングリスト<a class="anchor" aria-label="anchor" href="#%E9%96%8B%E7%99%BA%E8%80%85%E3%83%A1%E3%83%BC%E3%83%AA%E3%83%B3%E3%82%B0%E3%83%AA%E3%82%B9%E3%83%88"></a>
</h4>
<p><a href="mailto:bioc-devel@r-project.org">bioc-devel@r-project.org</a>
メーリングリストは、パッケージ開発者間のコミュニケーションや
Bioconductor コアチームからの発表に使用されます。</p>
</div>
</div>
<div class="section level3">
<h3 id="科学技術コミュニティ">科学技術コミュニティ<a class="anchor" aria-label="anchor" href="#%E7%A7%91%E5%AD%A6%E6%8A%80%E8%A1%93%E3%82%B3%E3%83%9F%E3%83%A5%E3%83%8B%E3%83%86%E3%82%A3"></a>
</h3>
<ul>
<li>
<a href="https://bioconductor.org/about/scientific-advisory-board/" class="external-link">科学諮問委員会
(SAB)</a>
年に1度の会合、プロジェクトアドバイザーとして活動する外部および内部のリーダー。
任期の制限なし。</li>
<li>
<a href="https://bioconductor.org/about/technical-advisory-board/" class="external-link">技術技術委員会
(TAB)</a>。
月に一度、プロジェクトのコアインフラストラクチャと科学的方向性の技術的側面を検討します。
メンバーは 15 人、任期は 3 年。
年に一度の公開選挙でメンバーを交代します。
現在の役員は、ビンス・ケアリー（議長）、リーヴァイ・ウォルドロン（副議長）、シャーロット・ソネソン（書記）。</li>
<li>
<a href="https://bioconductor.org/about/community-advisory-board/" class="external-link">コミュニティ諮問委員会
(CAB)</a> 月に一度、地域社会への
outreach、イベント、教育、トレーニングを考慮します。 メンバーは 15
人、任期は 3 年。 年に一度の公開選挙でメンバーを交代します。
現在の役員は、エイディン・カルハン（議長）、マット・リッチー（共同議長）、ローリ・カーン（書記）。</li>
<li><a href="https://www.bioconductor.org/about/code-of-conduct/" class="external-link">行動規範委員会</a></li>
</ul>
<div id="注意-1" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="注意-1" class="callout-inner">
<h3 class="callout-title">注意</h3>
<div class="callout-content">
<p>TAB/CABの少なくとも1名のメンバーが両方に参加し、ボードのコミュニケーションを確保するための連絡役を果たします。</p>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="参考文献">参考文献<a class="anchor" aria-label="anchor" href="#%E5%8F%82%E8%80%83%E6%96%87%E7%8C%AE"></a>
</h2>
<hr class="half-width">
<p>[1] R. A. Amezquita, A. T. L. Lun, E. Becht, et al. “Orchestrating
single-cell analysis with Bioconductor”. In: <em>Nat Methods</em> 17.2
(2020), pp. 137-145. ISSN: 1548-7105 (Electronic) 1548-7091 (Linking).
DOI: 10.1038/s41592-019-0654-x. <a href="https://www.ncbi.nlm.nih.gov/pubmed/31792435" class="external-link uri">https://www.ncbi.nlm.nih.gov/pubmed/31792435</a>.</p>
<p>[2] R. C. Gentleman, V. J. Carey, D. M. Bates, et al. “Bioconductor:
open software development for computational biology and bioinformatics”.
In: <em>Genome Biol</em> 5.10 (2004), p. R80. ISSN: 1474-760X
(Electronic) 1474-7596 (Linking). DOI: 10.1186/gb-2004-5-10-r80. <a href="https://www.ncbi.nlm.nih.gov/pubmed/15461798" class="external-link uri">https://www.ncbi.nlm.nih.gov/pubmed/15461798</a>.</p>
<p>[3] W. Huber, V. J. Carey, R. Gentleman, et al. “Orchestrating
high-throughput genomic analysis with Bioconductor”. In: <em>Nat
Methods</em> 12.2 (2015), pp. 115-21. ISSN: 1548-7105 (Electronic)
1548-7091 (Linking). DOI: 10.1038/nmeth.3252. <a href="https://www.ncbi.nlm.nih.gov/pubmed/25633503" class="external-link uri">https://www.ncbi.nlm.nih.gov/pubmed/25633503</a>.</p>
<p>[4] S. Su, V. Carey, L. Shepherd, et al. “BiocPkgTools: Toolkit for
mining the Bioconductor package ecosystem [version 1; peer review: 2
approved, 1 approved with reservations]”. In: <em>F1000Research</em>
8.752 (2019). DOI: 10.12688/f1000research.19410.1.</p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>R パッケージは Bioconductor プロジェクトの一側面に過ぎません。</li>
<li>Bioconductor プロジェクトは CRAN
リポジトリを拡張し、補完します。</li>
<li>異なる種類のパッケージは、ソフトウェアだけでなく、アノテーション、実験データも提供し、統合されたワークフローで複数のパッケージの使用方法を示します。</li>
<li>Bioconductor
パッケージ間の相互運用性は、統合ワークフローの作成を容易にし、ユーザーの認知的負担を最小限に抑えます。</li>
<li>コースや会議の教育資料はアーカイブされ、Bioconductor ウェブサイトと
YouTube チャンネルでアクセス可能です。</li>
<li>異なるコミュニケーションのチャネルを使用して、コミュニティのメンバーが会話し、ユーザーとパッケージ開発者の両方として互いに助け合います。</li>
<li>Bioconductor
プロジェクトは、科学、技術、諮問の各委員会および行動規範委員会によって運営されています。</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-03-installing-bioconductor"><p>Content from <a href="03-installing-bioconductor.html">Installing Bioconductor packages</a></p>
<hr>
<p>Last updated on 2025-07-22 |

        <a href="https://github.com/carpentries-incubator/bioc-project/edit/main/episodes/03-installing-bioconductor.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> NA minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How do I install Bioconductor packages?</li>
<li>How do I check if newer versions of my installed packages are
available?</li>
<li>How do I update Bioconductor packages?</li>
<li>How do I find out the name of packages available from the
Bioconductor repositories?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Install BiocManager.</li>
<li>Install Bioconductor packages.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="biocmanager">BiocManager<a class="anchor" aria-label="anchor" href="#biocmanager"></a>
</h2>
<hr class="half-width">
<p>The <em><a href="https://bioconductor.org/packages/3.19/BiocManager" class="external-link">BiocManager</a></em>
package is the entry point into the Bioconductor package repository.
Technically, this is the only Bioconductor package distributed on the
CRAN repository.</p>
<p>It provides functions to safely install Bioconductor packages and
check for available updates.</p>
<p>Once the package is installed, the function
<code>BiocManager::install()</code> can be used to install packages from
the Bioconductor repository. The function is also capable of installing
packages from other repositories (e.g., CRAN), if those packages are not
found in the Bioconductor repository first.</p>
<figure><img src="../fig/bioc-install.svg" alt="The package BiocManager is available from the CRAN repository and used to install packages from the Bioconductor repository." class="figure mx-auto d-block"></figure><p><strong>The package BiocManager is available from the CRAN repository
and used to install packages from the Bioconductor repository.</strong>
The function <code>install.packages()</code> from the base R package
<code>utils</code> can be used to install the <em><a href="https://bioconductor.org/packages/3.19/BiocManager" class="external-link">BiocManager</a></em>
package distributed on the CRAN repository. In turn, the function
<code>BiocManager::install()</code> can be used to install packages
available on the Bioconductor repository. Notably, the
<code>BiocManager::install()</code> function will fall back on the CRAN
repository if a package cannot be found in the Bioconductor
repository.</p>
<p>Install the package using the code below.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">install.packages</span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span></code></pre>
</div>
<div id="going-further" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="going-further" class="callout-inner">
<h3 class="callout-title">Going further</h3>
<div class="callout-content">
<p>A number of packages that are not part of the base R installation
also provide functions to install packages from various repositories.
For instance:</p>
<ul>
<li><code>devtools::install()</code></li>
<li><code>remotes::install_bioc()</code></li>
<li><code>remotes::install_bitbucket()</code></li>
<li><code>remotes::install_cran()</code></li>
<li><code>remotes::install_dev()</code></li>
<li><code>remotes::install_github()</code></li>
<li><code>remotes::install_gitlab()</code></li>
<li><code>remotes::install_git()</code></li>
<li><code>remotes::install_local()</code></li>
<li><code>remotes::install_svn()</code></li>
<li><code>remotes::install_url()</code></li>
<li><code>renv::install()</code></li>
</ul>
<p>Those functions are beyond the scope of this lesson, and should be
used with caution and adequate knowledge of their specific behaviors.
The general recommendation is to use <code>BiocManager::install()</code>
over any other installation mechanism because it ensures proper
versioning of Bioconductor packages.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="bioconductor-releases-and-current-version">Bioconductor releases and current version<a class="anchor" aria-label="anchor" href="#bioconductor-releases-and-current-version"></a>
</h2>
<hr class="half-width">
<p>Once the <em><a href="https://bioconductor.org/packages/3.19/BiocManager" class="external-link">BiocManager</a></em>
package is installed, the <code>BiocManager::version()</code> function
displays the version (i.e., release) of the Bioconductor project that is
currently active in the R session.</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">version</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] '3.19'</code></pre>
</div>
<p>Using the correct version of R and Bioconductor packages is a key
aspect of reproducibility. The <em><a href="https://bioconductor.org/packages/3.19/BiocManager" class="external-link">BiocManager</a></em>
packages uses the version of R running in the current session to
determine the version of Biocondutor packages that can be installed in
the current R library.</p>
<p>The Bioconductor project produces two releases each year, one around
April and another one around October. The April release of Bioconductor
coincides with the annual release of R. The October release of
Bioconductor continues to use the same version of R for that annual
cycle (i.e., until the next release, in April).</p>
<figure><img src="../fig/bioc-release-cycle.svg" alt="Timeline of release dates for selected Bioconductor and R versions." class="figure mx-auto d-block"></figure><p><strong>Timeline of release dates for selected Bioconductor and R
versions.</strong> The upper section of the timeline indicates versions
and approximate release dates for the R project. The lower section of
the timeline indicates versions and release dates for the Bioconductor
project. Source: <a href="https://bioconductor.org/about/release-announcements/" class="external-link">Bioconductor</a>.</p>
<p>During each 6-month cycle of package development, Bioconductor tests
packages for compatibility with the version of R that will be available
for the next release cycle. Then, each time a new Bioconductor release
is produced, the version of every package in the Bioconductor repository
is incremented, including the package <em><a href="https://bioconductor.org/packages/3.19/BiocVersion" class="external-link">BiocVersion</a></em>
which determines the version of the Bioconductor project.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">packageVersion</span><span class="op">(</span><span class="st">"BiocVersion"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] '3.19.1'</code></pre>
</div>
<p>This is the case for every package, even those which have not been
updated at all since the previous release. That new version of each
package is earmarked for the corresponding version of R; in other words,
that version of the package can only be installed and accessed in an R
session that uses the correct version of R. This version increment is
essential to associate a each version of a Bioconductor package with a
unique release of the Bioconductor project.</p>
<p>Following the April release, this means that users must install the
new version of R to access the newly released versions of Bioconductor
packages.</p>
<p>Instead, in October, users can continue to use the same version of R
to access the newly released version of Bioconductor packages. However,
to update an R library from the April release to the October release of
Bioconductor, users need to call the function
<code>BiocManager::install()</code> specifying the correct version of
Bioconductor as the <code>version</code> option, for instance:</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span>version <span class="op">=</span> <span class="st">"3.14"</span><span class="op">)</span></span></code></pre>
</div>
<p>This needs to be done only once, as the <em><a href="https://bioconductor.org/packages/3.19/BiocVersion" class="external-link">BiocVersion</a></em>
package will be updated to the corresponding version, indicating the
version of Bioconductor in use in this R library.</p>
<div id="going-further-1" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="going-further-1" class="callout-inner">
<h3 class="callout-title">Going further</h3>
<div class="callout-content">
<p>The <a href="discuss.html#the-bioconductor-release-cycle">Discussion</a>
article of this lesson includes a section discussing the release cycle
of the Bioconductor project.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="check-for-updates">Check for updates<a class="anchor" aria-label="anchor" href="#check-for-updates"></a>
</h2>
<hr class="half-width">
<p>The <code>BiocManager::valid()</code> function inspects the version
of packages currently installed in the user library, and checks whether
a new version is available for any of them on the Bioconductor
repository.</p>
<p>If everything is up-to-date, the function will simply print
<code>TRUE</code>.</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">valid</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning: 5 packages out-of-date; 0 packages too new</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
* sessionInfo()

R version 4.4.3 (2025-02-28)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8
 [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8
 [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C
[10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C

time zone: UTC
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base

other attached packages:
[1] BiocStyle_2.32.1

loaded via a namespace (and not attached):
 [1] digest_0.6.37          assertthat_0.2.1       R6_2.6.1
 [4] fastmap_1.2.0          xfun_0.52              magrittr_2.0.3
 [7] glue_1.8.0             knitr_1.50             sandpaper_0.16.13.9000
[10] htmltools_0.5.8.1      rmarkdown_2.29         lifecycle_1.0.4
[13] xml2_1.3.8             ps_1.9.1               cli_3.6.5
[16] processx_3.8.6         callr_3.7.6            vctrs_0.6.5
[19] renv_1.1.4             withr_3.0.2            compiler_4.4.3
[22] purrr_1.1.0            tools_4.4.3            tinkr_0.3.0
[25] evaluate_1.0.4         yaml_2.3.10            BiocManager_1.30.26
[28] pegboard_0.7.9         rlang_1.1.6

Bioconductor version '3.19'

  * 5 packages out-of-date
  * 0 packages too new

create a valid installation with

  BiocManager::install(c(
    "httr2", "pillar", "purrr", "Rcpp", "RSQLite"
  ), update = TRUE, ask = FALSE, force = TRUE)

more details: BiocManager::valid()$too_new, BiocManager::valid()$out_of_date</code></pre>
</div>
<p>Conveniently, if any package can be updated, the function generates
and displays the command needed to update those packages. Users simply
need to copy-paste and run that command in their R console.</p>
<div id="example-of-out-of-date-package-library" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="example-of-out-of-date-package-library" class="callout-inner">
<h3 class="callout-title">Example of out-of-date package library</h3>
<div class="callout-content">
<p>In the example below, the <code>BiocManager::valid()</code> function
did not return <code>TRUE</code>. Instead, it includes information about
the active user session, and displays the exact call to
<code>BiocManager::install()</code> that the user should run to replace
all the outdated packages detected in the user library with the latest
version available in CRAN or Bioconductor.</p>
<pre><code>&gt; BiocManager::valid()

* sessionInfo()

R version 4.1.0 (2021-05-18)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Big Sur 11.6

Matrix products: default
LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib

locale:
[1] en_GB.UTF-8/en_GB.UTF-8/en_GB.UTF-8/C/en_GB.UTF-8/en_GB.UTF-8

attached base packages:
[1] stats     graphics  grDevices datasets  utils     methods   base

loaded via a namespace (and not attached):
[1] BiocManager_1.30.16 compiler_4.1.0      tools_4.1.0         renv_0.14.0

Bioconductor version '3.13'

  * 18 packages out-of-date
  * 0 packages too new

create a valid installation with

  BiocManager::install(c(
    "cpp11", "data.table", "digest", "hms", "knitr", "lifecycle", "matrixStats", "mime", "pillar", "RCurl",
    "readr", "remotes", "S4Vectors", "shiny", "shinyWidgets", "tidyr", "tinytex", "XML"
  ), update = TRUE, ask = FALSE)

more details: BiocManager::valid()$too_new, BiocManager::valid()$out_of_date

Warning message:
18 packages out-of-date; 0 packages too new </code></pre>
<p>Specifically, in this example, the message tells the user to run the
following command to bring their installation up to date:</p>
<pre><code><span>  <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span></span>
<span>    <span class="st">"cpp11"</span>, <span class="st">"data.table"</span>, <span class="st">"digest"</span>, <span class="st">"hms"</span>, <span class="st">"knitr"</span>, <span class="st">"lifecycle"</span>, <span class="st">"matrixStats"</span>, <span class="st">"mime"</span>, <span class="st">"pillar"</span>, <span class="st">"RCurl"</span>,</span>
<span>    <span class="st">"readr"</span>, <span class="st">"remotes"</span>, <span class="st">"S4Vectors"</span>, <span class="st">"shiny"</span>, <span class="st">"shinyWidgets"</span>, <span class="st">"tidyr"</span>, <span class="st">"tinytex"</span>, <span class="st">"XML"</span></span>
<span>  <span class="op">)</span>, update <span class="op">=</span> <span class="cn">TRUE</span>, ask <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="exploring-the-package-repository">Exploring the package repository<a class="anchor" aria-label="anchor" href="#exploring-the-package-repository"></a>
</h2>
<hr class="half-width">
<p>The Bioconductor <a href="reference.html#biocviews">biocViews</a>,
demonstrated in the earlier episode <a href="https://carpentries-incubator.github.io/bioc-project/02-introduction-to-bioconductor/index.html#package-classification-using-biocviews" class="external-link">Introduction
to Bioconductor</a>, are a great way to discover new packages by
thematically browsing the hierarchical classification of Bioconductor
packages.</p>
<p>In addition, the <code>BiocManager::available()</code> function
returns the complete list of package names that are can be installed
from the Bioconductor and CRAN repositories. For instance the total
number of numbers that could be installed using <em><a href="https://bioconductor.org/packages/3.19/BiocManager" class="external-link">BiocManager</a></em></p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">length</span><span class="op">(</span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">available</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 26059</code></pre>
</div>
<p>Specifically, the union of current Bioconductor repositories and
other repositories on the search path can be displayed as follows.</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">repositories</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span>                                                <span class="va">BioCsoft</span></span>
<span>           <span class="st">"https://bioconductor.org/packages/3.19/bioc"</span></span>
<span>                                                 <span class="va">BioCann</span></span>
<span><span class="st">"https://bioconductor.org/packages/3.19/data/annotation"</span></span>
<span>                                                 <span class="va">BioCexp</span></span>
<span><span class="st">"https://bioconductor.org/packages/3.19/data/experiment"</span></span>
<span>                                           <span class="va">BioCworkflows</span></span>
<span>      <span class="st">"https://bioconductor.org/packages/3.19/workflows"</span></span>
<span>                                               <span class="va">BioCbooks</span></span>
<span>          <span class="st">"https://bioconductor.org/packages/3.19/books"</span></span>
<span>                                             <span class="va">carpentries</span></span>
<span>                    <span class="st">"https://carpentries.r-universe.dev"</span></span>
<span>                                     <span class="va">carpentries_archive</span></span>
<span>                    <span class="st">"https://carpentries.github.io/drat"</span></span>
<span>                                                    <span class="va">CRAN</span></span>
<span>                           <span class="st">"https://cloud.r-project.org"</span> </span></code></pre>
</div>
<p>Each repository URL can be accessed in a web browser, displaying the
full list of packages available from that repository. For instance,
navigate to <a href="https://bioconductor.org/packages/3.14/bioc" class="external-link">https://bioconductor.org/packages/3.14/bioc</a>.</p>
<div id="going-further-2" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="going-further-2" class="callout-inner">
<h3 class="callout-title">Going further</h3>
<div class="callout-content">
<p>The function <code>BiocManager::repositories()</code> can be combined
with the base function <code>available.packages()</code> to query
packages available specifically from any package repository, e.g. the
Bioconductor <a href="reference.html#software-package">software
package</a> repository.</p>
<pre><code>&gt; db = available.packages(repos = BiocManager::repositories()["BioCsoft"])
&gt; dim(db)
[1] 1948   17
&gt; head(rownames(db))
[1] "a4"          "a4Base"      "a4Classif"   "a4Core"      "a4Preproc"
[6] "a4Reporting"</code></pre>
</div>
</div>
</div>
<p>Conveniently, <code>BiocManager::available()</code> includes a
<code>pattern=</code> argument, particularly useful to navigate
annotation resources (the original use case motivating it). For
instance, a range of <a href="reference.html#annotationdata-package">Annotation data
packages</a> available for the mouse model organism can be listed as
follows.</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">available</span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"*Mmusculus"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] "BSgenome.Mmusculus.UCSC.mm10"        "BSgenome.Mmusculus.UCSC.mm10.masked"
 [3] "BSgenome.Mmusculus.UCSC.mm39"        "BSgenome.Mmusculus.UCSC.mm8"
 [5] "BSgenome.Mmusculus.UCSC.mm8.masked"  "BSgenome.Mmusculus.UCSC.mm9"
 [7] "BSgenome.Mmusculus.UCSC.mm9.masked"  "EnsDb.Mmusculus.v75"
 [9] "EnsDb.Mmusculus.v79"                 "PWMEnrich.Mmusculus.background"
[11] "TxDb.Mmusculus.UCSC.mm10.ensGene"    "TxDb.Mmusculus.UCSC.mm10.knownGene"
[13] "TxDb.Mmusculus.UCSC.mm39.knownGene"  "TxDb.Mmusculus.UCSC.mm39.refGene"
[15] "TxDb.Mmusculus.UCSC.mm9.knownGene"  </code></pre>
</div>
</section><section><h2 class="section-heading" id="installing-packages">Installing packages<a class="anchor" aria-label="anchor" href="#installing-packages"></a>
</h2>
<hr class="half-width">
<p>The <code>BiocManager::install()</code> function is used to install
or update packages.</p>
<p>The function takes a character vector of package names, and attempts
to install them from the Bioconductor repository.</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="st">"S4Vectors"</span>, <span class="st">"BiocGenerics"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<p>However, if any package cannot be found in the Bioconductor
repository, the function also searches for those packages in
repositories listed in the global option <code>repos</code>.</p>
<div id="contribute" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="contribute" class="callout-inner">
<h3 class="callout-title">Contribute !</h3>
<div class="callout-content">
<p>Add an example of non-Bioconductor package that can be installed
using BioManager. Preferably, a package that will be used later in this
lesson.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="uninstalling-packages">Uninstalling packages<a class="anchor" aria-label="anchor" href="#uninstalling-packages"></a>
</h2>
<hr class="half-width">
<p>Bioconductor packages can be removed from the R library like any
other R package, using the base R function
<code>remove.packages()</code>. In essence, this function simply removes
installed packages and updates index information as necessary. As a
result, it will not be possible to attach the package to a session or
browse the documentation of that package anymore.</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remove.packages</span><span class="op">(</span><span class="st">"S4Vectors"</span><span class="op">)</span></span></code></pre>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>The BiocManager package is available from the CRAN repository.</li>
<li>
<code>BiocManager::install()</code> is used to install and update
Bioconductor packages (but also from CRAN and GitHub).</li>
<li>
<code>BiocManager::valid()</code> is used to check for available
package updates.</li>
<li>
<code>BiocManager::version()</code> reports the version of
Bioconductor currently installed.</li>
<li>
<code>BiocManager::install()</code> can also be used to update an
entire R library to a specific version of Bioconductor.</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-04-getting-help"><p>Content from <a href="04-getting-help.html">Getting help</a></p>
<hr>
<p>Last updated on 2025-07-22 |

        <a href="https://github.com/carpentries-incubator/bioc-project/edit/main/episodes/04-getting-help.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> NA minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>Where can I find help online?</li>
<li>Where can I ask questions to package developers and other
users?</li>
<li>Where can I find documentation for a specific package?</li>
<li>Where can I learn best practices to combine multiple package into a
coherent workflow?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Identify online resources for help.</li>
<li>Access package documentation.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="getting-help-with-bioconductor-packages">Getting help with Bioconductor packages<a class="anchor" aria-label="anchor" href="#getting-help-with-bioconductor-packages"></a>
</h2>
<hr class="half-width">
<p>Help about Bioconductor packages and best practices is available in
several places. Often, the best source of help depends on the situation
at hand:</p>
<ul>
<li>Are you trying to identify the best package for a particular
task?</li>
<li>Are you trying to use a package for the very first time?</li>
<li>Are you unsure about best practices to use and combine multiple
packages and functions in a sensible workflow for a particular type of
analysis?</li>
<li>Is a function throwing an error when you apply it to your data?</li>
<li>Do you have questions about the theory or methodology implemented in
a particular package or function?</li>
</ul>
<p>In the next sections, we describe different sources of help available
to Bioconductor users, and situations where each of them are most
useful.</p>
</section><section><h2 class="section-heading" id="the-bioconductor-website">The Bioconductor website<a class="anchor" aria-label="anchor" href="#the-bioconductor-website"></a>
</h2>
<hr class="half-width">
<p>The main <a href="https://www.bioconductor.org/" class="external-link">Bioconductor
website</a> provides a host of resources, all freely available without
even the need to install R or any Bioconductor package.</p>
<p>In particular, the <a href="https://www.bioconductor.org/packages/release/BiocViews.html" class="external-link">biocViews</a>
page is a great way to thematically explore the collection of packages
and identify Bioconductor packages providing a certain
functionality.</p>
<p>Furthermore, the website also collects materials from <a href="https://bioconductor.org/help/course-materials/" class="external-link">courses and
conferences</a>, including presentations, video recordings, and teaching
materials.</p>
<p>By nature, individual presentations and training materials are often
tied to a specific version of Bioconductor packages. As such, they
provide a snapshot of best practices at a particular point in time, and
may become outdated over time after successive Bioconductor releases.
Thus, it is important check the version of packages demonstrated in
teaching materials matches that in the user’s R library. Alternatively,
when referring to materials that employ package versions different from
those in the user’s library, it is important to carefully interpret any
discrepancy between the expected and actual results.</p>
</section><section><h2 class="section-heading" id="package-landing-pages">Package landing pages<a class="anchor" aria-label="anchor" href="#package-landing-pages"></a>
</h2>
<hr class="half-width">
<p>Each package accepted in the Bioconductor project is granted a
landing page on the main Bioconductor website, e.g. <em><a href="https://bioconductor.org/packages/3.19/S4Vectors" class="external-link">S4Vectors</a></em>.</p>
<p>Package landing pages contains useful information that can be
consulted without the need to install the package itself. This is
particularly useful while browsing the Bioconductor repository is search
of packages suitable for a specific task.</p>
<p>On the landing page, prospective users can find a short description
of the package functionality, and links to <a href="reference.html#vignette">package vignettes</a>. Package vignettes
available on the Bioconductor website are written by developers to
demonstrate how the functions available in the package are meant to be
used and combined into a complete workflow. Often, vignettes use
standard data sets preprocessed and freely available from public
repositories, including <a href="reference.html#experimentdata-package">ExperimentData</a> packages
or the Bioconductor <em><a href="https://bioconductor.org/packages/3.19/ExperimentHub" class="external-link">ExperimentHub</a></em>.</p>
<p>In the “Details” section of the landing page, many packages provide a
field labelled “BugReports”. That field provides a URL that users can
visit to report bugs to the package developer(s).</p>
<div id="note" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="note" class="callout-inner">
<h3 class="callout-title">Note</h3>
<div class="callout-content">
<p>It can be difficult to distinguish actual software bugs from
unwitting mistakes made by users not fully familiar with the package
yet. Later in this episode, we provide advice for reporting bugs and
including sufficient information to receive the fastest and most helpful
responses.</p>
<p>In doubt, the <a href="https://support.bioconductor.org/" class="external-link">Biocondutor
support site</a> can also be a great place to discuss individual
experiences and share knowledge about packages and best practices.</p>
</div>
</div>
</div>
<p>Additionally, the landing page provides many other pieces of
information, from daily build reports indicating whether the package
passed all tests on a range of operating systems, to software
dependencies indicating the number of other Bioconductor packages that
must be installed before the package itself can be installed and used on
the user’s system.</p>
<div id="going-further" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="going-further" class="callout-inner">
<h3 class="callout-title">Going further</h3>
<div class="callout-content">
<p>Each package has a landing page for each release of Bioconductor
since the package was added to the repository, e.g.:</p>
<ul>
<li><a href="https://www.bioconductor.org/packages/3.14/bioc/html/BiocVersion.html" class="external-link">https://www.bioconductor.org/packages/3.14/bioc/html/BiocVersion.html</a></li>
<li><a href="https://www.bioconductor.org/packages/3.13/bioc/html/BiocVersion.html" class="external-link">https://www.bioconductor.org/packages/3.13/bioc/html/BiocVersion.html</a></li>
<li><a href="https://www.bioconductor.org/packages/3.12/bioc/html/BiocVersion.html" class="external-link">https://www.bioconductor.org/packages/3.12/bioc/html/BiocVersion.html</a></li>
</ul>
<p>In the URL of a package landing page, we can replace the version
number by the word “release” or “devel” to access the landing page of
the latest stable release or development version, respectively.</p>
<ul>
<li><a href="https://www.bioconductor.org/packages/release/bioc/html/BiocPkgTools.html" class="external-link">https://www.bioconductor.org/packages/release/bioc/html/BiocPkgTools.html</a></li>
<li><a href="https://www.bioconductor.org/packages/devel/bioc/html/BiocPkgTools.html" class="external-link">https://www.bioconductor.org/packages/devel/bioc/html/BiocPkgTools.html</a></li>
</ul>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="package-vignettes">Package vignettes<a class="anchor" aria-label="anchor" href="#package-vignettes"></a>
</h2>
<hr class="half-width">
<p>Each Bioconductor package is required to include at least one
vignette. Many packages have more than one vignette, often separating
core functionality from specific use cases.</p>
<p>As we noted earlier in this episode, vignettes are available from <a href="..html#package-landing-pages">package landing pages</a> on the
Bioconductor website. However, the landing page only links to the
documentation of the most recent version of the package for each version
of Bioconductor. This may be a different version from the one that is
installed in the user’s R library and used in the R session.</p>
<p>When Bioconductor packages are installed in the user’s R library, the
vignettes associated with that particular version of the package are
also installed on the user’s computer. Those locally installed vignettes
are the gold standard reference for the version of the package that is
currently installed in the R library and used in the R session. They can
be accessed using the function <code>browseVignettes()</code>, for
instance:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">browseVignettes</span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span></code></pre>
</div>
<p>Specifically, the function <code>browseVignettes()</code> opens a
local web page in the user’s default web browser, listing all the
vignettes available for the requested package. Each vignette is
available in three formats:</p>
<ul>
<li>precompiled, in PDF or HTML format</li>
<li>source, in Sweave or R markdown format</li>
<li>as an R script</li>
</ul>
<p>The precompiled format is often the most comfortable format to read,
as the PDF and HTML formats allow the contents of the documents to be
preview in one integrated view. This includes plain text explanations,
as well as code and their outputs, both figures and console
messages.</p>
</section><section><h2 class="section-heading" id="package-help-pages">Package help pages<a class="anchor" aria-label="anchor" href="#package-help-pages"></a>
</h2>
<hr class="half-width">
<p>Bioconductor requires every user-facing package function to be
documented in one of the package help pages (often referred to as “man
pages”, after the name of the package sub-directory where they are
stored).</p>
<p>Help pages can be accessed using the <code>help()</code> function or
the question mark symbol <code>?</code>.</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">help</span><span class="op">(</span>topic <span class="op">=</span> <span class="st">"install"</span>, package <span class="op">=</span> <span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="op">?</span><span class="fu">BiocManager</span><span class="fu">::</span><span class="va">install</span></span></code></pre>
</div>
<p>Help pages for Bioconductor packages generally follow the same rules
as CRAN packages when it comes to formatting and essential contents.
However, Bioconductor also requires that most of the man pages
documenting exported objects must have runnable examples. Runnable
examples are particularly helpful to demonstrate the usage of individual
functions on small data sets immediately available to users - either
artificially simulated on the fly or programmatically imported from
public data repositories.</p>
<p>In particular, runnable examples demonstrate the usage of functions
in ideal cases, showcasing how the inputs inputs of the function should
be formatted, and what information will be available to users in the
outputs of the function. Running those examples and comparing the
example inputs and outputs with the user’s own data often provide
significant insights into the transformations that are needed before
applying the function to the user’s own data, and how the outputs of the
function can be interpreted and interacted with.</p>
</section><section><h2 class="section-heading" id="the-bioconductor-support-site">The Bioconductor support site<a class="anchor" aria-label="anchor" href="#the-bioconductor-support-site"></a>
</h2>
<hr class="half-width">
<p>The <a href="https://support.bioconductor.org/" class="external-link">Bioconductor support
site</a> provides a platform for the community of users and developers
to ask questions, and help each other through the doubts and challenges
related to Bioconductor packages and analytical workflows.</p>
<p>The support site can be freely browsed without an account to search
and read the many questions that were already asked and answered.
However, posting a new question does require an account on the platform.
Signing up to the platform is straightforward using an email address or
Open Authorization from a number of trusted providers.</p>
<p>A system of upvoting allows the most popular answers to feature more
prominently at the top of each page. Furthermore, the original poster
retains the right to mark one answer as the one that resolved their
issue.</p>
<p>Separately, a system of points granted to each user for providing
answers either popular or accepted by the original poster highlights the
most active and trusted contributors on the platform.</p>
<figure><img src="../fig/bioc-support-site.png" alt="The Bioconductor support site." class="figure mx-auto d-block"></figure><p><strong>The Bioconductor support site.</strong> The Bioconductor
support site tracks questions and answers posted by registered users.
The platform can be freely browsed and searched by non-registered
users.</p>
</section><section><h2 class="section-heading" id="workflow-packages">Workflow packages<a class="anchor" aria-label="anchor" href="#workflow-packages"></a>
</h2>
<hr class="half-width">
<p>Bioconductor workflow packages are special in the way that they are
only expected to contain vignettes, without any additional code or
functionality of their own. Instead, the vignettes of workflow packages
exclusively import functionality from other packages, and demonstrate
how to combine functions from those packages into an integrated workflow
that users are likely to face in their day-to-day work.</p>
<p>Like regular vignettes, data is typically fetched from publicly
available sources, including Bioconductor <a href="reference.html#experimentdata-package">ExperimentData</a> packages
or the Bioconductor <em><a href="https://bioconductor.org/packages/3.19/ExperimentHub" class="external-link">ExperimentHub</a></em>.
Those freely available standard data sets allow users to interactively
reproduce outputs while they read and follow along the vignette.</p>
<p>Workflow packages can be browsed in a dedicated section of the <a href="https://www.bioconductor.org/packages/release/BiocViews.html#___Workflow" class="external-link">biocViews
page</a>.</p>
</section><section><h2 class="section-heading" id="slack-workspace">Slack workspace<a class="anchor" aria-label="anchor" href="#slack-workspace"></a>
</h2>
<hr class="half-width">
<p>The <a href="https://community-bioc.slack.com/" class="external-link">Bioconductor Slack
workspace</a> was created in 2016.</p>
<p>The workspace can be freely joined using this <a href="https://bioc-community.herokuapp.com/" class="external-link">Heroku app</a> to generate
invitations for individual email addresses.</p>
<p>The Bioconductor Slack workspace is a lively online platform for
official announcements by the <a href="https://www.bioconductor.org/about/core-team/" class="external-link">Bioconductor Core
Team</a> (e.g., Bioconductor release, conferences &amp; events), as well
as informal discussions between groups of users subscribed to thematic
channels, and direct messages between community members.</p>
<p>The workspace features a large number of channels dedicated to
particular topics and areas of interest in the community. Those channels
range from active fields of research (e.g., single-cell genomics), to
time-limited events (e.g., conferences), but also community outreach
(e.g., diversity and representation).</p>
<p>Private channels also exist for governance (e.g., event organisation,
advisory boards).</p>
<div id="note-1" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="note-1" class="callout-inner">
<h3 class="callout-title">Note</h3>
<div class="callout-content">
<p>The Slack workspace allows for rapid day-to-day communication and
discussion with fellow community members through channels and direct
messages.</p>
<p>However, popular channels can reach up to hundreds of users in many
different time zones and should be used with parsimony and mindfulness.
Conversely, direct messages and private channels are limited to the
users invited in the discussion, and any outcome relevant to the
community then needs to be re-posted in a public channel.</p>
<p>As a consequence, the <a href="https://support.bioconductor.org/" class="external-link">Bioconductor support site</a>
remains the preferred way to publicly ask questions of interest to the
community, in a way that both the question, discussion, and answers are
easily searchable and indexed by major search engines.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="how-to-efficiently-ask-for-help">How to efficiently ask for help<a class="anchor" aria-label="anchor" href="#how-to-efficiently-ask-for-help"></a>
</h2>
<hr class="half-width">
<p>Most community members provide help voluntarily, on their own spare
time and without any form of compensation. As such, it is important to
ask questions as clearly as possible, providing relevant information, to
help those volunteers identify the source of the issue as rapidly as
possible, thus saving time for both the helper and the original
poster.</p>
<p>Depending on the question, some key information includes:</p>
<ul>
<li>Operating system</li>
<li>Version of R</li>
<li>Version of Bioconductor</li>
<li>Version of individual packages installed in the R library</li>
<li>Version of individual packages attached to the current session</li>
<li>Third-party software, libraries, and compilers installed on the
user’s system</li>
<li>Source of packages installed in the R library (e.g., Bioconductor,
CRAN, GitHub)</li>
<li>Code executed in the R session leading up to the issue</li>
<li>Global options active in the session (accessible using
<code>options()</code>)</li>
</ul>
<p>When the issue relates to code being run and producing unexpected
outputs, it is paramount to include sufficient information for others to
reproduce the issue on their own computer. Indeed, many issues require a
live R session to properly investigate the source of the issue, test
fixes, or provide workaround and advice.</p>
<p>Crucially, when providing code as part of your post, it is important
that this code be executable by readers, including data that are
processed by the code. Often, the code itself may look correct, while
the issue relates to the interaction between the code and a particular
data set. If sharing sensitive data is not an option, then the issue
should be reformulated and presented using a data set publicly available
on the internet, or including code to generate simulated data randomly
generated in a reproducible way (e.g., <code>set.seed()</code>).</p>
<p>One option is to use the package <em><a href="https://CRAN.R-project.org/package=reprex" class="external-link">reprex</a></em> to
collate the code and outputs that describe the issue into formatted text
that is easy to post of many online forums, including the <a href="https://support.bioconductor.org/" class="external-link">Bioconductor support
site</a>.</p>
<p>Finally, the <a href="https://support.bioconductor.org/" class="external-link">Bioconductor
support site</a> is the preferred platform to post questions related to
Bioconductor packages. This is because questions are visible to the
entire community, including many experienced Bioconductor users who
regularly answer those questions, and other users who can find answers
to questions that were already posted and resolved by the time they run
into the issue themselves.</p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>The <code>browseVignettes()</code> function is recommended to access
the vignette(s) installed with each package.</li>
<li>Vignettes can also be accessed on the Bioconductor website, but
beware of differences between package versions!</li>
<li>The Bioconductor main website contains general information, package
documentation, and course materials.</li>
<li>The Bioconductor support site is the recommended place to contact
developers and ask questions.</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-05-s4"><p>Content from <a href="05-s4.html">BioconductorにおけるS4クラス</a></p>
<hr>
<p>Last updated on 2025-07-22 |

        <a href="https://github.com/carpentries-incubator/bioc-project/edit/main/episodes/05-s4.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> NA minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>S4クラスシステムとは？</li>
<li>BioconductorはS4クラスをどのように使用しますか？</li>
<li>Bioconductorの<code>DataFrame</code>は、基本の<code>data.frame</code>とどう違うのですか？</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>S4クラス、ジェネリックおよびメソッドとは何かを説明してください。</li>
<li>Bioconductorパッケージインフラストラクチャの中心にあるS4クラスを特定します。</li>
<li>さまざまなS4オブジェクトを作成し、関連するS4メソッドを適用します。</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="パッケージのインストール">パッケージのインストール<a class="anchor" aria-label="anchor" href="#%E3%83%91%E3%83%83%E3%82%B1%E3%83%BC%E3%82%B8%E3%81%AE%E3%82%A4%E3%83%B3%E3%82%B9%E3%83%88%E3%83%BC%E3%83%AB"></a>
</h2>
<hr class="half-width">
<p>以下のセクションに進む前に、必要なBioconductorパッケージをいくつかインストールします。
最初に、<em><a href="https://bioconductor.org/packages/3.19/BiocManager" class="external-link">BiocManager</a></em>パッケージがインストールされているかを確認し、それを使用しようとする前にインストールします。
次に、<code>BiocManager::install()</code>関数を使用して必要なパッケージをインストールします。</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu">requireNamespace</span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu">install.packages</span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"S4Vectors"</span><span class="op">)</span></span></code></pre>
</div>
<div id="講師のために" class="callout prereq">
<div class="callout-square">
<i class="callout-icon" data-feather="check"></i>
</div>
<div id="講師のために" class="callout-inner">
<h3 class="callout-title">講師のために</h3>
<div class="callout-content">
<p>このエピソードの最初の部分は、少し理論的に見えるかもしれません。
S4クラスシステムの内部動作に関する詳細に深入りしないでください（例えば、<code>new()</code>関数に言及する必要はありませんし、クラスを作成する具体的なコード例を示す必要もありません）。
代わりに、最初の図のキャプションでは、技術用語を単純な文で導入しながら、図を通じて段階的に説明し、S4クラスシステムの中心となるメソッドディスパッチの概念に至り、初心者ユーザーにとっての混乱の原因となることを示しています。</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="s4クラスとメソッド">S4クラスとメソッド<a class="anchor" aria-label="anchor" href="#s4%E3%82%AF%E3%83%A9%E3%82%B9%E3%81%A8%E3%83%A1%E3%82%BD%E3%83%83%E3%83%89"></a>
</h2>
<hr class="half-width">
<div class="section level3">
<h3 id="メソッドパッケージ">メソッドパッケージ<a class="anchor" aria-label="anchor" href="#%E3%83%A1%E3%82%BD%E3%83%83%E3%83%89%E3%83%91%E3%83%83%E3%82%B1%E3%83%BC%E3%82%B8"></a>
</h3>
<p>S4クラスシステムは、基本パッケージ<a href="https://stat.ethz.ch/R-manual/R-devel/library/methods/html/00Index.html" class="external-link">methods</a>に実装されています。
そのため、この概念はBioconductorプロジェクトに特有なものではなく、さまざまな独立したパッケージにも存在します。
この主題は、Hadley Wickhamによるオンライン書籍<a href="https://adv-r.had.co.nz/S4.html" class="external-link">Advanced
R</a>で詳しく文書化されています。
ほとんどのBioconductorユーザーは、S4クラスシステムの詳細を過度に熟知する必要がないでしょう。
むしろ、Bioconductorプロジェクトにおけるパッケージの効率的な使用の鍵は、S4クラスシステムを使用する重要な動機と、ユーザー向けの機能に関するベストプラクティス（クラス、ジェネリック、メソッドを含む）を十分に理解することにあります。
このエピソードの次のセクションでは、BioconductorプロジェクトにおけるS4クラスとメソッドの基本的な機能性とユーザーエクスペリエンスに焦点を当てます。</p>
<p>一方で、<a href="reference.html#s4-class">S4クラス</a>は、Rセッション内の変数名に割り当てることができる計算オブジェクトにおいて、任意に複雑な情報を格納できるデータ構造を提供します。
もう一方で、<a href="reference.html#s4-method">S4ジェネリックおよびメソッド</a>は、それらのオブジェクトを処理するために適用される関数を定義します。</p>
<p>これまでの数年で、BioconductorプロジェクトはS4クラスシステムを使用して、ほとんどの生物学的アッセイ（生アッセイデータや処理済みアッセイデータ、個別の特徴やサンプルに関する実験メタデータ、関連する他のアッセイ特有の情報を含む）のデータを格納および処理できるいくつかのクラスとメソッドを開発してきました。
Bioconductorパッケージ全体で一般的に使用される標準S4クラスに慣れることは、分析ワークフローを開発したいユーザーがベストプラクティスに従う自信を高めるための重要なステップです。</p>
<figure><img src="../fig/bioc-s4.svg" alt="S4クラス、ジェネリックおよびメソッド。" class="figure mx-auto d-block"></figure><p><strong>S4クラス、ジェネリック、およびメソッド。</strong>
左側には、<code>S4Class1</code>と<code>S4Class2</code>という2つの例のクラスが、継承の概念を示しています。
クラス<code>S4Class1</code>には、データを格納するための<code>SlotName1</code>と<code>SlotName2</code>という2つのスロットがあります。
これらの2つのスロットは、それぞれ<code>SlotType1</code>および<code>SlotType2</code>型のオブジェクトを格納するように制限されています。
このクラスは、オブジェクトが更新されるたびにデータの整合性を確認する有効性ルールも定義しています。
クラス<code>S4Class2</code>は、<code>S4Class1</code>からすべてのスロットと有効性ルールを継承し、新しいスロット<code>SlotName3</code>と新しい有効性ルールを定義します。
例のコードは、それぞれのクラスのオブジェクトが通常、対応するクラスにちなんで命名されたコンストラクタ関数を使用してどのように作成されるかを示しています。
右側には、1つのジェネリック関数と2つのメソッドが多様性の概念とS4メソッドディスパッチのプロセスを示しています。
ジェネリック関数<code>S4Generic1()</code>は、関数の名前とその引数を定義します。
ただし、その関数の実装は提供されていません。
代わりに、2つのメソッドが定義され、それぞれ特定のクラスの入力に対して単純な実装を提供します。
具体的には、最初のメソッドは、<code>S4Class1</code>のオブジェクトが引数<code>x</code>として与えられた場合の<code>S4Generic1()</code>の実装を定義し、2番目のメソッドは、<code>S4Class2</code>のオブジェクトが引数<code>x</code>として与えられた場合の<code>S4Generic1()</code>の別の実装を提供します。
ジェネリック関数<code>S4Generic1()</code>が呼び出されると、<a href="reference.html#s4-method-dispatch">メソッドディスパッチ</a>と呼ばれるプロセスが発生し、渡されたオブジェクトのクラスに応じて<code>S4Generic1()</code>メソッドの適切な実装が呼び出されます。</p>
</div>
<div class="section level3">
<h3 id="スロットと有効性">スロットと有効性<a class="anchor" aria-label="anchor" href="#%E3%82%B9%E3%83%AD%E3%83%83%E3%83%88%E3%81%A8%E6%9C%89%E5%8A%B9%E6%80%A7"></a>
</h3>
<p>S3クラスシステムは基本Rに直接用意されているのに対し（このレッスンでは説明されていません）、S4クラスシステムは、Rにおけるオブジェクト指向プログラミング（OOP）のためのクラスとメソッドのより厳格な定義を提供します。
OOPモデルを実装している多くのプログラミング言語と同様に、S4クラスは実世界のエンティティを、情報を<code>*slots*</code>と呼ばれる1つ以上の内部コンポーネントに格納する計算オブジェクトとして表現するために使用されます。
クラスの定義は、各スロットに格納される可能性のあるデータのタイプを明示します。不適切なデータを格納しようとした場合にはエラーが発生します。
さらに、クラスの定義には、型を超えてオブジェクトに格納されるデータの有効性を確認するコードを含めることもできます。
例えば、<code>numeric</code>型のスロットを使用して人の年齢を格納することはできますが、有効性メソッドは格納される値が実際に正であることを確認できます。</p>
</div>
<div class="section level3">
<h3 id="継承">継承<a class="anchor" aria-label="anchor" href="#%E7%B6%99%E6%89%BF"></a>
</h3>
<p>OOPモデルの中核的な柱の1つは、既存のクラスの機能を継承して拡張できる新しいクラスを開発する可能性です。
S4クラスシステムはこのパラダイムを実装しています。</p>
<p>新しいS4クラスの定義は、継承する他のクラスの名前を宣言します。
新しいクラスには、親クラスのすべてのスロットが含まれ、新しいクラスの定義に追加される新しいスロットが含まれます。</p>
<p>新しいクラスの定義では、新しい有効性チェックも定義でき、親クラスの各クラスに実装された各有効性チェックに追加されます。</p>
</div>
<div class="section level3">
<h3 id="ジェネリックおよびメソッド">ジェネリックおよびメソッド<a class="anchor" aria-label="anchor" href="#%E3%82%B8%E3%82%A7%E3%83%8D%E3%83%AA%E3%83%83%E3%82%AF%E3%81%8A%E3%82%88%E3%81%B3%E3%83%A1%E3%82%BD%E3%83%83%E3%83%89"></a>
</h3>
<p>クラスが情報を格納するデータ構造を定義する一方で、ジェネリックおよびメソッドは、それらのクラスからインスタンス化されたオブジェクトに適用される関数を定義します。</p>
<p>S4ジェネリック関数は、いくつかの重要な引数に依存して、異なる振る舞いが期待される関数の名前を宣言するために使用されます。
その代わりに、S4メソッドは、各特定の入力の組み合わせに対してジェネリック関数の異なる実装を定義するために使用されます。</p>
<p>ジェネリック関数が呼び出され、S4オブジェクトが渡されると、<a href="reference.html#s4-method-dispatch">メソッドディスパッチ</a>と呼ばれるプロセスが発生し、オブジェクトのクラスが実行すべき適切なメソッドを決定するために使用されます。</p>
</div>
</section><section><h2 class="section-heading" id="s4vectorsパッケージ">S4Vectorsパッケージ<a class="anchor" aria-label="anchor" href="#s4vectors%E3%83%91%E3%83%83%E3%82%B1%E3%83%BC%E3%82%B8"></a>
</h2>
<hr class="half-width">
<p><em><a href="https://bioconductor.org/packages/3.19/S4Vectors" class="external-link">S4Vectors</a></em>パッケージは、<code>Vector</code>および<code>List</code>の仮想クラスと、Rの通常のベクトルおよびリストのセマンティクスを拡張する一連のジェネリック関数を定義しています。
S4クラスシステムを使用することで、パッケージ開発者は<code>Vector</code>や<code>List</code>の具体的なサブクラスとして、ベクトルやリストに似たオブジェクトを簡単に実装できます。</p>
<p>仮想クラス（<code>Vector</code>と<code>List</code>など）は、オブジェクト自体としてインスタンス化することはできません。
むしろ、これらの仮想クラスは、それらから派生したすべての具体的なクラスが継承する基本機能を提供します。</p>
<p>代わりに、一般的に興味のあるいくつかの低レベルの具体的なサブクラス（例えば、<code>DataFrame</code>、<code>Rle</code>、<code>Hits</code>）は、<em><a href="https://bioconductor.org/packages/3.19/S4Vectors" class="external-link">S4Vectors</a></em>パッケージ自体に実装されています。また、Bioconductorプロジェクト全体の他のパッケージ（例えば、<em><a href="https://bioconductor.org/packages/3.19/IRanges" class="external-link">IRanges</a></em>）にも、多くの他のパッケージが実装されています。</p>
<p>次のように、パッケージを現在のRセッションにアタッチします。</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">S4Vectors</span><span class="op">)</span></span></code></pre>
</div>
<div id="注意" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="注意" class="callout-inner">
<h3 class="callout-title">注意</h3>
<div class="callout-content">
<p>パッケージをセッションにアタッチした場合、コンソールに印刷されるパッケージ起動メッセージには、<em><a href="https://bioconductor.org/packages/3.19/S4Vectors" class="external-link">S4Vectors</a></em>パッケージが、セッションにアタッチされると、<a href="https://stat.ethz.ch/R-manual/R-devel/library/base/html/00Index.html" class="external-link">base</a>パッケージから数機能をマスクすることが記載されています。
これは、<em><a href="https://bioconductor.org/packages/3.19/S4Vectors" class="external-link">S4Vectors</a></em>パッケージが、これらの関数の実装を含んでいることを意味し、Rセッションにアタッチされた最新のパッケージであるため、その関数の独自の実装がR検索パス上で最初に見つかり、<a href="https://stat.ethz.ch/R-manual/R-devel/library/base/html/00Index.html" class="external-link">base</a>パッケージの元の実装ではなく、使用されます。</p>
<p>多くの場合、マスクされた関数は、特に問題なく以前のように使用できます。
時折、パッケージ名に加えて関数名を使ってマスクされた関数を呼び出す必要があるかもしれません。例：<code>base::anyDuplicated()</code>。</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="dataframeクラス">DataFrameクラス<a class="anchor" aria-label="anchor" href="#dataframe%E3%82%AF%E3%83%A9%E3%82%B9"></a>
</h2>
<hr class="half-width">
<div class="section level3">
<h3 id="長方形データの概念への拡張">長方形データの概念への拡張<a class="anchor" aria-label="anchor" href="#%E9%95%B7%E6%96%B9%E5%BD%A2%E3%83%87%E3%83%BC%E3%82%BF%E3%81%AE%E6%A6%82%E5%BF%B5%E3%81%B8%E3%81%AE%E6%8B%A1%E5%BC%B5"></a>
</h3>
<p><em><a href="https://bioconductor.org/packages/3.19/S4Vectors" class="external-link">S4Vectors</a></em>パッケージで実装された<code>DataFrame</code>クラスは、基本Rの<code>data.frame</code>クラスやtidyverseの<code>tibble</code>に慣れているユーザーに馴染みのある長方形データの概念を拡張しています。
具体的には、<code>DataFrame</code>は、カラムとしてあらゆるタイプのオブジェクト（<code>length</code>および<code>[</code>メソッドを持つ）を格納することをサポートしています。</p>
<p>全体として、<code>DataFrame</code>クラスは、構築、部分的選択、分割、結合などの観点から、非常に似た形で振る舞うS4クラスの正式な定義を提供します。</p>
<p>新しいオブジェクトを作成するには、<code>DataFrame()</code>コンストラクタ関数を使用する必要があります。これは、基本Rの<code>data.frame()</code>に相当します。
その関数のヘルプページは、<code>?DataFrame</code>でアクセスできますので、詳細情報を参照できます。</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DF1</span> <span class="op">&lt;-</span> <span class="fu">DataFrame</span><span class="op">(</span></span>
<span>    Integers <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">1L</span>, <span class="fl">2L</span>, <span class="fl">3L</span><span class="op">)</span>,</span>
<span>    Letters <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"C"</span><span class="op">)</span>,</span>
<span>    Floats <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">1.2</span>, <span class="fl">2.3</span>, <span class="fl">3.4</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">DF1</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DataFrame with 3 rows and 3 columns
   Integers     Letters    Floats
  &lt;integer&gt; &lt;character&gt; &lt;numeric&gt;
1         1           A       1.2
2         2           B       2.3
3         3           C       3.4</code></pre>
</div>
<p>実際、<code>DataFrame</code>オブジェクトは、同等の<code>data.frame</code>オブジェクトに簡単に変換できます。</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df1</span> <span class="op">&lt;-</span> <span class="fu">as.data.frame</span><span class="op">(</span><span class="va">DF1</span><span class="op">)</span></span>
<span><span class="va">df1</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>  Integers Letters Floats
1        1       A    1.2
2        2       B    2.3
3        3       C    3.4</code></pre>
</div>
<p>逆に、<code>as()</code>関数を使用して<code>data.frame</code>オブジェクトを<code>DataFrame</code>に変換することもできます。</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">as</span><span class="op">(</span><span class="va">df1</span>, <span class="st">"DataFrame"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DataFrame with 3 rows and 3 columns
   Integers     Letters    Floats
  &lt;integer&gt; &lt;character&gt; &lt;numeric&gt;
1         1           A       1.2
2         2           B       2.3
3         3           C       3.4</code></pre>
</div>
</div>
<div class="section level3">
<h3 id="基本のdata-frameとの違い">基本のdata.frameとの違い<a class="anchor" aria-label="anchor" href="#%E5%9F%BA%E6%9C%AC%E3%81%AEdata-frame%E3%81%A8%E3%81%AE%E9%81%95%E3%81%84"></a>
</h3>
<p>最も注目すべき例外は、行名の処理に関するものです。
まず、行名はオプションです。
これは、行名がない場合、<code>rownames(x)</code>が<code>NULL</code>を返すことを意味します。</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">rownames</span><span class="op">(</span><span class="va">DF1</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="cn">NULL</span></span></code></pre>
</div>
<p>これは、<code>data.frame</code>とは異なり、<code>rownames(x)</code>が<code>as.character(seq_len(nrow(x)))</code>の等価物を返すことになります。</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">rownames</span><span class="op">(</span><span class="va">df1</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "1" "2" "3"</code></pre>
</div>
<p>ただし、<code>NULL</code>を返すことは、例えば、組み合わせ関数に対して、行名が不要であることを通知します（大規模なデータを扱うときには、しばしば贅沢です）。</p>
<p>さらに、<code>DataFrame</code>オブジェクトの行名は一意である必要はありません。これは、基本Rの<code>data.frame</code>とは対照的です。
行名は、長方形データで観測結果を一意に特定してインデックス付けするために使用されるため、しばしば議論の原因となります。
設定されている場合、行名は<code>[</code>演算子を使用して長方形データを部分選択するために使用できます。
非一意の行名はその目的を妨げ、選択された各行名の最初の出現だけが抽出されるため、予期しない結果をもたらす可能性があります。
その代わりに、tidyverseの<code>tibble</code>は、行名を完全に設定する機能を排除し、ユーザーに特定のカラムに情報を明示的にストアさせるように強制しています。また、行を効率的にフィルタリングするための関数を提供し、<code>[</code>演算子を使用する必要はありません。</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DF2</span> <span class="op">&lt;-</span> <span class="fu">DataFrame</span><span class="op">(</span></span>
<span>    Integers <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">1L</span>, <span class="fl">2L</span>, <span class="fl">3L</span><span class="op">)</span>,</span>
<span>    Letters <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"C"</span><span class="op">)</span>,</span>
<span>    Floats <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">1.2</span>, <span class="fl">2.3</span>, <span class="fl">3.4</span><span class="op">)</span>,</span>
<span>    row.names <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"name1"</span>, <span class="st">"name1"</span>, <span class="st">"name2"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">DF2</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DataFrame with 3 rows and 3 columns
       Integers     Letters    Floats
      &lt;integer&gt; &lt;character&gt; &lt;numeric&gt;
name1         1           A       1.2
name1         2           B       2.3
name2         3           C       3.4</code></pre>
</div>
<div id="チャレンジ" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="チャレンジ" class="callout-inner">
<h3 class="callout-title">チャレンジ</h3>
<div class="callout-content">
<p>上記の例を使用して、<code>DF2["name1", ]</code>は何を返しますか？
なぜですか？</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  ソリューション
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<pre><code>&gt; DF2["name1", ]
DataFrame with 1 row and 3 columns
       Integers     Letters    Floats
      &lt;integer&gt; &lt;character&gt; &lt;numeric&gt;
name1         1           A       1.2</code></pre>
<p>行名<code>name1</code>に一致した行の最初の出現のみが返されます。</p>
<p>この場合、行名には特別な意味がなく、それらの必要性を正当化することが難しいです。
代わりに、ユーザーは、<code>DF2[rownames(DF2) == "name1", ]</code>のように、行名<code>name1</code>に一致するすべての行をより明示的に抽出することができます。</p>
<p>ユーザーは、特定の状況で行名を使用する動機、何を表し、分析中にどのように使用されるべきかを意識するべきです。</p>
</div>
</div>
</div>
</div>
<p>最後に、<code>DataFrame</code>の行名は、部分一致による選択をサポートしません。これは、基本の<code>data.frame</code>とは対照的です。
<code>DataFrame</code>のより厳格な動作により、予期しない結果が生じることは、意図しないユーザーによって防止されます。</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DF3</span> <span class="op">&lt;-</span> <span class="fu">DataFrame</span><span class="op">(</span></span>
<span>    Integers <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">1L</span>, <span class="fl">2L</span>, <span class="fl">3L</span><span class="op">)</span>,</span>
<span>    Letters <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"C"</span><span class="op">)</span>,</span>
<span>    Floats <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">1.2</span>, <span class="fl">2.3</span>, <span class="fl">3.4</span><span class="op">)</span>,</span>
<span>    row.names <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"alpha"</span>, <span class="st">"beta"</span>, <span class="st">"gamma"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">df3</span> <span class="op">&lt;-</span> <span class="fu">as.data.frame</span><span class="op">(</span><span class="va">DF3</span><span class="op">)</span></span></code></pre>
</div>
<div id="チャレンジ-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="チャレンジ-1" class="callout-inner">
<h3 class="callout-title">チャレンジ</h3>
<div class="callout-content">
<p>上記の例を使用して、<code>DF3["a", ]</code>と<code>df3["a", ]</code>の出力は何ですか？
なぜ違うのですか？</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">
  ソリューション
  </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<pre><code>&gt; DF3["a", ]
DataFrame with 1 row and 3 columns
      Integers     Letters    Floats
     &lt;integer&gt; &lt;character&gt; &lt;numeric&gt;
&lt;NA&gt;        NA          NA        NA
&gt; df3["a", ]
      Integers Letters Floats
alpha        1       A    1.2</code></pre>
<p><code>DataFrame</code>オブジェクトは、部分一致による行名のマッチングを実行せず、そのため、行がマッチせず、<code>NA</code>値のいっぱいの<code>DataFrame</code>を返しました。
その代わりに、<code>data.frame</code>オブジェクトは部分一致による行名マッチングを実行し、要求された<code>"a"</code>を<code>"alpha"</code>行名に一致させ、新しい<code>data.frame</code>オブジェクトとして該当行を返しました。</p>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="インデックス付け">インデックス付け<a class="anchor" aria-label="anchor" href="#%E3%82%A4%E3%83%B3%E3%83%87%E3%83%83%E3%82%AF%E3%82%B9%E4%BB%98%E3%81%91"></a>
</h3>
<p>通常の<code>data.frame</code>と同様に、カラムは<code>$</code>、<code>[</code>、および<code>[[</code>を使用してアクセスできます。
各演算子には異なる目的があり、最も適切なものは、達成しようとしていることによく依存します。</p>
<p>例えば、ドル演算子<code>$</code>は、名前で単一列を抽出するために使用できます。
それはしばしばベクターであることが多いですが、それはその列内のデータの性質によって異なる場合もあります。
この演算子は、インタラクティブなRセッションで非常に便利で、利用可能な列名の自動補完を提供します。</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DF3</span><span class="op">$</span><span class="va">Integers</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 1 2 3</code></pre>
</div>
<p>同様に、ダブルブラケット演算子<code>[[</code>も単一の列を抽出するために使用できます。
これは<code>$</code>よりも柔軟性があり、文字列名と整数インデックスの両方を処理できます。</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DF3</span><span class="op">[[</span><span class="st">"Letters"</span><span class="op">]</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "A" "B" "C"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DF3</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "A" "B" "C"</code></pre>
</div>
<p>演算子<code>[</code>は、行と列の同時選択や、単一列の選択が<code>DataFrame</code>または<code>vector</code>として返されるかどうかを制御する際に最も便利です。</p>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DF3</span><span class="op">[</span><span class="fl">2</span><span class="op">:</span><span class="fl">3</span>, <span class="st">"Letters"</span>, drop<span class="op">=</span><span class="cn">FALSE</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DataFrame with 2 rows and 1 column
          Letters
      &lt;character&gt;
beta            B
gamma           C</code></pre>
</div>
</div>
<div class="section level3">
<h3 id="メタデータ列">メタデータ列<a class="anchor" aria-label="anchor" href="#%E3%83%A1%E3%82%BF%E3%83%87%E3%83%BC%E3%82%BF%E5%88%97"></a>
</h3>
<p><code>DataFrame</code>における最も注目すべき新機能は、メタデータを別の<code>DataFrame</code>でカラムに保持できる能力です。</p>
<figure><img src="../fig/dataframe-mcols.svg" alt="メタデータ列。" class="figure mx-auto d-block"></figure><p><strong>メタデータ列。</strong>
メタデータ列は、<code>DataFrame</code>オブジェクトの文脈で示されます。
左側には、<code>A</code>と<code>B</code>という名前の列を持つ<code>DataFrame</code>オブジェクト<code>DF</code>が作成されます。
右側には、<code>DF</code>のメタデータ列を<code>mcols(DF)</code>を使ってアクセスします。
この例では、名前<code>meta1</code>と<code>meta2</code>の2つのメタデータ列が作成されます。
メタデータ列は、親<code>DataFrame</code>の各列に対して1行を含む<code>DataFrame</code>として格納されます。</p>
<p>メタデータ列は、関数<code>mcols()</code>を使用してアクセスされます。
メタデータ列が定義されていない場合、<code>mcols()</code>は単に<code>NULL</code>を返します。</p>
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DF4</span> <span class="op">&lt;-</span> <span class="fu">DataFrame</span><span class="op">(</span></span>
<span>    Integers <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">1L</span>, <span class="fl">2L</span>, <span class="fl">3L</span><span class="op">)</span>,</span>
<span>    Letters <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"C"</span><span class="op">)</span>,</span>
<span>    Floats <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">1.2</span>, <span class="fl">2.3</span>, <span class="fl">3.4</span><span class="op">)</span>,</span>
<span>    row.names <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"alpha"</span>, <span class="st">"beta"</span>, <span class="st">"gamma"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu">mcols</span><span class="op">(</span><span class="va">DF4</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="cn">NULL</span></span></code></pre>
</div>
<p>関数<code>mcols()</code>は、メタデータ列を追加、編集、または削除するためにも使用できます。
例えば、2つの列を持つ<code>DataFrame</code>として、メタデータ列を初期化できます。</p>
<ul>
<li>対応する列に格納されている値のタイプを示す1つの列</li>
<li>対応する列で観察された異なる値の数を示す1つの列</li>
</ul>
<div class="codewrapper sourceCode" id="cb28">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">mcols</span><span class="op">(</span><span class="va">DF4</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">DataFrame</span><span class="op">(</span></span>
<span>    Type <span class="op">=</span> <span class="fu">sapply</span><span class="op">(</span><span class="va">DF4</span>, <span class="va">typeof</span><span class="op">)</span>,</span>
<span>    Distinct <span class="op">=</span> <span class="fu">sapply</span><span class="op">(</span><span class="va">DF4</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span> <span class="fu">length</span><span class="op">(</span><span class="fu">unique</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span> <span class="op">}</span> <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu">mcols</span><span class="op">(</span><span class="va">DF4</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DataFrame with 3 rows and 2 columns
                Type  Distinct
         &lt;character&gt; &lt;integer&gt;
Integers     integer         3
Letters    character         3
Floats        double         3</code></pre>
</div>
<div id="注意-1" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="注意-1" class="callout-inner">
<h3 class="callout-title">注意</h3>
<div class="callout-content">
<p>メタデータ列の行名は、親<code>DataFrame</code>の列名と一致するように自動的に設定され、列とメタデータのペアを明確に示します。</p>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="実行長エンコーディング-rle">実行長エンコーディング (RLE)<a class="anchor" aria-label="anchor" href="#%E5%AE%9F%E8%A1%8C%E9%95%B7%E3%82%A8%E3%83%B3%E3%82%B3%E3%83%BC%E3%83%87%E3%82%A3%E3%83%B3%E3%82%B0-rle"></a>
</h2>
<hr class="half-width">
<div class="section level3">
<h3 id="ベクターの概念への拡張">ベクターの概念への拡張<a class="anchor" aria-label="anchor" href="#%E3%83%99%E3%82%AF%E3%82%BF%E3%83%BC%E3%81%AE%E6%A6%82%E5%BF%B5%E3%81%B8%E3%81%AE%E6%8B%A1%E5%BC%B5"></a>
</h3>
<p><em><a href="https://bioconductor.org/packages/3.19/S4Vectors" class="external-link">S4Vectors</a></em>で実装された<code>DataFrame</code>クラスと同様に、<code>Rle</code>クラスは基本パッケージの<code>rle()</code>関数にS4の拡張を提供します。
具体的には、<code>Rle</code>クラスは、原子ベクターを実行長エンコーディング形式で保存することをサポートしています。</p>
<figure><img src="../fig/rle.svg" alt="実行長エンコーディング。" class="figure mx-auto d-block"></figure><p><strong>実行長エンコーディング。</strong>
実行長エンコーディングの概念は、核酸の配列の例を使用してここに示されています。
エンコードの前に、配列の各位置にある各ヌクレオチドはメモリに明示的に保存されます。
エンコーディング中に、同一のヌクレオチドの連続したランは、ヌクレオチドの同一性とランの長さの2ビットの情報に圧縮されます。</p>
<p>実行長エンコーディングは、同一の情報の頻繁な連続が含まれるベクターのメモリ使用量を劇的に削減できます。
たとえば、実行長エンコーディングの有力な応用は、シーケンシング実験におけるゲノムカバレッジの表現です。ここで、大きなゲノム領域にマッピングされたリードが存在しない場合、長い<code>0</code>の値の連続が発生します。
各個別の値を保存することは、メモリ使用量の観点から非常に非効率的です。
その代わりに、実行長エンコーディングプロセスは、冗長な情報のそのような連続を、同一の情報の任意の長い連続から2つの値、すなわち繰り返される値自体と、それが繰り返される回数に圧縮します。</p>
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">v1</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span>, <span class="fl">2</span>, <span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">rle1</span> <span class="op">&lt;-</span> <span class="fu">Rle</span><span class="op">(</span><span class="va">v1</span><span class="op">)</span></span>
<span><span class="va">rle1</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>numeric-Rle of length 17 with 7 runs
  Lengths: 7 1 1 1 1 1 5
  Values : 0 1 2 3 2 1 0</code></pre>
</div>
</div>
<div class="section level3">
<h3 id="インデクシング">インデクシング<a class="anchor" aria-label="anchor" href="#%E3%82%A4%E3%83%B3%E3%83%87%E3%82%AF%E3%82%B7%E3%83%B3%E3%82%B0"></a>
</h3>
<p>通常の<code>vector</code>と同様に、<code>Rle</code>オブジェクトは<code>[</code>を使用してインデックス付けすることができます。</p>
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rle1</span><span class="op">[</span><span class="fl">2</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>numeric-Rle of length 3 with 1 run
  Lengths: 3
  Values : 0</code></pre>
</div>
</div>
<div class="section level3">
<h3 id="使用法">使用法<a class="anchor" aria-label="anchor" href="#%E4%BD%BF%E7%94%A8%E6%B3%95"></a>
</h3>
<p>ベクターのようなオブジェクトとして、<code>Rle</code>オブジェクトも<code>DataFrame</code>オブジェクトの列として保存でき、他のベクター状のオブジェクトとともに格納できます。</p>
<div class="codewrapper sourceCode" id="cb34">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">v2</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="fu">rep</span><span class="op">(</span><span class="fl">1</span>, <span class="fl">5</span><span class="op">)</span>, <span class="fu">rep</span><span class="op">(</span><span class="fl">2</span>, <span class="fl">5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">rle2</span> <span class="op">&lt;-</span> <span class="fu">Rle</span><span class="op">(</span><span class="va">v2</span><span class="op">)</span></span>
<span><span class="va">DF5</span> <span class="op">&lt;-</span> <span class="fu">DataFrame</span><span class="op">(</span></span>
<span>    vector <span class="op">=</span> <span class="va">v2</span>,</span>
<span>    rle <span class="op">=</span> <span class="va">rle2</span>,</span>
<span>    equal <span class="op">=</span> <span class="va">v2</span> <span class="op">==</span> <span class="va">rle2</span></span>
<span><span class="op">)</span></span>
<span><span class="va">DF5</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DataFrame with 10 rows and 3 columns
      vector   rle equal
   &lt;numeric&gt; &lt;Rle&gt; &lt;Rle&gt;
1          1     1  TRUE
2          1     1  TRUE
3          1     1  TRUE
4          1     1  TRUE
5          1     1  TRUE
6          2     2  TRUE
7          2     2  TRUE
8          2     2  TRUE
9          2     2  TRUE
10         2     2  TRUE</code></pre>
</div>
<div id="さらに進む" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="さらに進む" class="callout-inner">
<h3 class="callout-title">さらに進む</h3>
<div class="callout-content">
<p><code>Rle</code>オブジェクトに関する多くの標準操作は、<code>Rle</code>クラスのヘルプページに文書化されており、<code>?Rle</code>としてアクセス可能で、<em><a href="https://bioconductor.org/packages/3.19/S4Vectors" class="external-link">S4Vectors</a></em>パッケージのビネットにおいても、<code>browseVignettes("S4Vectors")</code>を使ってアクセスできます。</p>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>S4クラスはスロットに情報を保存し、オブジェクトが更新されるたびに情報の有効性を確認します。</li>
<li>S4オブジェクトの継続的な整合性を確保するために、ユーザーはスロットに直接アクセスせず、専用の関数を使用する必要があります。</li>
<li>S4ジェネリックは、与えられたオブジェクトのクラスに応じてメソッドの異なる実装を呼び出します。</li>
<li>S4クラス<code>DataFrame</code>は、<code>data.frame</code>の基本機能を拡張しており、メタデータ列の各列に関する情報を保持できる能力を持っています。</li>
<li>S4クラス<code>Rle</code>は、基本の<code>vector</code>の機能を拡張しており、メモリ効率の良い形式で繰り返しベクターをエンコードする能力を持っています。</li>
</ul>
</div>
</div>
</div>
</div>
</section></section><section id="aio-06-biological-sequences"><p>Content from <a href="06-biological-sequences.html">Working with biological sequences</a></p>
<hr>
<p>Last updated on 2025-07-22 |

        <a href="https://github.com/carpentries-incubator/bioc-project/edit/main/episodes/06-biological-sequences.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> NA minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What is the recommended way to represent biological sequences in
Bioconductor?</li>
<li>What Bioconductor packages provides methods to efficiently process
biological sequences?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Explain how biological sequences are represented in the Bioconductor
project.</li>
<li>Identify Bioconductor packages and methods available to process
biological sequences.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="install-packages">Install packages<a class="anchor" aria-label="anchor" href="#install-packages"></a>
</h2>
<hr class="half-width">
<p>Before we can proceed into the following sections, we install some
Bioconductor packages that we will need. First, we check that the <em><a href="https://bioconductor.org/packages/3.19/BiocManager" class="external-link">BiocManager</a></em>
package is installed before trying to use it; otherwise we install it.
Then we use the <code>BiocManager::install()</code> function to install
the necessary packages.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu">requireNamespace</span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu">install.packages</span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"Biostrings"</span><span class="op">)</span></span></code></pre>
</div>
</section><section><h2 class="section-heading" id="the-biostrings-package-and-classes">The Biostrings package and classes<a class="anchor" aria-label="anchor" href="#the-biostrings-package-and-classes"></a>
</h2>
<hr class="half-width">
<div class="section level3">
<h3 id="why-do-we-need-classes-for-biological-sequences">Why do we need classes for biological sequences?<a class="anchor" aria-label="anchor" href="#why-do-we-need-classes-for-biological-sequences"></a>
</h3>
<p>Biological sequences are arguably some of the simplest biological
entities to represent computationally. Examples include nucleic acid
sequences (e.g., DNA, RNA) and protein sequences composed of amino
acids.</p>
<p>That is because alphabets have been designed and agreed upon to
represent individual monomers using character symbols.</p>
<p>For instance, using the alphabet for amino acids, the reference
protein sequence for the <a href="https://www.uniprot.org/uniprot/P68133#sequences" class="external-link">Actin, alpha
skeletal muscle protein sequence</a> is represented as follows.</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "MCDEDETTALVCDNGSGLVKAGFAGDDAPRAVFPSIVGRPRHQGVMVGMGQKDSYVGDEAQSKRGILTLKYPIEHGIITNWDDMEKIWHHTFYNELRVAPEEHPTLLTEAPLNPKANREKMTQIMFETFNVPAMYVAIQAVLSLYASGRTTGIVLDSGDGVTHNVPIYEGYALPHAIMRLDLAGRDLTDYLMKILTERGYSFVTTAEREIVRDIKEKLCYVALDFENEMATAASSSSLEKSYELPDGQVITIGNERFRCPETLFQPSFIGMESAGIHETTYNSIMKCDIDIRKDLYANNVMSGGTTMYPGIADRMQKEITALAPSTMKIKIIAPPERKYSVWIGGSILASLSTFQQMWITKQEYDEAGPSIVHRKCF"</code></pre>
</div>
<p>However, a major limitation of regular character vectors is that they
do not check the validity of the sequences that they contain.
Practically, it is possible to store meaningless sequences of symbols in
character strings, including symbols that are not part of the official
alphabet for the relevant type of polymer. In those cases, the burden of
checking the validity of sequences falls on the programs that process
them, or causing those programs to run into errors when they
unexpectedly encounter invalid symbols in a sequence.</p>
<p>Instead, <a href="reference.html#s4-class">S4 classes</a> –
demonstrated in the earlier episode <a href="05-s4.html">The S4 class
system</a> – provide a way to label objects as distinct “DNA”, “RNA”, or
“protein” varieties of biological sequences. This label is an extremely
powerful way to inform programs on the set of character symbols they can
expect in the sequence, but also the range of computational operations
that can be applied to those sequences. For instance, a function
designed to translate nucleic acid sequences into the corresponding
amino acid sequence should only be allowed to run on sequences that
represent nucleic acids.</p>
<div id="challenge" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Can you tell whether this character string is a valid DNA
sequence?</p>
<pre><code><span><span class="va">AATTGGCCRGGCCAATT</span></span></code></pre>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<p>Yes, this is a valid DNA sequence using ambiguity codes defined in
the <a href="https://en.wikipedia.org/wiki/Nucleic_acid_notation#IUPAC_notation" class="external-link">IUPAC</a>
notation. In this case, <code>A</code>, <code>T</code>, <code>C</code>,
and <code>G</code> represents the four standard types of nucleotides,
while the <code>R</code> symbol acts as a regular expression
representing either of the two purine nucleotide bases, <code>A</code>
and <code>G</code>.</p>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="the-biostrings-package">The Biostrings package<a class="anchor" aria-label="anchor" href="#the-biostrings-package"></a>
</h2>
<hr class="half-width">
<div class="section level3">
<h3 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h3>
<p>In the Bioconductor project, the <em><a href="https://bioconductor.org/packages/3.19/Biostrings" class="external-link">Biostrings</a></em>
package implements S4 classes to represent biological sequences as S4
objects, e.g. <code>DNAString</code> for sequences of nucleotides in
deoxyribonucleic acid polymers, and <code>AAString</code> for sequences
of amino acids in protein polymers. Those S4 classes provide
memory-efficient containers for character strings, automatic
validity-checking functionality for each class of biological molecules,
and methods implementing various string matching algorithms and other
utilities for fast manipulation and processing of large biological
sequences or sets of sequences.</p>
<p>A short presentation of the basic classes defined in the <em><a href="https://bioconductor.org/packages/3.19/Biostrings" class="external-link">Biostrings</a></em>
package is available in one of the package vignettes, accessible as
<code>vignette("Biostrings2Classes")</code>, while more detailed
information is provided in the other package vignettes, accessible as
<code>browseVignettes("Biostrings")</code>.</p>
</div>
<div class="section level3">
<h3 id="first-steps">First steps<a class="anchor" aria-label="anchor" href="#first-steps"></a>
</h3>
<p>To get started, we load the package.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">Biostrings</span><span class="op">)</span></span></code></pre>
</div>
<p>With the package loaded and attached to the session, we have access
to the package functions. Those include functions that let us create new
objects of the classes defined in the package. For instance, we can
create an object that represents a DNA sequence, using the
<code>DNAString()</code> constructor function. Without assigning the
output to an object, we let the resulting object be printed in the
console.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">DNAString</span><span class="op">(</span><span class="st">"ATCG"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>4-letter DNAString object
seq: ATCG</code></pre>
</div>
<p>Notably, DNA sequences may only contain the symbols <code>A</code>,
<code>T</code>, <code>C</code>, and <code>G</code>, to represent the
four DNA nucleotide bases, the symbol <code>N</code> as a placeholder
for an unknown or unspecified base, and a restricted set of additional
symbols with special meaning defined in the <a href="https://www.bioinformatics.org/sms/iupac.html" class="external-link">IUPAC Extended
Genetic Alphabet</a>. Notice that the constructor function does not let
us create objects that contain invalid characters,
e.g. <code>Z</code>.</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">DNAString</span><span class="op">(</span><span class="st">"ATCGZ"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in .Call2("new_XString_from_CHARACTER", class(x0), string, start, : key 90 (char 'Z') not in lookup table</code></pre>
</div>
<p>Specifically, the <a href="https://www.bioinformatics.org/sms/iupac.html" class="external-link">IUPAC Extended
Genetic Alphabet</a> defines ambiguity codes that represent sets of
nucleotides, in a way similar to regular expressions. The
<code>IUPAC_CODE_MAP</code> named character vector contains the mapping
from the IUPAC nucleotide ambiguity codes to their meaning.</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">IUPAC_CODE_MAP</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>     A      C      G      T      M      R      W      S      Y      K      V
   "A"    "C"    "G"    "T"   "AC"   "AG"   "AT"   "CG"   "CT"   "GT"  "ACG"
     H      D      B      N
 "ACT"  "AGT"  "CGT" "ACGT" </code></pre>
</div>
<p>Any of those nucleotide codes are allowed in the sequence of a
<code>DNAString</code> object. For instance, the symbol <code>M</code>
represents either of the two nucleotides <code>A</code> or
<code>C</code> at a given position in a nucleic acid sequence.</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">DNAString</span><span class="op">(</span><span class="st">"ATCGM"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>5-letter DNAString object
seq: ATCGM</code></pre>
</div>
<p>In particular, pattern matching methods implemented in the <em><a href="https://bioconductor.org/packages/3.19/Biostrings" class="external-link">Biostrings</a></em>
package recognize the meaning of ambiguity codes for each class of
biological sequence, allowing them to efficiently match motifs queried
by users without the need to design elaborate regular expressions. For
instance, the method <code>matchPattern()</code> takes a
<code>pattern=</code> and a <code>subject=</code> argument, and returns
a <code>Views</code> object that reports and displays any match of the
<code>pattern</code> expression at any position in the
<code>subject</code> sequence.</p>
<p>Note that the default option <code>fixed = TRUE</code> instructs the
method to match the query exactly – i.e., ignore ambiguity codes – which
in this case does not report any exact match.</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dna1</span> <span class="op">&lt;-</span> <span class="fu">DNAString</span><span class="op">(</span><span class="st">"ATCGCTTTGA"</span><span class="op">)</span></span>
<span><span class="fu">matchPattern</span><span class="op">(</span><span class="st">"GM"</span>, <span class="va">dna1</span>, fixed <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Views on a 10-letter DNAString subject
subject: ATCGCTTTGA
views: NONE</code></pre>
</div>
<p>Instead, to indicate that the pattern includes some ambiguity code,
the argument <code>fixed</code> must be set to <code>FALSE</code>.</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">matchPattern</span><span class="op">(</span><span class="st">"GM"</span>, <span class="va">dna1</span>, fixed <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Views on a 10-letter DNAString subject
subject: ATCGCTTTGA
views:
      start end width
  [1]     4   5     2 [GC]
  [2]     9  10     2 [GA]</code></pre>
</div>
<p>In this particular example, two views describe matches of the pattern
in the subject sequence. Specifically, the pattern <code>GM</code> first
matched the sequence <code>GC</code> spanning positions 4 to 5 in the
subject sequence, and then also matched the sequence <code>GA</code>
from positions 9 to 10.</p>
<p>Similarly to the method <code>matchPattern()</code>, the method
<code>countPattern()</code> can be applied to simply count the number of
matches of the <code>pattern</code> in the <code>subject</code>
sequence. And again, the option <code>fixed</code> controls whether to
respect ambiguity codes, or match them exactly.</p>
<div id="challenge-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-1" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>How many hits does the following code return? Why?</p>
<pre><code><span><span class="va">dna2</span> <span class="op">&lt;-</span> <span class="fu">DNAString</span><span class="op">(</span><span class="st">"TGATTGCTTGGTTGMTT"</span><span class="op">)</span></span>
<span><span class="fu">countPattern</span><span class="op">(</span><span class="st">"GM"</span>, <span class="va">dna2</span>, fixed <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<p>The method <code>countPattern()</code> reports 3 hits, because the
option <code>fixed = FALSE</code> allows the pattern <code>GM</code> to
match <code>GA</code>, <code>GC</code>, and <code>GM</code> sequences,
due to the use of the ambiguity code <code>M</code> in the
<code>pattern</code>.</p>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="importing-biological-strings-from-files">Importing biological strings from files<a class="anchor" aria-label="anchor" href="#importing-biological-strings-from-files"></a>
</h3>
<p>In practice, users rarely type the strings representing biological
sequences themselves. Most of the time, biological strings are imported
from files, either downloaded from public repositories or generated
locally using bioinformatics programs.</p>
<p>For instance, we can load the set of adapter sequences for the <a href="https://emea.illumina.com/products/by-type/sequencing-kits/library-prep-kits.html" class="external-link">TruSeq™
DNA PCR-Free whole-genome sequencing library preparation</a> kit from a
file that we downloaded during the lesson setup. Since adapter sequences
are nucleic acid sequences, we must use the function
<code>readDNAStringSet()</code>.</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">truseq_adapters</span> <span class="op">&lt;-</span> <span class="fu">readDNAStringSet</span><span class="op">(</span>filepath <span class="op">=</span> <span class="st">"data/TruSeq3-PE-2.fa"</span><span class="op">)</span></span>
<span><span class="va">truseq_adapters</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DNAStringSet object of length 6:
    width seq                                               names
[1]    34 TACACTCTTTCCCTACACGACGCTCTTCCGATCT                PrefixPE/1
[2]    34 GTGACTGGAGTTCAGACGTGTGCTCTTCCGATCT                PrefixPE/2
[3]    34 TACACTCTTTCCCTACACGACGCTCTTCCGATCT                PE1
[4]    34 AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGTA                PE1_rc
[5]    34 GTGACTGGAGTTCAGACGTGTGCTCTTCCGATCT                PE2
[6]    34 AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC                PE2_rc</code></pre>
</div>
<div id="going-further" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="going-further" class="callout-inner">
<h3 class="callout-title">Going further</h3>
<div class="callout-content">
<p>The help page of the function <code>readDNAStringSet()</code> –
accessible using <code>help(readDNAStringSet)</code> – documents related
functions designed to import other types of biological sequences, e.g
<code>readRNAStringSet()</code>, <code>readAAStringSet()</code>.</p>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="operations-on-biological-strings">Operations on biological strings<a class="anchor" aria-label="anchor" href="#operations-on-biological-strings"></a>
</h3>
<div class="section level4">
<h4 id="computing-the-frequency-of-symbols">Computing the frequency of symbols<a class="anchor" aria-label="anchor" href="#computing-the-frequency-of-symbols"></a>
</h4>
<p>The <em><a href="https://bioconductor.org/packages/3.19/Biostrings" class="external-link">Biostrings</a></em>
package provides several functions to process and manipulate classes of
biological strings. For example, we have come across
<code>matchPattern()</code> and <code>countPattern()</code> earlier in
this episode.</p>
<p>Another example of a method that can be applied to biological strings
is <code>letterFrequency()</code>, to compute the frequency of letters
in a biological sequence.</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">letterFrequency</span><span class="op">(</span><span class="va">truseq_adapters</span>, letters <span class="op">=</span> <span class="va">DNA_ALPHABET</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>      A  C  G  T M R W S Y K V H D B N - + .
[1,]  6 14  3 11 0 0 0 0 0 0 0 0 0 0 0 0 0 0
[2,]  5  8 10 11 0 0 0 0 0 0 0 0 0 0 0 0 0 0
[3,]  6 14  3 11 0 0 0 0 0 0 0 0 0 0 0 0 0 0
[4,] 11  3 14  6 0 0 0 0 0 0 0 0 0 0 0 0 0 0
[5,]  5  8 10 11 0 0 0 0 0 0 0 0 0 0 0 0 0 0
[6,] 11 10  8  5 0 0 0 0 0 0 0 0 0 0 0 0 0 0</code></pre>
</div>
<p>The output is a matrix with one row for each sequence in the
<code>DNAStringSet</code> object, and one column for each symbol in the
alphabet of deoxyribonucleic acids, provided by the <em><a href="https://bioconductor.org/packages/3.19/Biostrings" class="external-link">Biostrings</a></em>
package in a built-in object called <code>DNA_ALPHABET</code>.</p>
</div>
</div>
<div class="section level3">
<h3 id="amino-acid-sequences">Amino acid sequences<a class="anchor" aria-label="anchor" href="#amino-acid-sequences"></a>
</h3>
<p>Similarly to the <code>DNAString</code> and <code>DNAStringSet</code>
classes, the classes <code>AAString</code> and <code>AAStringSet</code>
allow efficient storage and manipulation of a long amino acid sequence,
or a set thereof.</p>
<p>Similarly to built-in objects for the DNA alphabet, the built-in
objects <code>AA_ALPHABET</code>, <code>AA_STANDARD</code> and
<code>AA_PROTEINOGENIC</code> describe different subsets of the alphabet
of valid symbols for amino acid sequences.</p>
<p>For instance, the <code>AA_ALPHABET</code> object describes the set
of symbols in the full amino acid alphabet.</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">AA_ALPHABET</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] "A" "R" "N" "D" "C" "Q" "E" "G" "H" "I" "L" "K" "M" "F" "P" "S" "T" "W" "Y"
[20] "V" "U" "O" "B" "J" "Z" "X" "*" "-" "+" "."</code></pre>
</div>
<div id="challenge-2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-2" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Use base R code to identify the two symbols present in the
<code>AA_PROTEINOGENIC</code> alphabet object that are absent from the
<code>AA_STANDARD</code> alphabet object. What do those two symbols
represent?</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">
<pre><code>&gt; setdiff(AA_PROTEINOGENIC, AA_STANDARD)
[1] "U" "O"</code></pre>
<p>The symbols <code>U</code> and <code>O</code> represent
selenocysteine and pyrrolysine, respectively. Those two amino acids are
in some species coded for by codons that are usually interpreted as stop
codons. As such, they are not included in the alphabet of “standard”
amino acids, and an alphabet of “proteinogenic” amino acids was defined
to acknowledge the special biology of those amino acids. Either of those
alphabets may be used to determine the validity of an amino acid
sequence, depending on its biological nature.</p>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="translating-nucleotide-sequences">Translating nucleotide sequences<a class="anchor" aria-label="anchor" href="#translating-nucleotide-sequences"></a>
</h3>
<p>One of the key motivations for the use of <a href="reference.html#s4-class">S4 classes</a> and the object-oriented
programming (OOP) model relies on the infrastructure of S4 generics and
methods. As described in the earlier episode <a href="05-s4.html">The S4
class system</a>, generics provide a mechanism for defining and applying
distinct implementations of the same generic function name, according to
the nature of the input object(s) provided to the function call.</p>
<p>For instance, the <em><a href="https://bioconductor.org/packages/3.19/Biostrings" class="external-link">Biostrings</a></em>
package provides multiple implementations of a generic called
<code>translate()</code>, for translating DNA or RNA sequences into
amino acid sequences. The set of input objects supported by the generic
<code>translate()</code> can be listed using the function
<code>showMethods()</code>, from the CRAN package <em><a href="https://CRAN.R-project.org/package=methods" class="external-link">methods</a></em>.</p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">showMethods</span><span class="op">(</span><span class="st">"translate"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Function: translate (package Biostrings)
x="DNAString"
x="DNAStringSet"
x="MaskedDNAString"
x="MaskedRNAString"
x="RNAString"
x="RNAStringSet"</code></pre>
</div>
<p>In the output above, we see that that the generic function
<code>translate()</code> includes methods capable of handling objects
representing DNA and RNA sequences in the <code>DNAString</code> and
<code>RNAString</code> classes, respectively; but also lists of DNA and
RNA sequences in objects of class <code>DNAStringSet</code> and
<code>RNAStringSet</code>, as well as other classes capable of storing
DNA and RNA sequences.</p>
<p>To demonstrate the use of the <code>translate()</code> method, we
first load a set of open reading frames (ORFs) identified by the <a href="https://www.ncbi.nlm.nih.gov/orffinder/" class="external-link">NIH Open Reading Frame
Finder</a> for the <em>Homo sapiens</em> actin beta (ACTB) mRNA (RefSeq:
NM\_001101), using the standard genetic code, a minimal ORF length of 75
nucleotides, and starting with the “ATG” start codon only.</p>
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">actb_orf_nih</span> <span class="op">&lt;-</span> <span class="fu">readDNAStringSet</span><span class="op">(</span><span class="st">"data/actb_orfs.fasta"</span><span class="op">)</span></span>
<span><span class="va">actb_orf_nih</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DNAStringSet object of length 13:
     width seq                                              names
 [1]   222 ATGCCCACCATCACGCCCTGGTG...CGGGGCGGACGCGGTCTCGGCG gi|1519311456|ref...
 [2]  1128 ATGGATGATGATATCGCCGCGCT...CGTCCACCGCAAATGCTTCTAG gi|1519311456|ref...
 [3]   126 ATGATGATATCGCCGCGCTCGTC...CGCCCCAGGCACCAGGGCGTGA gi|1519311456|ref...
 [4]    90 ATGTCGTCCCAGTTGGTGACGAT...CTGGGCCTCGTCGCCCACATAG gi|1519311456|ref...
 [5]   225 ATGGGCACAGTGTGGGTGACCCC...AGCCACACGCAGCTCATTGTAG gi|1519311456|ref...
 ...   ... ...
 [9]   342 ATGAGATTGGCATGGCTTTATTT...ATGTAATGCAAAATTTTTTTAA gi|1519311456|ref...
[10]   168 ATGGCTTTATTTGTTTTTTTTGT...TTGCACATTGTTGTTTTTTTAA gi|1519311456|ref...
[11]   111 ATGACTATTAAAAAAACAACAAT...CCTTCACCGTTCCAGTTTTTAA gi|1519311456|ref...
[12]   105 ATGCAAAATTTTTTTAATCTTCG...CCTTTTTTGTCCCCCAACTTGA gi|1519311456|ref...
[13]   135 ATGATGAGCCTTCGTGCCCCCCC...TGACTTGAGACCAGTTGAATAA gi|1519311456|ref...</code></pre>
</div>
<p>Having imported the nucleotide sequences as a
<code>DNAStringSet</code> object, we can apply the
<code>translate()</code> method to that object to produce the amino acid
sequence that results from the translation process for each nucleotide
sequence.</p>
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">actb_aa</span> <span class="op">&lt;-</span> <span class="fu">translate</span><span class="op">(</span><span class="va">actb_orf_nih</span><span class="op">)</span></span>
<span><span class="va">actb_aa</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>AAStringSet object of length 13:
     width seq                                              names
 [1]    74 MPTITPWCLGRPTMEGKTARGAS...VWTGGGSAKARLCARGADAVSA gi|1519311456|ref...
 [2]   376 MDDDIAALVVDNGSGMCKAGFAG...MWISKQEYDESGPSIVHRKCF* gi|1519311456|ref...
 [3]    42 MMISPRSSSTTAPACARPASRATMPPGPSSPPSWGAPGTRA*       gi|1519311456|ref...
 [4]    30 MSSQLVTMPCSMGYFRVRMPLLLWASSPT*                   gi|1519311456|ref...
 [5]    75 MGTVWVTPSPESITMPVVRPEAY...GFRGASVSSTGCSSGATRSSL* gi|1519311456|ref...
 ...   ... ...
 [9]   114 MRLAWLYLFFLFCFGFFFFFGLT...QVHTGEVIALLSCKLCNAKFF* gi|1519311456|ref...
[10]    56 MALFVFFVLFWFFFFFWLDSGFK...ERASPKVHNVAEDFDCTLLFF* gi|1519311456|ref...
[11]    37 MTIKKTTMCNQSPRPHCELWGMLAPTDCCHLHRSSF*            gi|1519311456|ref...
[12]    35 MQNFFNLRLNTFLFCFILNDEPSCPPFPLFCPPT*              gi|1519311456|ref...
[13]    45 MMSLRAPPSPFFVPQLEMYEGFWSPWEWVEAARAYLYTDLRPVE*    gi|1519311456|ref...</code></pre>
</div>
<p>In the example above, all amino acid sequences visible start with the
typical methionin amino acid encoded by the “ATG” start codon. We also
see that all but one of the amino acid sequences visible end with the
<code>*</code> symbol, which indicates that the translation process
ended on a stop codon. In contrast, the first open reading frame above
reached the end of the nucleotide sequence without encoutering a stop
codon.</p>
<p>Conveniently, the number of amino acids in each sequence is stated
under the header <code>width</code>.</p>
<div id="challenge-3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-3" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Extract the length of each amino acid sequence above as an integer
vector. What is the length of the longest amino acid sequence translated
from any of those open reading frames?</p>
<p>Compare your result with the sequence information on the UniPro page
for ACTB (<a href="https://www.uniprot.org/uniprot/P60709#sequences" class="external-link">https://www.uniprot.org/uniprot/P60709#sequences</a>).</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" aria-labelledby="headingSolution4" data-bs-parent="#accordionSolution4">
<div class="accordion-body">
<pre><code><span><span class="fu">width</span><span class="op">(</span><span class="va">actb_aa</span><span class="op">)</span></span>
<span><span class="co"># or</span></span>
<span><span class="fu">max</span><span class="op">(</span><span class="fu">width</span><span class="op">(</span><span class="va">actb_aa</span><span class="op">)</span><span class="op">)</span></span></code></pre>
<p>The longest translated sequence contains 376 amino acids.</p>
<p>The Uniprot page reports a sequence of 375 amino acids. However, the
UniProt amino acid sequence does not comprise any symbol to represent
the stop codon. That difference aside, the UniPro amino acid sequence is
identical to the sequence that was produced by the
<code>translate()</code> method.</p>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="the-bsgenome-package">The BSgenome package<a class="anchor" aria-label="anchor" href="#the-bsgenome-package"></a>
</h2>
<hr class="half-width">
<div class="section level3">
<h3 id="overview-1">Overview<a class="anchor" aria-label="anchor" href="#overview-1"></a>
</h3>
<p>In the Bioconductor project, the <em><a href="https://bioconductor.org/packages/3.19/BSgenome" class="external-link">BSgenome</a></em>
package provides software infrastructure for efficient representation of
full genome and their single-nucleotide polymorphisms.</p>
<p>The <em><a href="https://bioconductor.org/packages/3.19/BSgenome" class="external-link">BSgenome</a></em>
package itself does not contain any genome sequence itself, but provides
functionality to access genome sequences available in other Bioconductor
packages, as we demonstrate in the next section.</p>
</div>
<div class="section level3">
<h3 id="first-steps-1">First steps<a class="anchor" aria-label="anchor" href="#first-steps-1"></a>
</h3>
<p>To get started, we load the package.</p>
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">BSgenome</span><span class="op">)</span></span></code></pre>
</div>
<p>With the package loaded and attached to the session, we have access
to the package functions.</p>
<p>In particular, the function
<code>BSgenome::available.genomes()</code> can be used to display the
names of Bioconductor packages that contain genome sequences.</p>
<div class="codewrapper sourceCode" id="cb33">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">available.genomes</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>'getOption("repos")' replaces Bioconductor standard repositories, see
'help("repositories", package = "BiocManager")' for details.
Replacement repositories:
    BioCsoft: https://bioconductor.org/packages/3.19/bioc
    BioCann: https://bioconductor.org/packages/3.19/data/annotation
    BioCexp: https://bioconductor.org/packages/3.19/data/experiment
    BioCworkflows: https://bioconductor.org/packages/3.19/workflows
    BioCbooks: https://bioconductor.org/packages/3.19/books
    CRAN: https://cloud.r-project.org</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>  [1] "BSgenome.Alyrata.JGI.v1"
  [2] "BSgenome.Amellifera.BeeBase.assembly4"
  [3] "BSgenome.Amellifera.NCBI.AmelHAv3.1"
  [4] "BSgenome.Amellifera.UCSC.apiMel2"
  [5] "BSgenome.Amellifera.UCSC.apiMel2.masked"
  [6] "BSgenome.Aofficinalis.NCBI.V1"
  [7] "BSgenome.Athaliana.TAIR.04232008"
  [8] "BSgenome.Athaliana.TAIR.TAIR9"
  [9] "BSgenome.Btaurus.UCSC.bosTau3"
 [10] "BSgenome.Btaurus.UCSC.bosTau3.masked"
 [11] "BSgenome.Btaurus.UCSC.bosTau4"
 [12] "BSgenome.Btaurus.UCSC.bosTau4.masked"
 [13] "BSgenome.Btaurus.UCSC.bosTau6"
 [14] "BSgenome.Btaurus.UCSC.bosTau6.masked"
 [15] "BSgenome.Btaurus.UCSC.bosTau8"
 [16] "BSgenome.Btaurus.UCSC.bosTau9"
 [17] "BSgenome.Btaurus.UCSC.bosTau9.masked"
 [18] "BSgenome.Carietinum.NCBI.v1"
 [19] "BSgenome.Celegans.UCSC.ce10"
 [20] "BSgenome.Celegans.UCSC.ce11"
 [21] "BSgenome.Celegans.UCSC.ce2"
 [22] "BSgenome.Celegans.UCSC.ce6"
 [23] "BSgenome.Cfamiliaris.UCSC.canFam2"
 [24] "BSgenome.Cfamiliaris.UCSC.canFam2.masked"
 [25] "BSgenome.Cfamiliaris.UCSC.canFam3"
 [26] "BSgenome.Cfamiliaris.UCSC.canFam3.masked"
 [27] "BSgenome.Cjacchus.UCSC.calJac3"
 [28] "BSgenome.Cjacchus.UCSC.calJac4"
 [29] "BSgenome.CneoformansVarGrubiiKN99.NCBI.ASM221672v1"
 [30] "BSgenome.Creinhardtii.JGI.v5.6"
 [31] "BSgenome.Dmelanogaster.UCSC.dm2"
 [32] "BSgenome.Dmelanogaster.UCSC.dm2.masked"
 [33] "BSgenome.Dmelanogaster.UCSC.dm3"
 [34] "BSgenome.Dmelanogaster.UCSC.dm3.masked"
 [35] "BSgenome.Dmelanogaster.UCSC.dm6"
 [36] "BSgenome.Drerio.UCSC.danRer10"
 [37] "BSgenome.Drerio.UCSC.danRer11"
 [38] "BSgenome.Drerio.UCSC.danRer5"
 [39] "BSgenome.Drerio.UCSC.danRer5.masked"
 [40] "BSgenome.Drerio.UCSC.danRer6"
 [41] "BSgenome.Drerio.UCSC.danRer6.masked"
 [42] "BSgenome.Drerio.UCSC.danRer7"
 [43] "BSgenome.Drerio.UCSC.danRer7.masked"
 [44] "BSgenome.Dvirilis.Ensembl.dvircaf1"
 [45] "BSgenome.Ecoli.NCBI.20080805"
 [46] "BSgenome.Gaculeatus.UCSC.gasAcu1"
 [47] "BSgenome.Gaculeatus.UCSC.gasAcu1.masked"
 [48] "BSgenome.Ggallus.UCSC.galGal3"
 [49] "BSgenome.Ggallus.UCSC.galGal3.masked"
 [50] "BSgenome.Ggallus.UCSC.galGal4"
 [51] "BSgenome.Ggallus.UCSC.galGal4.masked"
 [52] "BSgenome.Ggallus.UCSC.galGal5"
 [53] "BSgenome.Ggallus.UCSC.galGal6"
 [54] "BSgenome.Gmax.NCBI.Gmv40"
 [55] "BSgenome.Hsapiens.1000genomes.hs37d5"
 [56] "BSgenome.Hsapiens.NCBI.GRCh38"
 [57] "BSgenome.Hsapiens.NCBI.T2T.CHM13v2.0"
 [58] "BSgenome.Hsapiens.UCSC.hg17"
 [59] "BSgenome.Hsapiens.UCSC.hg17.masked"
 [60] "BSgenome.Hsapiens.UCSC.hg18"
 [61] "BSgenome.Hsapiens.UCSC.hg18.masked"
 [62] "BSgenome.Hsapiens.UCSC.hg19"
 [63] "BSgenome.Hsapiens.UCSC.hg19.masked"
 [64] "BSgenome.Hsapiens.UCSC.hg38"
 [65] "BSgenome.Hsapiens.UCSC.hg38.dbSNP151.major"
 [66] "BSgenome.Hsapiens.UCSC.hg38.dbSNP151.minor"
 [67] "BSgenome.Hsapiens.UCSC.hg38.masked"
 [68] "BSgenome.Hsapiens.UCSC.hs1"
 [69] "BSgenome.Mdomestica.UCSC.monDom5"
 [70] "BSgenome.Mfascicularis.NCBI.5.0"
 [71] "BSgenome.Mfascicularis.NCBI.6.0"
 [72] "BSgenome.Mfuro.UCSC.musFur1"
 [73] "BSgenome.Mmulatta.UCSC.rheMac10"
 [74] "BSgenome.Mmulatta.UCSC.rheMac2"
 [75] "BSgenome.Mmulatta.UCSC.rheMac2.masked"
 [76] "BSgenome.Mmulatta.UCSC.rheMac3"
 [77] "BSgenome.Mmulatta.UCSC.rheMac3.masked"
 [78] "BSgenome.Mmulatta.UCSC.rheMac8"
 [79] "BSgenome.Mmusculus.UCSC.mm10"
 [80] "BSgenome.Mmusculus.UCSC.mm10.masked"
 [81] "BSgenome.Mmusculus.UCSC.mm39"
 [82] "BSgenome.Mmusculus.UCSC.mm8"
 [83] "BSgenome.Mmusculus.UCSC.mm8.masked"
 [84] "BSgenome.Mmusculus.UCSC.mm9"
 [85] "BSgenome.Mmusculus.UCSC.mm9.masked"
 [86] "BSgenome.Osativa.MSU.MSU7"
 [87] "BSgenome.Ppaniscus.UCSC.panPan1"
 [88] "BSgenome.Ppaniscus.UCSC.panPan2"
 [89] "BSgenome.Ptroglodytes.UCSC.panTro2"
 [90] "BSgenome.Ptroglodytes.UCSC.panTro2.masked"
 [91] "BSgenome.Ptroglodytes.UCSC.panTro3"
 [92] "BSgenome.Ptroglodytes.UCSC.panTro3.masked"
 [93] "BSgenome.Ptroglodytes.UCSC.panTro5"
 [94] "BSgenome.Ptroglodytes.UCSC.panTro6"
 [95] "BSgenome.Rnorvegicus.UCSC.rn4"
 [96] "BSgenome.Rnorvegicus.UCSC.rn4.masked"
 [97] "BSgenome.Rnorvegicus.UCSC.rn5"
 [98] "BSgenome.Rnorvegicus.UCSC.rn5.masked"
 [99] "BSgenome.Rnorvegicus.UCSC.rn6"
[100] "BSgenome.Rnorvegicus.UCSC.rn7"
[101] "BSgenome.Scerevisiae.UCSC.sacCer1"
[102] "BSgenome.Scerevisiae.UCSC.sacCer2"
[103] "BSgenome.Scerevisiae.UCSC.sacCer3"
[104] "BSgenome.Sscrofa.UCSC.susScr11"
[105] "BSgenome.Sscrofa.UCSC.susScr3"
[106] "BSgenome.Sscrofa.UCSC.susScr3.masked"
[107] "BSgenome.Tgondii.ToxoDB.7.0"
[108] "BSgenome.Tguttata.UCSC.taeGut1"
[109] "BSgenome.Tguttata.UCSC.taeGut1.masked"
[110] "BSgenome.Tguttata.UCSC.taeGut2"
[111] "BSgenome.Vvinifera.URGI.IGGP12Xv0"
[112] "BSgenome.Vvinifera.URGI.IGGP12Xv2"
[113] "BSgenome.Vvinifera.URGI.IGGP8X"                    </code></pre>
</div>
</div>
<div class="section level3">
<h3 id="installing-bsgenome-packages">Installing BSgenome packages<a class="anchor" aria-label="anchor" href="#installing-bsgenome-packages"></a>
</h3>
<p>To use one of the available genomes, the corresponding package must
be installed first. For instance, the example below demonstrates how the
data package <em><a href="https://bioconductor.org/packages/3.19/BSgenome.Hsapiens.UCSC.hg38.masked" class="external-link">BSgenome.Hsapiens.UCSC.hg38.masked</a></em>
can be installed using the function <code>BiocManager::install()</code>
that we have seen before.</p>
<div class="codewrapper sourceCode" id="cb36">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"BSgenome.Hsapiens.UCSC.hg38.masked"</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="using-bsgenome-packages">Using BSgenome packages<a class="anchor" aria-label="anchor" href="#using-bsgenome-packages"></a>
</h3>
<p>Once installed, BSgenome packages can be loaded like any other R
package, using the <code>library()</code> function.</p>
<div class="codewrapper sourceCode" id="cb37">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">BSgenome.Hsapiens.UCSC.hg38.masked</span><span class="op">)</span></span></code></pre>
</div>
<p>Each BSgenome package contains an object that is named identically to
the package and contains the genome sequence.</p>
<p>Having loaded the package <em><a href="https://bioconductor.org/packages/3.19/BSgenome.Hsapiens.UCSC.hg38.masked" class="external-link">BSgenome.Hsapiens.UCSC.hg38.masked</a></em>
above, we can display the BSgenome object as follows.</p>
<div class="codewrapper sourceCode" id="cb38">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">BSgenome.Hsapiens.UCSC.hg38.masked</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>| BSgenome object for Human
| - organism: Homo sapiens
| - provider: UCSC
| - genome: hg38
| - release date: 2023/01/31
| - 711 sequence(s):
|     chr1                    chr2                    chr3
|     chr4                    chr5                    chr6
|     chr7                    chr8                    chr9
|     chr10                   chr11                   chr12
|     chr13                   chr14                   chr15
|     ...                     ...                     ...
|     chr19_KV575256v1_alt    chr19_KV575257v1_alt    chr19_KV575258v1_alt
|     chr19_KV575259v1_alt    chr19_KV575260v1_alt    chr19_MU273387v1_alt
|     chr22_KN196485v1_alt    chr22_KN196486v1_alt    chr22_KQ458387v1_alt
|     chr22_KQ458388v1_alt    chr22_KQ759761v1_alt    chrX_KV766199v1_alt
|     chrX_MU273395v1_alt     chrX_MU273396v1_alt     chrX_MU273397v1_alt
|
| Tips: call 'seqnames()' on the object to get all the sequence names, call
| 'seqinfo()' to get the full sequence info, use the '$' or '[[' operator to
| access a given sequence, see '?BSgenome' for more information.</code></pre>
</div>
<p>Given the length and the complexity of the object name, it is common
practice to assign a copy of BSgenome objects to a new object simply
called <code>genome</code>.</p>
<div class="codewrapper sourceCode" id="cb40">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genome</span> <span class="op">&lt;-</span> <span class="va">BSgenome.Hsapiens.UCSC.hg38.masked</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="using-bsgenome-objects">Using BSgenome objects<a class="anchor" aria-label="anchor" href="#using-bsgenome-objects"></a>
</h3>
<p>When printing BSgenome objects in the console (see above), some
helpful tips are displayed under the object itself, hinting at functions
commonly used to access information in the object.</p>
<p>For instance, the function <code>seqnames()</code> can be used get
the list of sequence names (i.e., chromosomes and contigs) present in
the object.</p>
<div class="codewrapper sourceCode" id="cb41">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">seqnames</span><span class="op">(</span><span class="va">genome</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>  [1] "chr1"                    "chr2"
  [3] "chr3"                    "chr4"
  [5] "chr5"                    "chr6"
  [7] "chr7"                    "chr8"
  [9] "chr9"                    "chr10"
 [11] "chr11"                   "chr12"
 [13] "chr13"                   "chr14"
 [15] "chr15"                   "chr16"
 [17] "chr17"                   "chr18"
 [19] "chr19"                   "chr20"
 [21] "chr21"                   "chr22"
 [23] "chrX"                    "chrY"
 [25] "chrM"                    "chr1_GL383518v1_alt"
 [27] "chr1_GL383519v1_alt"     "chr1_GL383520v2_alt"
 [29] "chr1_KI270759v1_alt"     "chr1_KI270760v1_alt"
 [31] "chr1_KI270761v1_alt"     "chr1_KI270762v1_alt"
 [33] "chr1_KI270763v1_alt"     "chr1_KI270764v1_alt"
 [35] "chr1_KI270765v1_alt"     "chr1_KI270766v1_alt"
 [37] "chr1_KI270892v1_alt"     "chr2_GL383521v1_alt"
 [39] "chr2_GL383522v1_alt"     "chr2_GL582966v2_alt"
 [41] "chr2_KI270767v1_alt"     "chr2_KI270768v1_alt"
 [43] "chr2_KI270769v1_alt"     "chr2_KI270770v1_alt"
 [45] "chr2_KI270771v1_alt"     "chr2_KI270772v1_alt"
 [47] "chr2_KI270773v1_alt"     "chr2_KI270774v1_alt"
 [49] "chr2_KI270775v1_alt"     "chr2_KI270776v1_alt"
 [51] "chr2_KI270893v1_alt"     "chr2_KI270894v1_alt"
 [53] "chr3_GL383526v1_alt"     "chr3_JH636055v2_alt"
 [55] "chr3_KI270777v1_alt"     "chr3_KI270778v1_alt"
 [57] "chr3_KI270779v1_alt"     "chr3_KI270780v1_alt"
 [59] "chr3_KI270781v1_alt"     "chr3_KI270782v1_alt"
 [61] "chr3_KI270783v1_alt"     "chr3_KI270784v1_alt"
 [63] "chr3_KI270895v1_alt"     "chr3_KI270924v1_alt"
 [65] "chr3_KI270934v1_alt"     "chr3_KI270935v1_alt"
 [67] "chr3_KI270936v1_alt"     "chr3_KI270937v1_alt"
 [69] "chr4_GL000257v2_alt"     "chr4_GL383527v1_alt"
 [71] "chr4_GL383528v1_alt"     "chr4_KI270785v1_alt"
 [73] "chr4_KI270786v1_alt"     "chr4_KI270787v1_alt"
 [75] "chr4_KI270788v1_alt"     "chr4_KI270789v1_alt"
 [77] "chr4_KI270790v1_alt"     "chr4_KI270896v1_alt"
 [79] "chr4_KI270925v1_alt"     "chr5_GL339449v2_alt"
 [81] "chr5_GL383530v1_alt"     "chr5_GL383531v1_alt"
 [83] "chr5_GL383532v1_alt"     "chr5_GL949742v1_alt"
 [85] "chr5_KI270791v1_alt"     "chr5_KI270792v1_alt"
 [87] "chr5_KI270793v1_alt"     "chr5_KI270794v1_alt"
 [89] "chr5_KI270795v1_alt"     "chr5_KI270796v1_alt"
 [91] "chr5_KI270897v1_alt"     "chr5_KI270898v1_alt"
 [93] "chr6_GL000250v2_alt"     "chr6_GL000251v2_alt"
 [95] "chr6_GL000252v2_alt"     "chr6_GL000253v2_alt"
 [97] "chr6_GL000254v2_alt"     "chr6_GL000255v2_alt"
 [99] "chr6_GL000256v2_alt"     "chr6_GL383533v1_alt"
[101] "chr6_KB021644v2_alt"     "chr6_KI270758v1_alt"
[103] "chr6_KI270797v1_alt"     "chr6_KI270798v1_alt"
[105] "chr6_KI270799v1_alt"     "chr6_KI270800v1_alt"
[107] "chr6_KI270801v1_alt"     "chr6_KI270802v1_alt"
[109] "chr7_GL383534v2_alt"     "chr7_KI270803v1_alt"
[111] "chr7_KI270804v1_alt"     "chr7_KI270805v1_alt"
[113] "chr7_KI270806v1_alt"     "chr7_KI270807v1_alt"
[115] "chr7_KI270808v1_alt"     "chr7_KI270809v1_alt"
[117] "chr7_KI270899v1_alt"     "chr8_KI270810v1_alt"
[119] "chr8_KI270811v1_alt"     "chr8_KI270812v1_alt"
[121] "chr8_KI270813v1_alt"     "chr8_KI270814v1_alt"
[123] "chr8_KI270815v1_alt"     "chr8_KI270816v1_alt"
[125] "chr8_KI270817v1_alt"     "chr8_KI270818v1_alt"
[127] "chr8_KI270819v1_alt"     "chr8_KI270820v1_alt"
[129] "chr8_KI270821v1_alt"     "chr8_KI270822v1_alt"
[131] "chr8_KI270900v1_alt"     "chr8_KI270901v1_alt"
[133] "chr8_KI270926v1_alt"     "chr9_GL383539v1_alt"
[135] "chr9_GL383540v1_alt"     "chr9_GL383541v1_alt"
[137] "chr9_GL383542v1_alt"     "chr9_KI270823v1_alt"
[139] "chr10_GL383545v1_alt"    "chr10_GL383546v1_alt"
[141] "chr10_KI270824v1_alt"    "chr10_KI270825v1_alt"
[143] "chr11_GL383547v1_alt"    "chr11_JH159136v1_alt"
[145] "chr11_JH159137v1_alt"    "chr11_KI270826v1_alt"
[147] "chr11_KI270827v1_alt"    "chr11_KI270829v1_alt"
[149] "chr11_KI270830v1_alt"    "chr11_KI270831v1_alt"
[151] "chr11_KI270832v1_alt"    "chr11_KI270902v1_alt"
[153] "chr11_KI270903v1_alt"    "chr11_KI270927v1_alt"
[155] "chr12_GL383549v1_alt"    "chr12_GL383550v2_alt"
[157] "chr12_GL383551v1_alt"    "chr12_GL383552v1_alt"
[159] "chr12_GL383553v2_alt"    "chr12_GL877875v1_alt"
[161] "chr12_GL877876v1_alt"    "chr12_KI270833v1_alt"
[163] "chr12_KI270834v1_alt"    "chr12_KI270835v1_alt"
[165] "chr12_KI270836v1_alt"    "chr12_KI270837v1_alt"
[167] "chr12_KI270904v1_alt"    "chr13_KI270838v1_alt"
[169] "chr13_KI270839v1_alt"    "chr13_KI270840v1_alt"
[171] "chr13_KI270841v1_alt"    "chr13_KI270842v1_alt"
[173] "chr13_KI270843v1_alt"    "chr14_KI270844v1_alt"
[175] "chr14_KI270845v1_alt"    "chr14_KI270846v1_alt"
[177] "chr14_KI270847v1_alt"    "chr15_GL383554v1_alt"
[179] "chr15_GL383555v2_alt"    "chr15_KI270848v1_alt"
[181] "chr15_KI270849v1_alt"    "chr15_KI270850v1_alt"
[183] "chr15_KI270851v1_alt"    "chr15_KI270852v1_alt"
[185] "chr15_KI270905v1_alt"    "chr15_KI270906v1_alt"
[187] "chr16_GL383556v1_alt"    "chr16_GL383557v1_alt"
[189] "chr16_KI270853v1_alt"    "chr16_KI270854v1_alt"
[191] "chr16_KI270855v1_alt"    "chr16_KI270856v1_alt"
[193] "chr17_GL000258v2_alt"    "chr17_GL383563v3_alt"
[195] "chr17_GL383564v2_alt"    "chr17_GL383565v1_alt"
[197] "chr17_GL383566v1_alt"    "chr17_JH159146v1_alt"
[199] "chr17_JH159147v1_alt"    "chr17_JH159148v1_alt"
[201] "chr17_KI270857v1_alt"    "chr17_KI270858v1_alt"
[203] "chr17_KI270859v1_alt"    "chr17_KI270860v1_alt"
[205] "chr17_KI270861v1_alt"    "chr17_KI270862v1_alt"
[207] "chr17_KI270907v1_alt"    "chr17_KI270908v1_alt"
[209] "chr17_KI270909v1_alt"    "chr17_KI270910v1_alt"
[211] "chr18_GL383567v1_alt"    "chr18_GL383568v1_alt"
[213] "chr18_GL383569v1_alt"    "chr18_GL383570v1_alt"
[215] "chr18_GL383571v1_alt"    "chr18_GL383572v1_alt"
[217] "chr18_KI270863v1_alt"    "chr18_KI270864v1_alt"
[219] "chr18_KI270911v1_alt"    "chr18_KI270912v1_alt"
[221] "chr19_GL000209v2_alt"    "chr19_GL383573v1_alt"
[223] "chr19_GL383574v1_alt"    "chr19_GL383575v2_alt"
[225] "chr19_GL383576v1_alt"    "chr19_GL949746v1_alt"
[227] "chr19_GL949747v2_alt"    "chr19_GL949748v2_alt"
[229] "chr19_GL949749v2_alt"    "chr19_GL949750v2_alt"
[231] "chr19_GL949751v2_alt"    "chr19_GL949752v1_alt"
[233] "chr19_GL949753v2_alt"    "chr19_KI270865v1_alt"
[235] "chr19_KI270866v1_alt"    "chr19_KI270867v1_alt"
[237] "chr19_KI270868v1_alt"    "chr19_KI270882v1_alt"
[239] "chr19_KI270883v1_alt"    "chr19_KI270884v1_alt"
[241] "chr19_KI270885v1_alt"    "chr19_KI270886v1_alt"
[243] "chr19_KI270887v1_alt"    "chr19_KI270888v1_alt"
[245] "chr19_KI270889v1_alt"    "chr19_KI270890v1_alt"
[247] "chr19_KI270891v1_alt"    "chr19_KI270914v1_alt"
[249] "chr19_KI270915v1_alt"    "chr19_KI270916v1_alt"
[251] "chr19_KI270917v1_alt"    "chr19_KI270918v1_alt"
[253] "chr19_KI270919v1_alt"    "chr19_KI270920v1_alt"
[255] "chr19_KI270921v1_alt"    "chr19_KI270922v1_alt"
[257] "chr19_KI270923v1_alt"    "chr19_KI270929v1_alt"
[259] "chr19_KI270930v1_alt"    "chr19_KI270931v1_alt"
[261] "chr19_KI270932v1_alt"    "chr19_KI270933v1_alt"
[263] "chr19_KI270938v1_alt"    "chr20_GL383577v2_alt"
[265] "chr20_KI270869v1_alt"    "chr20_KI270870v1_alt"
[267] "chr20_KI270871v1_alt"    "chr21_GL383578v2_alt"
[269] "chr21_GL383579v2_alt"    "chr21_GL383580v2_alt"
[271] "chr21_GL383581v2_alt"    "chr21_KI270872v1_alt"
[273] "chr21_KI270873v1_alt"    "chr21_KI270874v1_alt"
[275] "chr22_GL383582v2_alt"    "chr22_GL383583v2_alt"
[277] "chr22_KB663609v1_alt"    "chr22_KI270875v1_alt"
[279] "chr22_KI270876v1_alt"    "chr22_KI270877v1_alt"
[281] "chr22_KI270878v1_alt"    "chr22_KI270879v1_alt"
[283] "chr22_KI270928v1_alt"    "chrX_KI270880v1_alt"
[285] "chrX_KI270881v1_alt"     "chrX_KI270913v1_alt"
[287] "chr1_KI270706v1_random"  "chr1_KI270707v1_random"
[289] "chr1_KI270708v1_random"  "chr1_KI270709v1_random"
[291] "chr1_KI270710v1_random"  "chr1_KI270711v1_random"
[293] "chr1_KI270712v1_random"  "chr1_KI270713v1_random"
[295] "chr1_KI270714v1_random"  "chr2_KI270715v1_random"
[297] "chr2_KI270716v1_random"  "chr3_GL000221v1_random"
[299] "chr4_GL000008v2_random"  "chr5_GL000208v1_random"
[301] "chr9_KI270717v1_random"  "chr9_KI270718v1_random"
[303] "chr9_KI270719v1_random"  "chr9_KI270720v1_random"
[305] "chr11_KI270721v1_random" "chr14_GL000009v2_random"
[307] "chr14_GL000194v1_random" "chr14_GL000225v1_random"
[309] "chr14_KI270722v1_random" "chr14_KI270723v1_random"
[311] "chr14_KI270724v1_random" "chr14_KI270725v1_random"
[313] "chr14_KI270726v1_random" "chr15_KI270727v1_random"
[315] "chr16_KI270728v1_random" "chr17_GL000205v2_random"
[317] "chr17_KI270729v1_random" "chr17_KI270730v1_random"
[319] "chr22_KI270731v1_random" "chr22_KI270732v1_random"
[321] "chr22_KI270733v1_random" "chr22_KI270734v1_random"
[323] "chr22_KI270735v1_random" "chr22_KI270736v1_random"
[325] "chr22_KI270737v1_random" "chr22_KI270738v1_random"
[327] "chr22_KI270739v1_random" "chrY_KI270740v1_random"
[329] "chrUn_GL000195v1"        "chrUn_GL000213v1"
[331] "chrUn_GL000214v1"        "chrUn_GL000216v2"
[333] "chrUn_GL000218v1"        "chrUn_GL000219v1"
[335] "chrUn_GL000220v1"        "chrUn_GL000224v1"
[337] "chrUn_GL000226v1"        "chrUn_KI270302v1"
[339] "chrUn_KI270303v1"        "chrUn_KI270304v1"
[341] "chrUn_KI270305v1"        "chrUn_KI270310v1"
[343] "chrUn_KI270311v1"        "chrUn_KI270312v1"
[345] "chrUn_KI270315v1"        "chrUn_KI270316v1"
[347] "chrUn_KI270317v1"        "chrUn_KI270320v1"
[349] "chrUn_KI270322v1"        "chrUn_KI270329v1"
[351] "chrUn_KI270330v1"        "chrUn_KI270333v1"
[353] "chrUn_KI270334v1"        "chrUn_KI270335v1"
[355] "chrUn_KI270336v1"        "chrUn_KI270337v1"
[357] "chrUn_KI270338v1"        "chrUn_KI270340v1"
[359] "chrUn_KI270362v1"        "chrUn_KI270363v1"
[361] "chrUn_KI270364v1"        "chrUn_KI270366v1"
[363] "chrUn_KI270371v1"        "chrUn_KI270372v1"
[365] "chrUn_KI270373v1"        "chrUn_KI270374v1"
[367] "chrUn_KI270375v1"        "chrUn_KI270376v1"
[369] "chrUn_KI270378v1"        "chrUn_KI270379v1"
[371] "chrUn_KI270381v1"        "chrUn_KI270382v1"
[373] "chrUn_KI270383v1"        "chrUn_KI270384v1"
[375] "chrUn_KI270385v1"        "chrUn_KI270386v1"
[377] "chrUn_KI270387v1"        "chrUn_KI270388v1"
[379] "chrUn_KI270389v1"        "chrUn_KI270390v1"
[381] "chrUn_KI270391v1"        "chrUn_KI270392v1"
[383] "chrUn_KI270393v1"        "chrUn_KI270394v1"
[385] "chrUn_KI270395v1"        "chrUn_KI270396v1"
[387] "chrUn_KI270411v1"        "chrUn_KI270412v1"
[389] "chrUn_KI270414v1"        "chrUn_KI270417v1"
[391] "chrUn_KI270418v1"        "chrUn_KI270419v1"
[393] "chrUn_KI270420v1"        "chrUn_KI270422v1"
[395] "chrUn_KI270423v1"        "chrUn_KI270424v1"
[397] "chrUn_KI270425v1"        "chrUn_KI270429v1"
[399] "chrUn_KI270435v1"        "chrUn_KI270438v1"
[401] "chrUn_KI270442v1"        "chrUn_KI270448v1"
[403] "chrUn_KI270465v1"        "chrUn_KI270466v1"
[405] "chrUn_KI270467v1"        "chrUn_KI270468v1"
[407] "chrUn_KI270507v1"        "chrUn_KI270508v1"
[409] "chrUn_KI270509v1"        "chrUn_KI270510v1"
[411] "chrUn_KI270511v1"        "chrUn_KI270512v1"
[413] "chrUn_KI270515v1"        "chrUn_KI270516v1"
[415] "chrUn_KI270517v1"        "chrUn_KI270518v1"
[417] "chrUn_KI270519v1"        "chrUn_KI270521v1"
[419] "chrUn_KI270522v1"        "chrUn_KI270528v1"
[421] "chrUn_KI270529v1"        "chrUn_KI270530v1"
[423] "chrUn_KI270538v1"        "chrUn_KI270539v1"
[425] "chrUn_KI270544v1"        "chrUn_KI270548v1"
[427] "chrUn_KI270579v1"        "chrUn_KI270580v1"
[429] "chrUn_KI270581v1"        "chrUn_KI270582v1"
[431] "chrUn_KI270583v1"        "chrUn_KI270584v1"
[433] "chrUn_KI270587v1"        "chrUn_KI270588v1"
[435] "chrUn_KI270589v1"        "chrUn_KI270590v1"
[437] "chrUn_KI270591v1"        "chrUn_KI270593v1"
[439] "chrUn_KI270741v1"        "chrUn_KI270742v1"
[441] "chrUn_KI270743v1"        "chrUn_KI270744v1"
[443] "chrUn_KI270745v1"        "chrUn_KI270746v1"
[445] "chrUn_KI270747v1"        "chrUn_KI270748v1"
[447] "chrUn_KI270749v1"        "chrUn_KI270750v1"
[449] "chrUn_KI270751v1"        "chrUn_KI270752v1"
[451] "chrUn_KI270753v1"        "chrUn_KI270754v1"
[453] "chrUn_KI270755v1"        "chrUn_KI270756v1"
[455] "chrUn_KI270757v1"        "chr1_KN196472v1_fix"
[457] "chr1_KN196473v1_fix"     "chr1_KN196474v1_fix"
[459] "chr1_KN538360v1_fix"     "chr1_KN538361v1_fix"
[461] "chr1_KQ031383v1_fix"     "chr1_KZ208906v1_fix"
[463] "chr1_KZ559100v1_fix"     "chr1_MU273333v1_fix"
[465] "chr1_MU273334v1_fix"     "chr1_MU273335v1_fix"
[467] "chr1_MU273336v1_fix"     "chr2_KN538362v1_fix"
[469] "chr2_KN538363v1_fix"     "chr2_KQ031384v1_fix"
[471] "chr2_ML143341v1_fix"     "chr2_ML143342v1_fix"
[473] "chr2_MU273341v1_fix"     "chr2_MU273342v1_fix"
[475] "chr2_MU273343v1_fix"     "chr2_MU273344v1_fix"
[477] "chr2_MU273345v1_fix"     "chr3_KN196475v1_fix"
[479] "chr3_KN196476v1_fix"     "chr3_KN538364v1_fix"
[481] "chr3_KQ031385v1_fix"     "chr3_KQ031386v1_fix"
[483] "chr3_KV766192v1_fix"     "chr3_KZ559104v1_fix"
[485] "chr3_MU273346v1_fix"     "chr3_MU273347v1_fix"
[487] "chr3_MU273348v1_fix"     "chr4_KQ983257v1_fix"
[489] "chr4_ML143344v1_fix"     "chr4_ML143345v1_fix"
[491] "chr4_ML143346v1_fix"     "chr4_ML143347v1_fix"
[493] "chr4_ML143348v1_fix"     "chr4_ML143349v1_fix"
[495] "chr4_MU273350v1_fix"     "chr4_MU273351v1_fix"
[497] "chr5_KV575244v1_fix"     "chr5_ML143350v1_fix"
[499] "chr5_MU273352v1_fix"     "chr5_MU273353v1_fix"
[501] "chr5_MU273354v1_fix"     "chr5_MU273355v1_fix"
[503] "chr6_KN196478v1_fix"     "chr6_KQ031387v1_fix"
[505] "chr6_KQ090016v1_fix"     "chr6_KV766194v1_fix"
[507] "chr6_KZ208911v1_fix"     "chr6_ML143351v1_fix"
[509] "chr7_KQ031388v1_fix"     "chr7_KV880764v1_fix"
[511] "chr7_KV880765v1_fix"     "chr7_KZ208912v1_fix"
[513] "chr7_ML143352v1_fix"     "chr8_KV880766v1_fix"
[515] "chr8_KV880767v1_fix"     "chr8_KZ208914v1_fix"
[517] "chr8_KZ208915v1_fix"     "chr8_MU273359v1_fix"
[519] "chr8_MU273360v1_fix"     "chr8_MU273361v1_fix"
[521] "chr8_MU273362v1_fix"     "chr8_MU273363v1_fix"
[523] "chr9_KN196479v1_fix"     "chr9_ML143353v1_fix"
[525] "chr9_MU273364v1_fix"     "chr9_MU273365v1_fix"
[527] "chr9_MU273366v1_fix"     "chr10_KN196480v1_fix"
[529] "chr10_KN538365v1_fix"    "chr10_KN538366v1_fix"
[531] "chr10_KN538367v1_fix"    "chr10_KQ090021v1_fix"
[533] "chr10_ML143354v1_fix"    "chr10_ML143355v1_fix"
[535] "chr10_MU273367v1_fix"    "chr11_KN196481v1_fix"
[537] "chr11_KQ090022v1_fix"    "chr11_KQ759759v1_fix"
[539] "chr11_KQ759759v2_fix"    "chr11_KV766195v1_fix"
[541] "chr11_KZ559108v1_fix"    "chr11_KZ559109v1_fix"
[543] "chr11_ML143356v1_fix"    "chr11_ML143357v1_fix"
[545] "chr11_ML143358v1_fix"    "chr11_ML143359v1_fix"
[547] "chr11_ML143360v1_fix"    "chr11_MU273369v1_fix"
[549] "chr11_MU273370v1_fix"    "chr11_MU273371v1_fix"
[551] "chr12_KN196482v1_fix"    "chr12_KN538369v1_fix"
[553] "chr12_KN538370v1_fix"    "chr12_KQ759760v1_fix"
[555] "chr12_KZ208916v1_fix"    "chr12_KZ208917v1_fix"
[557] "chr12_ML143361v1_fix"    "chr12_ML143362v1_fix"
[559] "chr12_MU273372v1_fix"    "chr13_KN196483v1_fix"
[561] "chr13_KN538371v1_fix"    "chr13_KN538372v1_fix"
[563] "chr13_KN538373v1_fix"    "chr13_ML143363v1_fix"
[565] "chr13_ML143364v1_fix"    "chr13_ML143365v1_fix"
[567] "chr13_ML143366v1_fix"    "chr14_KZ208920v1_fix"
[569] "chr14_ML143367v1_fix"    "chr14_MU273373v1_fix"
[571] "chr15_KN538374v1_fix"    "chr15_ML143369v1_fix"
[573] "chr15_ML143370v1_fix"    "chr15_ML143371v1_fix"
[575] "chr15_ML143372v1_fix"    "chr15_MU273374v1_fix"
[577] "chr16_KV880768v1_fix"    "chr16_KZ559113v1_fix"
[579] "chr16_ML143373v1_fix"    "chr16_MU273376v1_fix"
[581] "chr16_MU273377v1_fix"    "chr17_KV575245v1_fix"
[583] "chr17_KV766196v1_fix"    "chr17_ML143374v1_fix"
[585] "chr17_ML143375v1_fix"    "chr17_MU273379v1_fix"
[587] "chr17_MU273380v1_fix"    "chr17_MU273381v1_fix"
[589] "chr17_MU273382v1_fix"    "chr17_MU273383v1_fix"
[591] "chr18_KQ090028v1_fix"    "chr18_KZ208922v1_fix"
[593] "chr18_KZ559115v1_fix"    "chr19_KN196484v1_fix"
[595] "chr19_KQ458386v1_fix"    "chr19_ML143376v1_fix"
[597] "chr19_MU273384v1_fix"    "chr19_MU273385v1_fix"
[599] "chr19_MU273386v1_fix"    "chr20_MU273388v1_fix"
[601] "chr20_MU273389v1_fix"    "chr21_ML143377v1_fix"
[603] "chr21_MU273390v1_fix"    "chr21_MU273391v1_fix"
[605] "chr21_MU273392v1_fix"    "chr22_KQ759762v1_fix"
[607] "chr22_KQ759762v2_fix"    "chr22_ML143378v1_fix"
[609] "chr22_ML143379v1_fix"    "chr22_ML143380v1_fix"
[611] "chrX_ML143381v1_fix"     "chrX_ML143382v1_fix"
[613] "chrX_ML143383v1_fix"     "chrX_ML143384v1_fix"
[615] "chrX_ML143385v1_fix"     "chrX_MU273393v1_fix"
[617] "chrX_MU273394v1_fix"     "chrY_KN196487v1_fix"
[619] "chrY_KZ208923v1_fix"     "chrY_KZ208924v1_fix"
[621] "chrY_MU273398v1_fix"     "chr1_KQ458382v1_alt"
[623] "chr1_KQ458383v1_alt"     "chr1_KQ458384v1_alt"
[625] "chr1_KQ983255v1_alt"     "chr1_KV880763v1_alt"
[627] "chr1_KZ208904v1_alt"     "chr1_KZ208905v1_alt"
[629] "chr1_MU273330v1_alt"     "chr1_MU273331v1_alt"
[631] "chr1_MU273332v1_alt"     "chr2_KQ983256v1_alt"
[633] "chr2_KZ208907v1_alt"     "chr2_KZ208908v1_alt"
[635] "chr2_MU273337v1_alt"     "chr2_MU273338v1_alt"
[637] "chr2_MU273339v1_alt"     "chr2_MU273340v1_alt"
[639] "chr3_KZ208909v1_alt"     "chr3_KZ559101v1_alt"
[641] "chr3_KZ559102v1_alt"     "chr3_KZ559103v1_alt"
[643] "chr3_KZ559105v1_alt"     "chr3_ML143343v1_alt"
[645] "chr4_KQ090013v1_alt"     "chr4_KQ090014v1_alt"
[647] "chr4_KQ090015v1_alt"     "chr4_KQ983258v1_alt"
[649] "chr4_KV766193v1_alt"     "chr4_MU273349v1_alt"
[651] "chr5_KN196477v1_alt"     "chr5_KV575243v1_alt"
[653] "chr5_KZ208910v1_alt"     "chr5_MU273356v1_alt"
[655] "chr6_KQ090017v1_alt"     "chr6_MU273357v1_alt"
[657] "chr7_KZ208913v1_alt"     "chr7_KZ559106v1_alt"
[659] "chr7_MU273358v1_alt"     "chr8_KZ559107v1_alt"
[661] "chr9_KQ090018v1_alt"     "chr9_KQ090019v1_alt"
[663] "chr10_KQ090020v1_alt"    "chr11_KN538368v1_alt"
[665] "chr11_KZ559110v1_alt"    "chr11_KZ559111v1_alt"
[667] "chr11_MU273368v1_alt"    "chr12_KQ090023v1_alt"
[669] "chr12_KZ208918v1_alt"    "chr12_KZ559112v1_alt"
[671] "chr13_KQ090024v1_alt"    "chr13_KQ090025v1_alt"
[673] "chr14_KZ208919v1_alt"    "chr14_ML143368v1_alt"
[675] "chr15_KQ031389v1_alt"    "chr15_MU273375v1_alt"
[677] "chr16_KQ031390v1_alt"    "chr16_KQ090026v1_alt"
[679] "chr16_KQ090027v1_alt"    "chr16_KZ208921v1_alt"
[681] "chr17_KV766197v1_alt"    "chr17_KV766198v1_alt"
[683] "chr17_KZ559114v1_alt"    "chr17_MU273378v1_alt"
[685] "chr18_KQ458385v1_alt"    "chr18_KZ559116v1_alt"
[687] "chr19_KV575246v1_alt"    "chr19_KV575247v1_alt"
[689] "chr19_KV575248v1_alt"    "chr19_KV575249v1_alt"
[691] "chr19_KV575250v1_alt"    "chr19_KV575251v1_alt"
[693] "chr19_KV575252v1_alt"    "chr19_KV575253v1_alt"
[695] "chr19_KV575254v1_alt"    "chr19_KV575255v1_alt"
[697] "chr19_KV575256v1_alt"    "chr19_KV575257v1_alt"
[699] "chr19_KV575258v1_alt"    "chr19_KV575259v1_alt"
[701] "chr19_KV575260v1_alt"    "chr19_MU273387v1_alt"
[703] "chr22_KN196485v1_alt"    "chr22_KN196486v1_alt"
[705] "chr22_KQ458387v1_alt"    "chr22_KQ458388v1_alt"
[707] "chr22_KQ759761v1_alt"    "chrX_KV766199v1_alt"
[709] "chrX_MU273395v1_alt"     "chrX_MU273396v1_alt"
[711] "chrX_MU273397v1_alt"    </code></pre>
</div>
<p>Similarly, the function <code>seqinfo()</code> can be used to get the
full sequence information stored in the object.</p>
<div class="codewrapper sourceCode" id="cb43">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">seqinfo</span><span class="op">(</span><span class="va">genome</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Seqinfo object with 711 sequences (1 circular) from hg38 genome:
  seqnames             seqlengths isCircular genome
  chr1                  248956422      FALSE   hg38
  chr2                  242193529      FALSE   hg38
  chr3                  198295559      FALSE   hg38
  chr4                  190214555      FALSE   hg38
  chr5                  181538259      FALSE   hg38
  ...                         ...        ...    ...
  chr22_KQ759761v1_alt     145162      FALSE   hg38
  chrX_KV766199v1_alt      188004      FALSE   hg38
  chrX_MU273395v1_alt      619716      FALSE   hg38
  chrX_MU273396v1_alt      294119      FALSE   hg38
  chrX_MU273397v1_alt      330493      FALSE   hg38</code></pre>
</div>
<p>Finally, the nature of BSgenome objects being akin to a list of
sequences, the operators <code>$</code> and <code>[[]]</code> can both
be used to extract individual sequences from the BSgenome object.</p>
<div class="codewrapper sourceCode" id="cb45">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genome</span><span class="op">$</span><span class="va">chr1</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>248956422-letter MaskedDNAString object (# for masking)
seq: ####################################...####################################
masks:
  maskedwidth  maskedratio active names                               desc
1    18470101 7.419010e-02   TRUE AGAPS                      assembly gaps
2        5309 2.132502e-05   TRUE   AMB           intra-contig ambiguities
3   119060341 4.782377e-01  FALSE    RM                       RepeatMasker
4     1647959 6.619468e-03  FALSE   TRF Tandem Repeats Finder [period&lt;=12]
all masks together:
  maskedwidth maskedratio
    137685771   0.5530517
all active masks together:
  maskedwidth maskedratio
     18475410  0.07421142</code></pre>
</div>
<p>For instance, we can extract the sequence of the Y chromosome and
assign it to a new object <code>chrY</code>.</p>
<div class="codewrapper sourceCode" id="cb47">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chrY</span> <span class="op">&lt;-</span> <span class="va">genome</span><span class="op">[[</span><span class="st">"chrY"</span><span class="op">]</span><span class="op">]</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="using-genome-sequences">Using genome sequences<a class="anchor" aria-label="anchor" href="#using-genome-sequences"></a>
</h3>
<p>From this point, genome sequences can be treated very much like
biological strings (e.g. <code>DNAString</code>) described earlier, in
the <em><a href="https://bioconductor.org/packages/3.19/Biostrings" class="external-link">Biostrings</a></em>
package.</p>
<p>For instance, the function <code>countPattern()</code> can be used to
count the number of occurences of a given pattern in a given genome
sequence.</p>
<div class="codewrapper sourceCode" id="cb48">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">countPattern</span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"CANNTG"</span>, subject <span class="op">=</span> <span class="va">chrY</span>, fixed <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 141609</code></pre>
</div>
<div id="note" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="note" class="callout-inner">
<h3 class="callout-title">Note</h3>
<div class="callout-content">
<p>In the example above, the argument <code>fixed = FALSE</code> is used
to indicate that the pattern contain <a href="https://en.wikipedia.org/wiki/Nucleic_acid_notation#IUPAC_notation" class="external-link">IUPAC
ambiguity codes</a>.</p>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>The <code>Biostrings</code> package defines classes to represent
sequences of nucleotides and amino acids.</li>
<li>The <code>Biostrings</code> package also defines methods to
efficiently process biological sequences.</li>
<li>The <code>BSgenome</code> package provides genome sequences for a
range of model organisms immediately available as Bioconductor
objects.</li>
</ul>
</div>
</div>
</div>
</div>
</section></section><section id="aio-07-genomic-ranges"><p>Content from <a href="07-genomic-ranges.html">ゲノミクスの範囲での作業</a></p>
<hr>
<p>Last updated on 2025-07-22 |

        <a href="https://github.com/carpentries-incubator/bioc-project/edit/main/episodes/07-genomic-ranges.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> NA minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>Bioconductorでゲノムスケールの座標を表現する推奨方法は何ですか？</li>
<li>ゲノムの範囲を効率的に処理するためのメソッドを提供するBioconductorパッケージはどれですか？</li>
<li>さまざまなゲノムファイル形式から/へのゲノム座標のセットをインポート/エクスポートする方法は？</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>ゲノム座標と区間がBioconductorプロジェクトでどのように表現されるかを説明します。</li>
<li>ゲノム座標の範囲を処理するために利用可能なBioconductorパッケージおよびメソッドを特定します。</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="パッケージのインストール">パッケージのインストール<a class="anchor" aria-label="anchor" href="#%E3%83%91%E3%83%83%E3%82%B1%E3%83%BC%E3%82%B8%E3%81%AE%E3%82%A4%E3%83%B3%E3%82%B9%E3%83%88%E3%83%BC%E3%83%AB"></a>
</h2>
<hr class="half-width">
<p>次のセクションに進む前に、必要となるいくつかのBioconductorパッケージをインストールします。
まず、<em><a href="https://bioconductor.org/packages/3.19/BiocManager" class="external-link">BiocManager</a></em>パッケージがインストールされているか確認し、ない場合はインストールします。
次に、<code>BiocManager::install()</code>関数を使用して必要なパッケージをインストールします。</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu">requireNamespace</span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu">install.packages</span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"GenomicRanges"</span><span class="op">)</span></span></code></pre>
</div>
</section><section><h2 class="section-heading" id="genomicrangesパッケージとクラス">GenomicRangesパッケージとクラス<a class="anchor" aria-label="anchor" href="#genomicranges%E3%83%91%E3%83%83%E3%82%B1%E3%83%BC%E3%82%B8%E3%81%A8%E3%82%AF%E3%83%A9%E3%82%B9"></a>
</h2>
<hr class="half-width">
<div class="section level3">
<h3 id="ゲノム範囲のためにクラスが必要な理由は">ゲノム範囲のためにクラスが必要な理由は？<a class="anchor" aria-label="anchor" href="#%E3%82%B2%E3%83%8E%E3%83%A0%E7%AF%84%E5%9B%B2%E3%81%AE%E3%81%9F%E3%82%81%E3%81%AB%E3%82%AF%E3%83%A9%E3%82%B9%E3%81%8C%E5%BF%85%E8%A6%81%E3%81%AA%E7%90%86%E7%94%B1%E3%81%AF"></a>
</h3>
<p>ゲノミクスの時代に、多くの観察は遺伝子の座標の範囲、すなわちゲノムスケールの区間として報告されます。
アッセイの性質に応じて、これらのゲノム範囲は遺伝子、トランスクリプト、エクソン、一塩基多型（SNP）、転写因子結合部位、またはChIP-seqやATAC-seqなどの次世代シーケンシングアッセイからのピークを表す場合があります。</p>
<p>ゲノム範囲は、アッセイされた値（例：遺伝子発現）の観察をゲノムまたは生物内の物理的位置に結びつけます。
例えば、これらのゲノム範囲は、アッセイされた特徴と既知の調節領域のデータベースとの物理的近接性や重複を照会するために使用することができます。</p>
<p>最終的に報告される測定値に使用されるゲノム範囲は、既知のゲノム特徴のデータベース内のゲノム範囲のセットに対する組み合わせや操作の結果であることが多いです。
例えば、RNAシーケンシングにおいて、次世代シーケンシングリードはしばしば個々のエクソン内でカウントされ、それらのカウントは後に各遺伝子のすべてのエクソンで集約されます。
また、プロモーターは、既知の転写開始サイト（TSS）の上流および/または下流の任意の幅の領域として定義されることが多いです。</p>
<p>重要なことに、ゲノム範囲は必ずしも複数の座標を跨ぐ必要はありません。
範囲の概念は数学的な方法で意味されており、単一塩基のゲノム範囲（例：SNP）は、同じ座標で開閉するか（または右開区間の場合は次の座標で開閉する）と説明できます。</p>
<p>多くの生物において、遺伝物質は複数の別々の核酸分子（例：染色体、プラスミド）に分割されています。
そのため、ゲノム範囲は配列の名前とその配列上の座標の数値区間によって記述されます。</p>
<figure><img src="../fig/bioc-genomicranges.svg" alt="GenomicRanges代数の使用例。" class="figure mx-auto d-block"></figure><p><strong>GenomicRanges代数の使用例。</strong> Huber, Carey, Gentleman,
Anders, Carlson, Carvalho, Bravo, Davis, Gatto, Girke, Gottardo, Hahne,
Hansen, Irizarry, Lawrence, Love, MacDonald, Obenchain, Oles, Pages,
Reyes, Shannon, Smyth, Tenenbaum, Waldron, and Morgan (2015)から適応。
この図は、2つのトランスクリプトからなる遺伝子モデルの例と、その遺伝子モデルに対するさまざまなゲノム範囲の定義を示しています。
例えば、特定の図示において、スプライスされていないトランスクリプトは、最初のエクソンの開始から最後のエクソンの終了までのすべての座標の範囲を要約します。一方、遺伝子領域は、少なくとも1つのトランスクリプトの1エクソンに含まれる座標のセットとして定義されます。</p>
</div>
<div class="section level3">
<h3 id="区間の簡単な紹介">区間の簡単な紹介<a class="anchor" aria-label="anchor" href="#%E5%8C%BA%E9%96%93%E3%81%AE%E7%B0%A1%E5%8D%98%E3%81%AA%E7%B4%B9%E4%BB%8B"></a>
</h3>
<p>区間は、連続した座標の軸上の開始位置と終了位置を使用して数学的に説明されます。
区間は、その二つの座標の間のすべての実数を含み、各区間の幅は開始位置と終了位置の座標の差から計算できます。</p>
<p>一般的に、開始と終了の位置は、任意の有理数、包括浮動小数点数になり得ます。
ただし、ゲノム学では、整数座標は通常、ポリマー（例：核酸、タンパク質）内のモノマー（例：ヌクレオチド、アミノ酸）の位置を表すために使用されます。</p>
<p>さまざまなルールを使用して区間を定義するパッケージ、データベース、プログラミング言語に出会うかもしれません。
Rでは、インデックスは1ベース（すなわちシーケンスの最初の位置は1）ですが、Pythonは0ベース（すなわちシーケンスの最初の位置のインデックスは0）です。
同様に、UCSCゲノムブラウザの参照ファイルは0ベースであり、Ensemblゲノムブラウザのものは1ベースです。</p>
<p>共有座標系における区間の定義は、二つの区間間の距離を計算したり、重複する区間を特定したりするなどの計算を可能にします。</p>
<figure><img src="../fig/intervals.svg" alt="区間の例。" class="figure mx-auto d-block"></figure><p><strong>区間の例。</strong>
A、B、およびCという名前の3つの区間が表されています。
区間Aは位置5から始まり、位置9で終わり、幅は4ユニットです。
区間Bは位置1から始まり、位置3で終わり、幅は2ユニットです。
区間Cは位置3から始まり、位置6で終わり、幅は3ユニットです。
区間AとCは、座標5から6の間で重複し、一方、区間BとCは座標3で会い、厳密には互いに重複していません。</p>
</div>
<div class="section level3">
<h3 id="ゲノム範囲の簡単な紹介">ゲノム範囲の簡単な紹介<a class="anchor" aria-label="anchor" href="#%E3%82%B2%E3%83%8E%E3%83%A0%E7%AF%84%E5%9B%B2%E3%81%AE%E7%B0%A1%E5%8D%98%E3%81%AA%E7%B4%B9%E4%BB%8B"></a>
</h3>
<p>ゲノム範囲は、本質的に生物学的シーケンス（例：染色体）上の数学的な区間の概念を拡張します。
つまり、ゲノム範囲は、それが存在する生物学的配列の名前と、その配列内でゲノム範囲が跨ぐ整数座標の範囲を結合します。
これは、異なる生物学的配列上の重複する範囲のゲノム特徴を区別するための重要な要素です。</p>
<p>さらに、DNA配列の二重鎖特性も、ゲノム範囲に対してストランド性の概念を加えます。
もし知られている場合、ゲノム特徴のストランド情報は重要な情報であり、追跡されるべきで、後続の解析に使用されるかもしれません。
たとえば、同じDNA配列の反対のストランドで共通の座標範囲を跨ぐゲノム範囲は、重なっていると見なされない場合があります（例：ストランド特異的な次世代シーケンシングアッセイの目的のため）。</p>
<p>ゲノム範囲は_閉じた_区間です -
開始位置と終了位置は区間に含まれます。核酸の例では、開始位置は区間内の最初のヌクレオチドを示し、終了位置は区間内の最後のヌクレオチドを示します。</p>
<figure><img src="../fig/genomic-intervals.svg" alt="ゲノム区間の例。" class="figure mx-auto d-block"></figure><p><strong>ゲノム範囲の例。</strong>
ゲノム範囲は、存在する生物学的配列の名前（ここでは「chr1」）と、その配列内での開始と終了の位置で定義されます。
ここで、数値位置は明示的に示されていませんが、核酸のシーケンスと座標が左から右に増加していることを示す矢印によって暗示されています。
この例では、ゲノム範囲を使用して個々のエクソンを記述でき、メタデータはそれらのエクソンをトランスクリプトと遺伝子にグループ化します。
さらに、エクソン、トランスクリプト、遺伝子のストランド性は、各ゲノム範囲の正確な位置を記述するための重要な情報です。</p>
</div>
</section><section><h2 class="section-heading" id="genomicrangesパッケージ">GenomicRangesパッケージ<a class="anchor" aria-label="anchor" href="#genomicranges%E3%83%91%E3%83%83%E3%82%B1%E3%83%BC%E3%82%B8"></a>
</h2>
<hr class="half-width">
<div class="section level3">
<h3 id="概要">概要<a class="anchor" aria-label="anchor" href="#%E6%A6%82%E8%A6%81"></a>
</h3>
<p><em><a href="https://bioconductor.org/packages/3.19/GenomicRanges" class="external-link">GenomicRanges</a></em>パッケージは、ゲノム範囲をS4オブジェクトとして表現するための<a href="reference.html#s4-class">S4クラス</a>を実装しています。</p>
<p>具体的には、<code>GRanges</code>クラスは、機能が存在するシーケンスの名前と、そのシーケンス内で機能が跨ぐ整数座標の範囲を含む区間のセットを保存するために設計されています。</p>
<p>より一般的には、<code>IRanges</code>クラスは、シーケンス名の概念なしで、整数座標の範囲にわたる区間のセットを保存するために設計されています。
このように、<code>GRanges</code>オブジェクトは、<code>IRanges</code>オブジェクトとシーケンス名のベクターを組み合わせたものであるに過ぎません。</p>
<p>これらのS4クラスは、自動的な有効性チェック機能を提供し、整数区間およびゲノム範囲上の一般的な操作を実装するさまざまなメソッドを提供します。
区間間の距離の計算から重複するゲノム範囲の特定まで。</p>
<p><em><a href="https://bioconductor.org/packages/3.19/GenomicRanges" class="external-link">GenomicRanges</a></em>パッケージで定義されている基本クラスの簡単な説明は、パッケージのヴィネットの1つに利用可能で、<code>vignette("GenomicRangesIntroduction")</code>としてアクセス可能。本より詳細な情報は、他のパッケージのヴィネットに提供され、<code>browseVignettes("GenomicRanges")</code>としてアクセス可能です。</p>
</div>
<div class="section level3">
<h3 id="最初のステップ">最初のステップ<a class="anchor" aria-label="anchor" href="#%E6%9C%80%E5%88%9D%E3%81%AE%E3%82%B9%E3%83%86%E3%83%83%E3%83%97"></a>
</h3>
<p>始めるには、パッケージをロードします。</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">GenomicRanges</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="irangesクラス">IRangesクラス<a class="anchor" aria-label="anchor" href="#iranges%E3%82%AF%E3%83%A9%E3%82%B9"></a>
</h3>
<p>多くの生物のゲノム空間は複数の配列（例：染色体）に分割されていますが、ゲノム範囲上での多くの操作は、整数位置のみが重要な個々の配列内で行われます。
<code>IRanges</code>クラスは、次の三つの情報のうち二つに基づいて定義される「単純な」範囲のためのコンテナを提供します：</p>
<ul>
<li>範囲の開始位置</li>
<li>範囲の幅</li>
<li>範囲の終了位置</li>
</ul>
<p><code>IRanges()</code>コンストラクタ関数は、<code>start=</code>,
<code>width=</code>,
<code>end=</code>の引数でこれら三つの情報を受け入れます。
例えば、開始位置と幅から二つの整数範囲を作成します：</p>
<ul>
<li>ある範囲は位置10から始まり、幅は10です。</li>
<li>別の範囲は位置15から始まり、幅は5です。</li>
</ul>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">demo_iranges</span> <span class="op">&lt;-</span> <span class="fu">IRanges</span><span class="op">(</span>start <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">10</span>, <span class="fl">15</span><span class="op">)</span>, width <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">10</span>, <span class="fl">5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">demo_iranges</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>IRanges object with 2 ranges and 0 metadata columns:
          start       end     width
      &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
  [1]        10        19        10
  [2]        15        19         5</code></pre>
</div>
<p>オブジェクトは、要求した_開始_と_幅_の情報だけでなく、他の二つの情報から自然に算出される_終了_位置も表示されることに注意します。</p>
<div id="チャレンジ" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="チャレンジ" class="callout-inner">
<h3 class="callout-title">チャレンジ</h3>
<div class="callout-content">
<p><code>IRanges()</code>コンストラクタ関数の<code>start=</code>および<code>end=</code>の引数を使用して、上記と同じ二つの範囲を作成します。</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  ソリューション
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">IRanges</span><span class="op">(</span>start <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">10</span>, <span class="fl">15</span><span class="op">)</span>, end <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">19</span>, <span class="fl">19</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>IRanges object with 2 ranges and 0 metadata columns:
          start       end     width
      &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
  [1]        10        19        10
  [2]        15        19         5</code></pre>
</div>
</div>
</div>
</div>
</div>
<p>すべての区間の開始および終了位置、ならびに各区間の幅は、<code>start()</code>,
<code>end()</code>および<code>width()</code>関数を使用して数値ベクターとして抽出できます。</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">start</span><span class="op">(</span><span class="va">demo_iranges</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 10 15</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">end</span><span class="op">(</span><span class="va">demo_iranges</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 19 19</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">width</span><span class="op">(</span><span class="va">demo_iranges</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 10  5</code></pre>
</div>
<p><code>IRanges</code>ファミリーのオブジェクトは<code>Vector</code>クラスを拡張し、インデックスの観点から一次元ベクターとして扱われます。
そのため、個々の範囲は通常のベクターのように整数インデックスで抽出できます。</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">demo_iranges</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>IRanges object with 1 range and 0 metadata columns:
          start       end     width
      &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
  [1]        10        19        10</code></pre>
</div>
</div>
<div class="section level3">
<h3 id="irangesのメタデータ">IRangesのメタデータ<a class="anchor" aria-label="anchor" href="#iranges%E3%81%AE%E3%83%A1%E3%82%BF%E3%83%87%E3%83%BC%E3%82%BF"></a>
</h3>
<p><code>IRanges</code>クラスは、各範囲のメタデータ情報を収容でき、名前は<code>names=</code>引数に渡され、さまざまなメタデータは名前付きベクターとして渡されます。</p>
<p>例えば、“A”と”B”という名前の二つの範囲を作成します。
さらに、例として文字値と数値を保存するメタデータフィールドを定義します。
この例では、メタデータフィールドの名前と値は完全に任意です。</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">demo_with_metadata</span> <span class="op">&lt;-</span> <span class="fu">IRanges</span><span class="op">(</span></span>
<span>  start <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">10</span>,  <span class="fl">15</span><span class="op">)</span>,</span>
<span>  end   <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">19</span>,  <span class="fl">19</span><span class="op">)</span>,</span>
<span>  names <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span><span class="op">)</span>,</span>
<span>  character_metadata <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"control"</span>, <span class="st">"target"</span><span class="op">)</span>,</span>
<span>  numeric_metadata <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">100</span>, <span class="fl">200</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">demo_with_metadata</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>IRanges object with 2 ranges and 2 metadata columns:
        start       end     width | character_metadata numeric_metadata
    &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; |        &lt;character&gt;        &lt;numeric&gt;
  A        10        19        10 |            control              100
  B        15        19         5 |             target              200</code></pre>
</div>
<p>メタデータ列は、<code>mcols()</code>関数を使用して<code>DataFrame</code>として抽出できます（“メタデータ列”の略）。</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">mcols</span><span class="op">(</span><span class="va">demo_with_metadata</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DataFrame with 2 rows and 2 columns
  character_metadata numeric_metadata
         &lt;character&gt;        &lt;numeric&gt;
A            control              100
B             target              200</code></pre>
</div>
<p>名前の文字ベクターは、<code>names()</code>関数を使用して抽出できます。</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">names</span><span class="op">(</span><span class="va">demo_with_metadata</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "A" "B"</code></pre>
</div>
<p>基本データ型の名前付きベクターに類似して、個々の範囲は名前で抽出できます。</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">demo_with_metadata</span><span class="op">[</span><span class="st">"A"</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>IRanges object with 1 range and 2 metadata columns:
        start       end     width | character_metadata numeric_metadata
    &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; |        &lt;character&gt;        &lt;numeric&gt;
  A        10        19        10 |            control              100</code></pre>
</div>
</div>
<div class="section level3">
<h3 id="irangesでの操作">IRangesでの操作<a class="anchor" aria-label="anchor" href="#iranges%E3%81%A7%E3%81%AE%E6%93%8D%E4%BD%9C"></a>
</h3>
<p><code>IRanges</code>は、数値座標の範囲に対するほとんどの操作の基礎を提供します。</p>
<p>例えば、クエリセットとサブジェクトセットの二つの範囲が与えられた場合、<code>findOVerlaps()</code>関数を使用して、二つのセットの範囲のどのペアが互いに重なっているかを確認できます。</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">query_iranges</span> <span class="op">&lt;-</span> <span class="fu">IRanges</span><span class="op">(</span></span>
<span>  start <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">8</span>, <span class="fl">16</span><span class="op">)</span>,</span>
<span>  end   <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">14</span>, <span class="fl">18</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">overlaps_iranges</span> <span class="op">&lt;-</span> <span class="fu">findOverlaps</span><span class="op">(</span>query <span class="op">=</span> <span class="va">query_iranges</span>, subject <span class="op">=</span> <span class="va">demo_iranges</span><span class="op">)</span></span>
<span><span class="va">overlaps_iranges</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Hits object with 3 hits and 0 metadata columns:
      queryHits subjectHits
      &lt;integer&gt;   &lt;integer&gt;
  [1]         1           1
  [2]         2           1
  [3]         2           2
  -------
  queryLength: 2 / subjectLength: 2</code></pre>
</div>
<p>結果は、まだ紹介していない<code>Hits</code>オブジェクトの形式で返されます。
<code>Hits</code>オブジェクトは、<code>queryHits</code>および<code>subjectHits</code>という二つの整数列からなるテーブルとして視覚化されます。
テーブルの各行は、クエリセット内の1つの範囲とサブジェクトセット内の1つの範囲の重複を報告し、各列の整数値は重複に関与する各セット内の範囲のインデックスを示します。</p>
<p>この例では、クエリセットの最初の範囲がサブジェクトセットの最初の範囲と重なっていることを確認し、クエリセットの二番目の範囲がサブジェクトセット内の両範囲と重なっています。</p>
<div id="さらに深く" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="さらに深く" class="callout-inner">
<h3 class="callout-title">さらに深く</h3>
<div class="callout-content">
<p>後処理用に、<code>Hits</code>オブジェクトから二つのコンポーネントを、それぞれの名前を使用して抽出できます：</p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">queryHits</span><span class="op">(</span><span class="va">overlaps_iranges</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 1 2 2</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">subjectHits</span><span class="op">(</span><span class="va">overlaps_iranges</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 1 1 2</code></pre>
</div>
<p>テーブルとして表示される間、<code>Hits</code>オブジェクトは実際にはベクターのように処理されます。
クエリ範囲とサブジェクト範囲の間の個々のヒットは、インデックスで抽出できます：</p>
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">overlaps_iranges</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Hits object with 1 hit and 0 metadata columns:
      queryHits subjectHits
      &lt;integer&gt;   &lt;integer&gt;
  [1]         1           1
  -------
  queryLength: 2 / subjectLength: 2</code></pre>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="grangesクラス">GRangesクラス<a class="anchor" aria-label="anchor" href="#granges%E3%82%AF%E3%83%A9%E3%82%B9"></a>
</h3>
<p>整数範囲を定義した後、ゲノム範囲を定義するために必要な追加情報は、各範囲が位置するゲノム配列の名前です。</p>
<p>例えば、次のように二つのゲノム範囲を定義します。</p>
<ul>
<li>1番染色体（略称”chr1”）の位置10から25の範囲</li>
<li>2番染色体（略称”chr2”）の位置20から35の範囲</li>
</ul>
<p>それを行うために、<code>GRanges()</code>コンストラクタ関数を使用します。
シーケンス名を<code>seqnames=</code>引数に、開始位置と終了位置の両方を<code>ranges=</code>引数に、<code>IRanges</code>オブジェクトとして提供します。</p>
<div class="codewrapper sourceCode" id="cb31">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">demo_granges</span> <span class="op">&lt;-</span> <span class="fu">GRanges</span><span class="op">(</span></span>
<span>  seqnames <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"chr1"</span>, <span class="st">"chr2"</span><span class="op">)</span>,</span>
<span>  ranges <span class="op">=</span> <span class="fu">IRanges</span><span class="op">(</span></span>
<span>    start <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">10</span>, <span class="fl">20</span><span class="op">)</span>,</span>
<span>    end   <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">25</span>, <span class="fl">35</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">demo_granges</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>GRanges object with 2 ranges and 0 metadata columns:
      seqnames    ranges strand
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
  [1]     chr1     10-25      *
  [2]     chr2     20-35      *
  -------
  seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
<p>コンソールで、オブジェクトは<code>seqnames</code>コンポーネントにシーケンス名を表示し、<code>ranges</code>コンポーネントに<code>start-end</code>形式で範囲を表示します。
さらに、上記の例は、<code>GRanges</code>オブジェクトが<code>strand</code>というコンポーネントを持ち、シンボル<code>*</code>がストランドのないゲノム範囲を示すことも示しています。これは、その情報が提供されていないためです。</p>
<p>ストランド情報は、<code>strand=</code>引数に提供できます。例えば：</p>
<div class="codewrapper sourceCode" id="cb33">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">demo_granges2</span> <span class="op">&lt;-</span> <span class="fu">GRanges</span><span class="op">(</span></span>
<span>  seqnames <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"chr1"</span>, <span class="st">"chr2"</span><span class="op">)</span>,</span>
<span>  ranges <span class="op">=</span> <span class="fu">IRanges</span><span class="op">(</span></span>
<span>    start <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">10</span>, <span class="fl">20</span><span class="op">)</span>,</span>
<span>    end   <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">25</span>, <span class="fl">35</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  strand  <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"+"</span>, <span class="st">"-"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">demo_granges2</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>GRanges object with 2 ranges and 0 metadata columns:
      seqnames    ranges strand
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
  [1]     chr1     10-25      +
  [2]     chr2     20-35      -
  -------
  seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
<p>最後に、上記の例は、<code>GRanges</code>オブジェクトに<code>seqinfo</code>というコンポーネントが含まれていることを示しています。これは、<code>seqnames</code>コンポーネントに含まれる各シーケンスに関する情報を保存するために使用されます。
前の例では、シーケンスについての情報を提供していません。
そのため、<code>seqinfo</code>コンポーネントはオブジェクトを作成するために使用したシーケンスの名前で自動的に埋められ、残りの情報は指定されていないままとなっています。</p>
<div class="codewrapper sourceCode" id="cb35">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">seqinfo</span><span class="op">(</span><span class="va">demo_granges2</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Seqinfo object with 2 sequences from an unspecified genome; no seqlengths:
  seqnames seqlengths isCircular genome
  chr1             NA         NA   &lt;NA&gt;
  chr2             NA         NA   &lt;NA&gt;</code></pre>
</div>
<p>上記の例は、シーケンスに関する情報に、それぞれの名前と長さだけでなく、それらが環状ポリマー（例：プラスミド）を表すかどうか、またそれらが属するゲノムの名前が含まれていることも示しています。</p>
<p>この情報は、オブジェクトの作成時にコンストラクタに直接提供したり、既存のオブジェクトに対して<code>seqinfo()</code>アクセサと<code>Seqinfo()</code>コンストラクタを使用して編集することができます。</p>
<div class="codewrapper sourceCode" id="cb37">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">seqinfo</span><span class="op">(</span><span class="va">demo_granges2</span><span class="op">)</span> <span class="op">&lt;-</span>  <span class="fu">Seqinfo</span><span class="op">(</span></span>
<span>    seqnames <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"chr1"</span>, <span class="st">"chr2"</span><span class="op">)</span>,</span>
<span>    seqlengths <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">1234</span>, <span class="fl">5678</span><span class="op">)</span>,</span>
<span>    isCircular <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="cn">FALSE</span>, <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>    genome <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"homo_sapiens"</span>, <span class="st">"homo_sapiens"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">demo_granges2</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>GRanges object with 2 ranges and 0 metadata columns:
      seqnames    ranges strand
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
  [1]     chr1     10-25      +
  [2]     chr2     20-35      -
  -------
  seqinfo: 2 sequences (1 circular) from homo_sapiens genome</code></pre>
</div>
<p>各範囲の個々の開始および終了位置、ならびにすべての区間の幅は、<code>start()</code>,
<code>end()</code>および<code>width()</code>関数を使用して数値ベクターとして抽出できます。</p>
<div class="codewrapper sourceCode" id="cb39">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">start</span><span class="op">(</span><span class="va">demo_granges2</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 10 20</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb41">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">end</span><span class="op">(</span><span class="va">demo_granges2</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 25 35</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb43">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">width</span><span class="op">(</span><span class="va">demo_granges2</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 16 16</code></pre>
</div>
<p>シーケンス名とストランド情報は、それぞれ<code>seqnames()</code>および<code>strand()</code>の関数を使用して抽出できます。</p>
<div class="codewrapper sourceCode" id="cb45">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">seqnames</span><span class="op">(</span><span class="va">demo_granges2</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>factor-Rle of length 2 with 2 runs
  Lengths:    1    1
  Values : chr1 chr2
Levels(2): chr1 chr2</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb47">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">strand</span><span class="op">(</span><span class="va">demo_granges2</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>factor-Rle of length 2 with 2 runs
  Lengths: 1 1
  Values : + -
Levels(3): + - *</code></pre>
</div>
</div>
<div class="section level3">
<h3 id="grangesのメタデータ">GRangesのメタデータ<a class="anchor" aria-label="anchor" href="#granges%E3%81%AE%E3%83%A1%E3%82%BF%E3%83%87%E3%83%BC%E3%82%BF"></a>
</h3>
<p><code>IRanges</code>と同様に、メタデータは直接<code>GRanges</code>コンストラクタ関数に渡すことができます。
例えば：</p>
<div class="codewrapper sourceCode" id="cb49">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">demo_granges3</span> <span class="op">&lt;-</span> <span class="fu">GRanges</span><span class="op">(</span></span>
<span>  seqnames <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"chr1"</span>, <span class="st">"chr2"</span><span class="op">)</span>,</span>
<span>  ranges <span class="op">=</span> <span class="fu">IRanges</span><span class="op">(</span></span>
<span>    start <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">10</span>, <span class="fl">20</span><span class="op">)</span>,</span>
<span>    end <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">25</span>, <span class="fl">35</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  metadata1 <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"control"</span>, <span class="st">"target"</span><span class="op">)</span>,</span>
<span>  metadata2 <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">demo_granges3</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>GRanges object with 2 ranges and 2 metadata columns:
      seqnames    ranges strand |   metadata1 metadata2
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;numeric&gt;
  [1]     chr1     10-25      * |     control         1
  [2]     chr2     20-35      * |      target         2
  -------
  seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<div class="section level3">
<h3 id="ファイルからのゲノム範囲のインポート">ファイルからのゲノム範囲のインポート<a class="anchor" aria-label="anchor" href="#%E3%83%95%E3%82%A1%E3%82%A4%E3%83%AB%E3%81%8B%E3%82%89%E3%81%AE%E3%82%B2%E3%83%8E%E3%83%A0%E7%AF%84%E5%9B%B2%E3%81%AE%E3%82%A4%E3%83%B3%E3%83%9D%E3%83%BC%E3%83%88"></a>
</h3>
<p>頻繁に、大規模なゲノム範囲のコレクションは、手動で記述されたコードではなく、ファイルからインポートされます。
特に、既知の遺伝子機能のゲノム全体の注釈は、<a href="https://www.ensembl.org/info/data/ftp/" class="external-link">Ensembl FTP</a> や <a href="https://hgdownload.soe.ucsc.edu/downloads.html" class="external-link">UCSC Genome
Data</a> のようなウェブサイトにファイルとして配布されています。</p>
<p>さまざまなファイル形式が、バイオインフォマティクスのワークフローでゲノム範囲を格納するために一般的に使用されています。
例えば、BED (Browser Extensible Data)
形式は、クロマチン免疫沈降シーケンシング (ChIP-Seq)
で一般的に見られますが、GTF (Gene Transfer Format, GTF2.2)
は、エクソン、転写物、遺伝子などのゲノム機能を記述するための_デファクト_標準ファイル形式です。</p>
<p>次の例では、私たちは小さなGTFファイルからアクチンベータ (ACTB)
の遺伝子モデルを一連のゲノム範囲としてインポートします。
例示のファイルは、<a href="https://www.ensembl.org/info/data/ftp/" class="external-link">Ensembl FTP</a>
サイトからダウンロードされた_Homo
sapiens_種のGTFファイルのサブセットを表しています。
元のファイルは300万行以上と22のメタデータフィールドを持っており、その中からサブセットが抽出されてこのレッスン用の小さなファイルになりました。</p>
<p>特に、私たちは <code>import()</code> ジェネリックを使用します。
<em><a href="https://bioconductor.org/packages/3.19/BiocIO" class="external-link">BiocIO</a></em>
パッケージで定義され、<em><a href="https://bioconductor.org/packages/3.19/rtracklayer" class="external-link">rtracklayer</a></em>
パッケージで実装されたメソッドを用いて、一般的なファイル拡張子を認識し、それぞれの特定のファイル形式を解析するための適切なメソッドに関連付けることができる多目的関数です。</p>
<div class="codewrapper sourceCode" id="cb51">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">rtracklayer</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning: replacing previous import 'S4Arrays::makeNindexFromArrayViewport' by
'DelayedArray::makeNindexFromArrayViewport' when loading 'SummarizedExperiment'</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb53">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">actb_gtf_data</span> <span class="op">&lt;-</span> <span class="fu">rtracklayer</span><span class="fu">::</span><span class="fu">import</span><span class="op">(</span><span class="st">"data/actb.gtf"</span><span class="op">)</span></span>
<span><span class="va">actb_gtf_data</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>GRanges object with 267 ranges and 7 metadata columns:
        seqnames          ranges strand |      source           type     score
           &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; |    &lt;factor&gt;       &lt;factor&gt; &lt;numeric&gt;
    [1]        7 5526409-5563902      - | rtracklayer     gene              NA
    [2]        7 5526409-5530601      - | rtracklayer     transcript        NA
    [3]        7 5530542-5530601      - | rtracklayer     exon              NA
    [4]        7 5529535-5529684      - | rtracklayer     exon              NA
    [5]        7 5529535-5529657      - | rtracklayer     CDS               NA
    ...      ...             ...    ... .         ...            ...       ...
  [263]        7 5540676-5540771      - | rtracklayer five_prime_utr        NA
  [264]        7 5529658-5529663      - | rtracklayer five_prime_utr        NA
  [265]        7 5561852-5562716      - | rtracklayer transcript            NA
  [266]        7 5562390-5562716      - | rtracklayer exon                  NA
  [267]        7 5561852-5561949      - | rtracklayer exon                  NA
            phase         gene_id   gene_name   transcript_id
        &lt;integer&gt;     &lt;character&gt; &lt;character&gt;     &lt;character&gt;
    [1]      &lt;NA&gt; ENSG00000075624        ACTB            &lt;NA&gt;
    [2]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000674681
    [3]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000674681
    [4]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000674681
    [5]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000674681
    ...       ...             ...         ...             ...
  [263]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000414620
  [264]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000414620
  [265]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000646584
  [266]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000646584
  [267]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000646584
  -------
  seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
</div>
<div id="さらに進む" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="さらに進む" class="callout-inner">
<h3 class="callout-title">さらに進む</h3>
<div class="callout-content">
<p>特定のファイル形式を解析するための個別のメソッドを直接呼び出すことができます。
例えば、この場合、GTFファイル形式はGFFバージョン2ファイル形式と同一であるため、私たちは
<code>rtracklayer::import.gff2()</code>
関数を直接呼び出すことができます。同じ効果を得ることができます。</p>
<p>すべてのメソッドのフルリストは、<em><a href="https://bioconductor.org/packages/3.19/rtracklayer" class="external-link">rtracklayer</a></em>
パッケージのドキュメントを参照してください。</p>
</div>
</div>
</div>
<p>上記の例では、GTFファイルの内容が<code>GRanges</code>オブジェクトにインポートされました。
ファイルの各エントリについて、配列名、開始位置と終了位置、ストランド情報がオブジェクトの専用コンポーネントをポピュレートするために使用され、他のすべての情報はメタデータの別のカラムとして保存されます。</p>
<p>ここから、この<code>GRanges</code>オブジェクトは、私たちがこのエピソードの初めに作成した他の<code>GRanges</code>オブジェクトのように操作できます。</p>
</div>
<div class="section level3">
<h3 id="grangesおよびgrangeslistクラスの操作">GRangesおよびGRangesListクラスの操作<a class="anchor" aria-label="anchor" href="#granges%E3%81%8A%E3%82%88%E3%81%B3grangeslist%E3%82%AF%E3%83%A9%E3%82%B9%E3%81%AE%E6%93%8D%E4%BD%9C"></a>
</h3>
<p>これまでに示したように、<code>GRanges</code>オブジェクトは手動で定義またはファイルからインポートできます。
それらは、興味のあるゲノム領域を表し、既知のゲノム機能のデータベースをそれぞれ表します。
いずれにせよ、生物情報学のワークフロー全体で<code>GRanges</code>オブジェクトに一般的に適用される多数の操作があります。</p>
<div class="section level4">
<h4 id="サブセット">サブセット<a class="anchor" aria-label="anchor" href="#%E3%82%B5%E3%83%96%E3%82%BB%E3%83%83%E3%83%88"></a>
</h4>
<p>例えば、<code>subset()</code>メソッドは、配列名、開始位置と終了位置、ストランド、または任意のメタデータフィールドの条件に一致するゲノム範囲のセットを抽出するのに非常に便利です。
以下の例では、位置<code>5527147</code>で開始する<code>transcript</code>型のすべてのレコードを抽出します。</p>
<div class="codewrapper sourceCode" id="cb55">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">subset</span><span class="op">(</span><span class="va">actb_gtf_data</span>, <span class="va">type</span> <span class="op">==</span> <span class="st">"transcript"</span> <span class="op">&amp;</span> <span class="va">start</span> <span class="op">==</span> <span class="fl">5527147</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>GRanges object with 5 ranges and 7 metadata columns:
      seqnames          ranges strand |      source       type     score
         &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; |    &lt;factor&gt;   &lt;factor&gt; &lt;numeric&gt;
  [1]        7 5527147-5529949      - | rtracklayer transcript        NA
  [2]        7 5527147-5530581      - | rtracklayer transcript        NA
  [3]        7 5527147-5530604      - | rtracklayer transcript        NA
  [4]        7 5527147-5530604      - | rtracklayer transcript        NA
  [5]        7 5527147-5530604      - | rtracklayer transcript        NA
          phase         gene_id   gene_name   transcript_id
      &lt;integer&gt;     &lt;character&gt; &lt;character&gt;     &lt;character&gt;
  [1]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000642480
  [2]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000676397
  [3]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000676319
  [4]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000676189
  [5]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000473257
  -------
  seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<div class="section level4">
<h4 id="分割">分割<a class="anchor" aria-label="anchor" href="#%E5%88%86%E5%89%B2"></a>
</h4>
<p>別々に、<code>split()</code>メソッドは、最初に単一の<code>GRanges</code>オブジェクトに格納されていたゲノム範囲のセットを、名前付きの<code>GRanges</code>オブジェクトのリストに分割するのに役立ちます。
便利なことに、<code>GRangesList</code>クラスは、<code>GRanges</code>オブジェクトのリストを効率的に表示および処理するためのコンテナを提供します。</p>
<p>以下の例では、最初にエクソンを表すエントリのサブセットを抽出し、次にそれらのエクソンを転写物識別子で分割し、結果を<code>GRangesList</code>オブジェクトとして得ます。</p>
<div class="codewrapper sourceCode" id="cb57">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">actb_exons</span> <span class="op">&lt;-</span> <span class="fu">subset</span><span class="op">(</span><span class="va">actb_gtf_data</span>, <span class="va">type</span> <span class="op">==</span> <span class="st">"exon"</span><span class="op">)</span></span>
<span><span class="va">actb_exons_by_transcript</span> <span class="op">&lt;-</span> <span class="fu">split</span><span class="op">(</span><span class="va">actb_exons</span>, <span class="va">actb_exons</span><span class="op">$</span><span class="va">transcript_id</span><span class="op">)</span></span>
<span><span class="va">actb_exons_by_transcript</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>GRangesList object of length 23:
$ENST00000414620
GRanges object with 4 ranges and 7 metadata columns:
      seqnames          ranges strand |      source     type     score
         &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; |    &lt;factor&gt; &lt;factor&gt; &lt;numeric&gt;
  [1]        7 5562574-5562790      - | rtracklayer     exon        NA
  [2]        7 5540676-5540771      - | rtracklayer     exon        NA
  [3]        7 5529535-5529663      - | rtracklayer     exon        NA
  [4]        7 5529282-5529400      - | rtracklayer     exon        NA
          phase         gene_id   gene_name   transcript_id
      &lt;integer&gt;     &lt;character&gt; &lt;character&gt;     &lt;character&gt;
  [1]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000414620
  [2]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000414620
  [3]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000414620
  [4]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000414620
  -------
  seqinfo: 1 sequence from an unspecified genome; no seqlengths

$ENST00000417101
GRanges object with 3 ranges and 7 metadata columns:
      seqnames          ranges strand |      source     type     score
         &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; |    &lt;factor&gt; &lt;factor&gt; &lt;numeric&gt;
  [1]        7 5529806-5529982      - | rtracklayer     exon        NA
  [2]        7 5529535-5529663      - | rtracklayer     exon        NA
  [3]        7 5529235-5529400      - | rtracklayer     exon        NA
          phase         gene_id   gene_name   transcript_id
      &lt;integer&gt;     &lt;character&gt; &lt;character&gt;     &lt;character&gt;
  [1]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000417101
  [2]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000417101
  [3]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000417101
  -------
  seqinfo: 1 sequence from an unspecified genome; no seqlengths

$ENST00000425660
GRanges object with 7 ranges and 7 metadata columns:
      seqnames          ranges strand |      source     type     score
         &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; |    &lt;factor&gt; &lt;factor&gt; &lt;numeric&gt;
  [1]        7 5530524-5530601      - | rtracklayer     exon        NA
  [2]        7 5529535-5529663      - | rtracklayer     exon        NA
  [3]        7 5529161-5529400      - | rtracklayer     exon        NA
  [4]        7 5529019-5529059      - | rtracklayer     exon        NA
  [5]        7 5528281-5528719      - | rtracklayer     exon        NA
  [6]        7 5528004-5528185      - | rtracklayer     exon        NA
  [7]        7 5527156-5527891      - | rtracklayer     exon        NA
          phase         gene_id   gene_name   transcript_id
      &lt;integer&gt;     &lt;character&gt; &lt;character&gt;     &lt;character&gt;
  [1]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000425660
  [2]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000425660
  [3]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000425660
  [4]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000425660
  [5]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000425660
  [6]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000425660
  [7]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000425660
  -------
  seqinfo: 1 sequence from an unspecified genome; no seqlengths

...
&lt;20 more elements&gt;</code></pre>
</div>
<p>コンソールで上記のオブジェクトを印刷すると、最初の行はそのオブジェクトのクラスを<code>GRrangesList</code>として確認し、そのリスト内の各名前付き<code>GRanges</code>がドル記号とそのアイテムの名前によって紹介され、通常のRの名前付きリストのように表示されます。</p>
</div>
<div class="section level4">
<h4 id="長さ">長さ<a class="anchor" aria-label="anchor" href="#%E9%95%B7%E3%81%95"></a>
</h4>
<p>性質上、<code>list</code>オブジェクトに適用可能な多くのメソッドは、直接<code>GRangesList</code>オブジェクトに適用できます。
例えば、<code>lengths()</code>関数を<code>GRangesList</code>に使用して、リスト内の各<code>GRanges</code>オブジェクトの長さを整数ベクトルとして表示できます。</p>
<p>上記の最新の例では、各転写物内のエクソンの数を、<code>GRangesList</code>内の各<code>GRanges</code>オブジェクトの長さとして計算できます：</p>
<div class="codewrapper sourceCode" id="cb59">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">lengths</span><span class="op">(</span><span class="va">actb_exons_by_transcript</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>ENST00000414620 ENST00000417101 ENST00000425660 ENST00000432588 ENST00000443528
              4               3               7               5               3
ENST00000462494 ENST00000464611 ENST00000473257 ENST00000477812 ENST00000480301
              5               3               5               5               2
ENST00000484841 ENST00000493945 ENST00000642480 ENST00000645025 ENST00000645576
              5               6               5               4               5
ENST00000646584 ENST00000646664 ENST00000647275 ENST00000674681 ENST00000675515
              2               6               3               6               6
ENST00000676189 ENST00000676319 ENST00000676397
              6               3               6 </code></pre>
</div>
<div id="チャレンジ-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="チャレンジ-1" class="callout-inner">
<h3 class="callout-title">チャレンジ</h3>
<div class="callout-content">
<p>重要なことに、上記の <code>lengths()</code> 関数 (最終的に
<code>s</code> が付いています) は、<code>length()</code> 関数 (
<code>s</code> が付いていないもの) とは異なります。
前者はリストオブジェクトに対して使用されることを目的としており、リスト内の各要素の長さを返すベクトルを返します。後者はリスト自体の長さを表す単一の数値スカラーを返します
(つまり、リスト内の要素数)。</p>
<p><code>length(actb_exons_by_transcript)</code>
は何を返し、この数は生物学的に何を表しますか？</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">
  ソリューション
  </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb61">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">length</span><span class="op">(</span><span class="va">actb_exons_by_transcript</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 23</code></pre>
</div>
<p>このコードは、単一の整数値 <code>23</code>
を返します。それは、<code>GRangesList</code>オブジェクト内の<code>GRanges</code>の数であり、遺伝子ACTBの転写物の数でもあります。</p>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="オーバーラップによるサブセット">オーバーラップによるサブセット<a class="anchor" aria-label="anchor" href="#%E3%82%AA%E3%83%BC%E3%83%90%E3%83%BC%E3%83%A9%E3%83%83%E3%83%97%E3%81%AB%E3%82%88%E3%82%8B%E3%82%B5%E3%83%96%E3%82%BB%E3%83%83%E3%83%88"></a>
</h3>
<p>ゲノム範囲で作業するときの最も一般的な操作の一つは、特定のゲノム領域に位置するゲノム範囲の任意の大きなコレクションをサブセット化することです。たとえば、ゲノムブラウザに情報をトラックとして視覚化するときなどです。</p>
<p>示すために、私たちは興味のある領域を表す新しい<code>GRanges</code>を手動で定義し、その領域に重なる以前にインポートされたすべてのゲノム範囲をGTFファイルから抽出するために使用します。</p>
<div class="codewrapper sourceCode" id="cb63">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">region_of_interest</span> <span class="op">&lt;-</span> <span class="fu">GRanges</span><span class="op">(</span></span>
<span>    seqnames <span class="op">=</span> <span class="st">"7"</span>,</span>
<span>    ranges <span class="op">=</span> <span class="fu">IRanges</span><span class="op">(</span>start <span class="op">=</span> <span class="fl">5525830</span>, end <span class="op">=</span> <span class="fl">5531239</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">actb_in_region</span> <span class="op">&lt;-</span> <span class="fu">subsetByOverlaps</span><span class="op">(</span>x <span class="op">=</span> <span class="va">actb_gtf_data</span>, ranges <span class="op">=</span> <span class="va">region_of_interest</span><span class="op">)</span></span>
<span><span class="va">actb_in_region</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>GRanges object with 256 ranges and 7 metadata columns:
        seqnames          ranges strand |      source           type     score
           &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; |    &lt;factor&gt;       &lt;factor&gt; &lt;numeric&gt;
    [1]        7 5526409-5563902      - | rtracklayer     gene              NA
    [2]        7 5526409-5530601      - | rtracklayer     transcript        NA
    [3]        7 5530542-5530601      - | rtracklayer     exon              NA
    [4]        7 5529535-5529684      - | rtracklayer     exon              NA
    [5]        7 5529535-5529657      - | rtracklayer     CDS               NA
    ...      ...             ...    ... .         ...            ...       ...
  [252]        7 5529535-5529657      - | rtracklayer CDS                   NA
  [253]        7 5529655-5529657      - | rtracklayer start_codon           NA
  [254]        7 5529282-5529400      - | rtracklayer exon                  NA
  [255]        7 5529282-5529400      - | rtracklayer CDS                   NA
  [256]        7 5529658-5529663      - | rtracklayer five_prime_utr        NA
            phase         gene_id   gene_name   transcript_id
        &lt;integer&gt;     &lt;character&gt; &lt;character&gt;     &lt;character&gt;
    [1]      &lt;NA&gt; ENSG00000075624        ACTB            &lt;NA&gt;
    [2]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000674681
    [3]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000674681
    [4]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000674681
    [5]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000674681
    ...       ...             ...         ...             ...
  [252]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000414620
  [253]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000414620
  [254]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000414620
  [255]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000414620
  [256]      &lt;NA&gt; ENSG00000075624        ACTB ENST00000414620
  -------
  seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
</div>
<p><code>subset()</code>メソッドと同様に、<code>subsetByOverlaps()</code>メソッドは新しい<code>GRanges</code>オブジェクトを返します。
私たちは、新しいサブセットオブジェクト内の情報（新しいサブセットされたオブジェクト内に256の範囲があり、元のオブジェクトには267の範囲がある）を視覚的に比較することができ、あるいは新しい<code>GRanges</code>オブジェクトが元の<code>GRanges</code>オブジェクトよりも小さいかどうかを確認するために、プログラム的に2つのオブジェクトの長さを比較することもできます。</p>
<div class="codewrapper sourceCode" id="cb65">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">length</span><span class="op">(</span><span class="va">actb_in_region</span><span class="op">)</span> <span class="op">-</span> <span class="fu">length</span><span class="op">(</span><span class="va">actb_gtf_data</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] -11</code></pre>
</div>
<p>上記の例では、新しい<code>GRanges</code>オブジェクトは、元の<code>GRanges</code>オブジェクトよりも11レコード少ないことがわかります。</p>
<div id="さらに進む-1" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="さらに進む-1" class="callout-inner">
<h3 class="callout-title">さらに進む</h3>
<div class="callout-content">
<p><code>GRanges</code>および<code>GRangesList</code>オブジェクトで動作するための方法が他にもたくさんありますが、ここで示すことができるのはそれだけです。</p>
<p><code>GenomicRanges</code>パッケージで定義された関数の完全なリストは、パッケージのドキュメントのインデックスページで見つけることができ、<code><a href="https://rdrr.io/pkg/GenomicRanges/man" class="external-link">help(package="GenomicRanges")</a></code>を使用してアクセスできます。
<code>GenomicRanges</code>パッケージのビネットには、より多くの例やユースケースも見つけることができ、<code>browseVignettes("GenomicRanges")</code>を使用してアクセスできます。</p>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>
<code>GenomicRanges</code>パッケージは、ゲノムスケールでの座標の範囲を表すクラスを定義します。</li>
<li>
<code>GenomicRanges</code>パッケージは、ゲノム範囲を効率よく処理するためのメソッドも定義します。</li>
<li>
<code>rtracklayer</code>パッケージは、一般的なゲノムファイル形式からゲノム範囲をインポートおよびエクスポートするための関数を提供します。</li>
</ul>
</div>
</div>
</div>
</div>
</section></section><section id="aio-08-annotations"><p>Content from <a href="08-annotations.html">Working with annotations</a></p>
<hr>
<p>Last updated on 2025-07-22 |

        <a href="https://github.com/carpentries-incubator/bioc-project/edit/main/episodes/08-annotations.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> NA minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<hr>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What Bioconductor packages provides methods to efficiently fetch and
use gene annotations?</li>
<li>How can I use gene annotation packages to convert between different
gene identifiers?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Explain how gene annotations are managed in the Bioconductor
project.</li>
<li>Identify Bioconductor packages and methods available to fetch and
use gene annotations.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="install-packages">Install packages<a class="anchor" aria-label="anchor" href="#install-packages"></a>
</h2>
<hr class="half-width">
<p>Before we can proceed into the following sections, we install some
Bioconductor packages that we will need. First, we check that the <em><a href="https://bioconductor.org/packages/3.19/BiocManager" class="external-link">BiocManager</a></em>
package is installed before trying to use it; otherwise we install it.
Then we use the <code>BiocManager::install()</code> function to install
the necessary packages.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu">requireNamespace</span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu">install.packages</span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="st">"biomaRt"</span>, <span class="st">"org.Hs.eg.db"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
</section><section><h2 class="section-heading" id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<hr class="half-width">
<p>Packages dedicated to query gene annotations exist in the ‘Software’
and ‘Annotation’ categories of the Bioconductor <a href="https://www.bioconductor.org/packages/release/BiocViews.html" class="external-link">biocViews</a>,
according to their nature.</p>
<p>In the ‘Software’ section, we find packages that do not actually
contain gene annotations, but rather dynamically <em>query</em> them
from online resources (e.g.,<a href="https://www.ensembl.org/biomart/martview" class="external-link">Ensembl BioMart</a>).
One such Bioconductor package is <em><a href="https://bioconductor.org/packages/3.19/biomaRt" class="external-link">biomaRt</a></em>.</p>
<p>Instead, in the ‘Annotation’ section, we find packages that do
contain annotations. Examples include <em><a href="https://bioconductor.org/packages/3.19/org.Hs.eg.db" class="external-link">org.Hs.eg.db</a></em>,
<em><a href="https://bioconductor.org/packages/3.19/EnsDb.Hsapiens.v86" class="external-link">EnsDb.Hsapiens.v86</a></em>,
and <em><a href="https://bioconductor.org/packages/3.19/TxDb.Hsapiens.UCSC.hg38.knownGene" class="external-link">TxDb.Hsapiens.UCSC.hg38.knownGene</a></em>.</p>
<p>In this episode, we will demonstrate the two approaches:</p>
<ul>
<li>Querying annotations from the Ensembl Biomart API using the <em><a href="https://bioconductor.org/packages/3.19/biomaRt" class="external-link">biomaRt</a></em>
package.</li>
<li>Querying annotations from the <em><a href="https://bioconductor.org/packages/3.19/org.Hs.eg.db" class="external-link">org.Hs.eg.db</a></em>
annotation package.</li>
</ul></section><section><h2 class="section-heading" id="online-resources-or-bioconductor-annotation-packages">Online resources or Bioconductor annotation packages?<a class="anchor" aria-label="anchor" href="#online-resources-or-bioconductor-annotation-packages"></a>
</h2>
<hr class="half-width">
<div class="section level3">
<h3 id="accessing-the-latest-information">Accessing the latest information<a class="anchor" aria-label="anchor" href="#accessing-the-latest-information"></a>
</h3>
<p>Bioconductor’s 6-month release cycle implies that packages available
from the latest stable release branch will not be updated for six months
(only bugfixes are allowed, not functional updates). As a result,
annotation packages may contain information that is out-of-date by up to
six months.</p>
<p>Instead, independent online resources often have different policies
driving the release of updated information. Some databases are
frequently updated, while others may not have been updated in years.</p>
<p>Accessing up-to-date information must also be balanced with
reproducibility. Having downloaded the ‘latest’ information at one point
is time is no good if one hasn’t recorded at least <em>when</em> that
information was downloaded.</p>
</div>
<div class="section level3">
<h3 id="storage-requirements">Storage requirements<a class="anchor" aria-label="anchor" href="#storage-requirements"></a>
</h3>
<p>By nature, Bioconductor annotation packages are larger than software
packages. Just like any other R package, annotation packages must be
installed on the user’s computer before they can be used. This can
rapidly use add up to using an amount of disk space that is not
negligible.</p>
<p>Conversely, online resources are generally accessed programmatically,
and generally only require users to record code to replicate analyses
reproducibly.</p>
</div>
<div class="section level3">
<h3 id="internet-connectivity">Internet connectivity<a class="anchor" aria-label="anchor" href="#internet-connectivity"></a>
</h3>
<p>When using online resources, it is often a good idea to write
annotations ownloaded from online resources to a local file, and refer
to that local file during analyses.</p>
<p>If online resources were to become unavailable for any reason (e.g.,
downtime, loss of internet connection), analyses that use local files
can carry on while those that rely on those online resources cannot.</p>
<p>In contrast, Bioconductor annotation packages only require internet
connectivity at the time of installation. Once installed, they do not
require internet connectivity, as they rely on information stored
locally.</p>
</div>
<div class="section level3">
<h3 id="reproducibility">Reproducibility<a class="anchor" aria-label="anchor" href="#reproducibility"></a>
</h3>
<p>Bioconductor annotation packages are naturally versioned, meaning
that users can confidently report the version of the package used in
their analysis. Just like software packages, users control if and when
annotation packages should be updated on their computer.</p>
<p>Online resources have different policies to facilitate reproducible
analyses. Some online resources keep archived versions of their
annotations, allowing users to consistently access the same information
over time. When this is not the case, it may be necessary to download a
copy of the annotation at one point in time, and preciously keep that
copy throughout the lifetime of the project to ensure the use of a
consistent set of annotations.</p>
</div>
<div class="section level3">
<h3 id="consistency">Consistency<a class="anchor" aria-label="anchor" href="#consistency"></a>
</h3>
<p>As we will see in the practical examples of this episode,
Bioconductor annotation packages generally re-use a consistent set of
data structures. This allows users familiar with one annotation package
to rapidly get started with others.</p>
<p>Independent online resources often organise their data in different
ways, which requires users to write custom code to access, retrieve, and
process their respective data.</p>
</div>
</section><section><h2 class="section-heading" id="querying-annotations-from-ensembl-biomart">Querying annotations from Ensembl BioMart<a class="anchor" aria-label="anchor" href="#querying-annotations-from-ensembl-biomart"></a>
</h2>
<hr class="half-width">
<div class="section level3">
<h3 id="the-ensembl-biomart">The Ensembl BioMart<a class="anchor" aria-label="anchor" href="#the-ensembl-biomart"></a>
</h3>
<p><a href="https://www.ensembl.org/biomart/martview" class="external-link">Ensembl
BioMart</a> is a robust data mining tool designed to facilitate access
to the vast array of biological data available through the Ensembl
project.</p>
<p>The <a href="https://www.ensembl.org/biomart/martview" class="external-link">BioMart web
interface</a> enables researchers to efficiently query and retrieve data
on genes, proteins, and other genomic features across multiple species.
It allows users to filter, sort, and export data based on various
attributes such as gene IDs, chromosomal locations, and functional
annotations.</p>
</div>
<div class="section level3">
<h3 id="the-bioconductor-biomart-package">The Bioconductor <code>biomaRt</code> package<a class="anchor" aria-label="anchor" href="#the-bioconductor-biomart-package"></a>
</h3>
<p><em><a href="https://bioconductor.org/packages/3.19/biomaRt" class="external-link">biomaRt</a></em>
is a Bioconductor software package that enables retrieval of large
amounts of data from Ensembl BioMart tables directly from an R session
where those annotations can be used.</p>
<p>Let us first load the package:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">biomaRt</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="listing-available-marts">Listing available marts<a class="anchor" aria-label="anchor" href="#listing-available-marts"></a>
</h3>
<p>Ensembl BioMart organises its diverse biological information into
four databases also known as ‘marts’ or ‘biomarts’. Each mart focuses on
a different type of data.</p>
<p>Users must select the mart corresponds to the type of data they are
interested in before they can query any information from it.</p>
<p>The function <code>listMarts()</code> can be used to display the
names of those marts. This is convenient as users do not need to
memorise the name of the marts, and the function will also return an
updated list of names if any mart is renamed, added, or removed.</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">listMarts</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>               biomart                version
1 ENSEMBL_MART_ENSEMBL      Ensembl Genes 114
2   ENSEMBL_MART_MOUSE      Mouse strains 114
3     ENSEMBL_MART_SNP  Ensembl Variation 114
4 ENSEMBL_MART_FUNCGEN Ensembl Regulation 114</code></pre>
</div>
<p>In this demonstration, we will use the biomart called
<code>ENSEMBL_MART_ENSEMBL</code>, which contains the Ensembl gene
set.</p>
<p>Notably, the <code>version</code> columns also indicates the version
of the biomart. The Ensembl BioMart is updated regularly (multiple times
per year). By default, <em><a href="https://bioconductor.org/packages/3.19/biomaRt" class="external-link">biomaRt</a></em>
functions access the latest version of each biomart. This is not ideal
for reproducibility.</p>
<p>Thankfully, Ensembl BioMart archives past versions of its mars in a
way that is accessible both programmatically, and on its website.</p>
<p>The function <code>listEnsemblArchives()</code> can be used to
display all the versions of Ensembl Biomart accessible.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">listEnsemblArchives</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>             name     date                                 url version
1  Ensembl GRCh37 Feb 2014          https://grch37.ensembl.org  GRCh37
2     Ensembl 114 May 2025 https://may2025.archive.ensembl.org     114
3     Ensembl 113 Oct 2024 https://oct2024.archive.ensembl.org     113
4     Ensembl 112 May 2024 https://may2024.archive.ensembl.org     112
5     Ensembl 111 Jan 2024 https://jan2024.archive.ensembl.org     111
6     Ensembl 110 Jul 2023 https://jul2023.archive.ensembl.org     110
7     Ensembl 109 Feb 2023 https://feb2023.archive.ensembl.org     109
8     Ensembl 108 Oct 2022 https://oct2022.archive.ensembl.org     108
9     Ensembl 107 Jul 2022 https://jul2022.archive.ensembl.org     107
10    Ensembl 106 Apr 2022 https://apr2022.archive.ensembl.org     106
11    Ensembl 105 Dec 2021 https://dec2021.archive.ensembl.org     105
12    Ensembl 104 May 2021 https://may2021.archive.ensembl.org     104
13    Ensembl 103 Feb 2021 https://feb2021.archive.ensembl.org     103
14    Ensembl 102 Nov 2020 https://nov2020.archive.ensembl.org     102
15    Ensembl 101 Aug 2020 https://aug2020.archive.ensembl.org     101
16    Ensembl 100 Apr 2020 https://apr2020.archive.ensembl.org     100
17     Ensembl 99 Jan 2020 https://jan2020.archive.ensembl.org      99
18     Ensembl 80 May 2015 https://may2015.archive.ensembl.org      80
19     Ensembl 77 Oct 2014 https://oct2014.archive.ensembl.org      77
20     Ensembl 75 Feb 2014 https://feb2014.archive.ensembl.org      75
21     Ensembl 54 May 2009 https://may2009.archive.ensembl.org      54
   current_release
1
2                *
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21                </code></pre>
</div>
<p>In the output above, the key piece of information is the
<code>url</code> column, which provides the URL that <em><a href="https://bioconductor.org/packages/3.19/biomaRt" class="external-link">biomaRt</a></em>
functions will need to access data from the corresponding snapshot of
the Ensembl BioMart.</p>
<p>At the time of writing, the current release is Ensembl 112, so let us
use the corresponding url
<code>https://may2024.archive.ensembl.org</code> to ensure reproducible
results no matter when this lesson is delivered.</p>
</div>
<div class="section level3">
<h3 id="connecting-to-a-biomart">Connecting to a biomart<a class="anchor" aria-label="anchor" href="#connecting-to-a-biomart"></a>
</h3>
<p>The two pieces of information collected above – the name of a biomart
and the URL of a snapshot – is all that is needed to connect to a
BioMart database reproducibly.</p>
<p>The function <code>useMart()</code> can then be used to create a
connection. The connection is traditionally stored in an object called
<code>mart</code>, to be reused in subsequent steps for querying
information from the online mart.</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mart</span> <span class="op">&lt;-</span> <span class="fu">useMart</span><span class="op">(</span>biomart <span class="op">=</span> <span class="st">"ENSEMBL_MART_ENSEMBL"</span>, host <span class="op">=</span> <span class="st">"https://may2024.archive.ensembl.org"</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="listing-available-data-sets">Listing available data sets<a class="anchor" aria-label="anchor" href="#listing-available-data-sets"></a>
</h3>
<p>Each biomart contains a number of data sets.</p>
<p>The function <code>listDatasets()</code> can be used to display
information about those data sets. This is convenient as users do not
need to memorise the name of the data sets, and the information returned
by the function includes a short description of each data set, as well
as its version.</p>
<p>In the example below, we restrict the output table to the first few
rows, as the full table comprises 214 rows.</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="fu">listDatasets</span><span class="op">(</span><span class="va">mart</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                       dataset                           description
1 abrachyrhynchus_gene_ensembl Pink-footed goose genes (ASM259213v1)
2     acalliptera_gene_ensembl      Eastern happy genes (fAstCal1.3)
3   acarolinensis_gene_ensembl       Green anole genes (AnoCar2.0v2)
4    acchrysaetos_gene_ensembl       Golden eagle genes (bAquChr1.2)
5    acitrinellus_gene_ensembl        Midas cichlid genes (Midas_v5)
6    amelanoleuca_gene_ensembl       Giant panda genes (ASM200744v2)
      version
1 ASM259213v1
2  fAstCal1.3
3 AnoCar2.0v2
4  bAquChr1.2
5    Midas_v5
6 ASM200744v2</code></pre>
</div>
<p>In the output above, the key piece of information is the
<code>dataset</code> column, which provides the identifier that <em><a href="https://bioconductor.org/packages/3.19/biomaRt" class="external-link">biomaRt</a></em>
functions will need to access data from the corresponding biomart
table.</p>
<p>In this demonstration, we will use the Ensembl gene set for Homo
sapiens, which is not visible in the output above.</p>
<p>Given the number of data sets available, let us programmatically
filter the table of information using pattern matching rather than
searching the table manually:</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">subset</span><span class="op">(</span><span class="fu">listDatasets</span><span class="op">(</span><span class="va">mart</span><span class="op">)</span>, <span class="fu">grepl</span><span class="op">(</span><span class="st">"sapiens"</span>, <span class="va">dataset</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                 dataset              description    version
80 hsapiens_gene_ensembl Human genes (GRCh38.p14) GRCh38.p14</code></pre>
</div>
<p>From the output above, we identify the desired data set identifier as
<code>hsapiens_gene_ensembl</code>.</p>
</div>
<div class="section level3">
<h3 id="connecting-to-a-data-set">Connecting to a data set<a class="anchor" aria-label="anchor" href="#connecting-to-a-data-set"></a>
</h3>
<p>Having chosen the data set that we want to use, we need to call the
function <code>useMart()</code> again, this time specifying the selected
data set.</p>
<p>Typically, one would copy paste the previous call to
<code>useMart()</code> and edit as needed. It is also common practice to
replace the <code>mart</code> object with the new connection.</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mart</span> <span class="op">&lt;-</span> <span class="fu">useMart</span><span class="op">(</span></span>
<span>  biomart <span class="op">=</span> <span class="st">"ENSEMBL_MART_ENSEMBL"</span>,</span>
<span>  dataset <span class="op">=</span> <span class="st">"hsapiens_gene_ensembl"</span>,</span>
<span>  host <span class="op">=</span> <span class="st">"https://may2024.archive.ensembl.org"</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="listing-information-available-in-a-data-set">Listing information available in a data set<a class="anchor" aria-label="anchor" href="#listing-information-available-in-a-data-set"></a>
</h3>
<p>BioMart tables contain many pieces of information also known as
‘attributes’. So many, in fact, that they have been grouped into
categories also known as ‘pages’.</p>
<p>The function <code>listAttributes()</code> can be used to display
information about those attributes. This is convenient as users do not
need to memorise the name of the attributes, and the information
returned by the function includes a short description of each attribute,
as well as its page categorisation.</p>
<p>In the example below, we restrict the output table to the first few
rows, as the full table comprises 3157 rows.</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="fu">listAttributes</span><span class="op">(</span><span class="va">mart</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                           name                  description         page
1               ensembl_gene_id               Gene stable ID feature_page
2       ensembl_gene_id_version       Gene stable ID version feature_page
3         ensembl_transcript_id         Transcript stable ID feature_page
4 ensembl_transcript_id_version Transcript stable ID version feature_page
5            ensembl_peptide_id            Protein stable ID feature_page
6    ensembl_peptide_id_version    Protein stable ID version feature_page</code></pre>
</div>
<p>In the output above, the key piece of information is the
<code>name</code> column, which provides the identifier that <em><a href="https://bioconductor.org/packages/3.19/biomaRt" class="external-link">biomaRt</a></em>
functions will need to query that information from the corresponding
biomart data set.</p>
<p>The choice of attributes to query now depends on what it is we wish
to achieve.</p>
<p>For instance, let us imagine that we have a set of gene identifiers,
for which we wish to query:</p>
<ul>
<li>The gene symbol</li>
<li>The name of the chromosome where the gene is located</li>
<li>The start and end position of the gene on that chromosome</li>
<li>The strand on which the gene is encoded</li>
</ul>
<p>Users would often manually explore the full table of attributes to
identify the ones they wish to include in their query. It is also
possible to programmatically filter the table of attribute, based on
experience and intuition, to narrow down the search:</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">subset</span><span class="op">(</span><span class="fu">listAttributes</span><span class="op">(</span><span class="va">mart</span><span class="op">)</span>, <span class="fu">grepl</span><span class="op">(</span><span class="st">"position"</span>, <span class="va">name</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu">grepl</span><span class="op">(</span><span class="st">"feature"</span>, <span class="va">page</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>             name     description         page
10 start_position Gene start (bp) feature_page
11   end_position   Gene end (bp) feature_page</code></pre>
</div>
</div>
<div class="section level3">
<h3 id="querying-information-from-a-biomart-table">Querying information from a BioMart table<a class="anchor" aria-label="anchor" href="#querying-information-from-a-biomart-table"></a>
</h3>
<p>We have now all the information that we need to perform the actual
query:</p>
<ul>
<li>A connection to a BioMart data set</li>
<li>The list of attributes available in that data set</li>
</ul>
<p>The function <code>getBM()</code> is the main <em><a href="https://bioconductor.org/packages/3.19/biomaRt" class="external-link">biomaRt</a></em>
query function. Given a set of filters and corresponding values, it
retrieves the attributes requested by the user from the BioMart data set
it is connected to.</p>
<p>In the example below, we manually create a vector of arbitrary gene
identifiers for our query. In practice, the query will often originate
from an earlier analysis (e.g., differential gene expression).</p>
<p>The example below also queries attributes that we have not introduced
yet. In the previous section, we described how one may search the table
of attributes returned by <code>listAttributes()</code> to identify
attributes to include in their query.</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">query_gene_ids</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span></span>
<span>  <span class="st">"ENSG00000133101"</span>,</span>
<span>  <span class="st">"ENSG00000145386"</span>,</span>
<span>  <span class="st">"ENSG00000134057"</span>,</span>
<span>  <span class="st">"ENSG00000157456"</span>,</span>
<span>  <span class="st">"ENSG00000147082"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu">getBM</span><span class="op">(</span></span>
<span>  attributes <span class="op">=</span> <span class="fu">c</span><span class="op">(</span></span>
<span>    <span class="st">"ensembl_gene_id"</span>,</span>
<span>    <span class="st">"hgnc_symbol"</span>,</span>
<span>    <span class="st">"chromosome_name"</span>,</span>
<span>    <span class="st">"start_position"</span>,</span>
<span>    <span class="st">"end_position"</span>,</span>
<span>    <span class="st">"strand"</span></span>
<span>  <span class="op">)</span>,</span>
<span>  filters <span class="op">=</span> <span class="st">"ensembl_gene_id"</span>,</span>
<span>  values <span class="op">=</span> <span class="va">query_gene_ids</span>,</span>
<span>  mart <span class="op">=</span> <span class="va">mart</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in .processResults(postRes, mart = mart, hostURLsep = sep, fullXmlQuery = fullXmlQuery, : Query ERROR: caught BioMart::Exception::Database: Error during query execution: Table 'ensembl_mart_112.hsapiens_gene_ensembl__ox_hgnc__dm' doesn't exist</code></pre>
</div>
<p>Note that we also included the filtering attribute
<code>ensembl_gene_id</code> to the attributes retrieved from the data
set. This is key to reliably match the newly retrieved attributes to
those used in the query.</p>
</div>
</section><section><h2 class="section-heading" id="querying-annotations-from-annotation-packages">Querying annotations from annotation packages<a class="anchor" aria-label="anchor" href="#querying-annotations-from-annotation-packages"></a>
</h2>
<hr class="half-width">
<div class="section level3">
<h3 id="families-of-annotation-packages">Families of annotation packages<a class="anchor" aria-label="anchor" href="#families-of-annotation-packages"></a>
</h3>
<p>To balance the need for comprehensive information while maintaining
reasonable package sizes, Bioconductor annotation packages are organised
by release, data type, and species.</p>
<p>The major families of Bioconductor annotation packages are:</p>
<ul>
<li>
<code>OrgDb</code> packages provide mapping between various types of
gene identifiers and pathway information.</li>
<li>
<code>EnsDb</code> packages provide individual releases of Ensembl
annotations.</li>
<li>
<code>TxDb</code> packages provide individual releases of UCSC
annotations.</li>
</ul>
<p>All those families of annotations derive from the
<code>AnnotationDb</code> base class defined in the <em><a href="https://bioconductor.org/packages/3.19/AnnotationDbi" class="external-link">AnnotationDbi</a></em>
package. As a result, any of those annotation packages can be accessed
using the same set of R functions, as demonstrated in the following
sections.</p>
</div>
<div class="section level3">
<h3 id="using-an-orgdb-package">Using an OrgDb package<a class="anchor" aria-label="anchor" href="#using-an-orgdb-package"></a>
</h3>
<p>In this example, we will use the <em><a href="https://bioconductor.org/packages/3.19/org.Hs.eg.db" class="external-link">org.Hs.eg.db</a></em>
package to demonstrate the use of gene annotations for the human
species.</p>
<p>Let us first load the package:</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span></span></code></pre>
</div>
<p>Each <code>OrgDb</code> package contains an object named identically
to the package itself. That object contains the annotations that the
package is meant to disseminate.</p>
<p>Aside from querying information, the whole object can be called to
print information about the annotations it contains, including the date
at which the snapshots of annotations that it contains were made.</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">org.Hs.eg.db</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>OrgDb object:
| DBSCHEMAVERSION: 2.1
| Db type: OrgDb
| Supporting package: AnnotationDbi
| DBSCHEMA: HUMAN_DB
| ORGANISM: Homo sapiens
| SPECIES: Human
| EGSOURCEDATE: 2024-Mar12
| EGSOURCENAME: Entrez Gene
| EGSOURCEURL: ftp://ftp.ncbi.nlm.nih.gov/gene/DATA
| CENTRALID: EG
| TAXID: 9606
| GOSOURCENAME: Gene Ontology
| GOSOURCEURL: http://current.geneontology.org/ontology/go-basic.obo
| GOSOURCEDATE: 2024-01-17
| GOEGSOURCEDATE: 2024-Mar12
| GOEGSOURCENAME: Entrez Gene
| GOEGSOURCEURL: ftp://ftp.ncbi.nlm.nih.gov/gene/DATA
| KEGGSOURCENAME: KEGG GENOME
| KEGGSOURCEURL: ftp://ftp.genome.jp/pub/kegg/genomes
| KEGGSOURCEDATE: 2011-Mar15
| GPSOURCENAME: UCSC Genome Bioinformatics (Homo sapiens)
| GPSOURCEURL:
| GPSOURCEDATE: 2024-Feb29
| ENSOURCEDATE: 2023-Nov22
| ENSOURCENAME: Ensembl
| ENSOURCEURL: ftp://ftp.ensembl.org/pub/current_fasta
| UPSOURCENAME: Uniprot
| UPSOURCEURL: http://www.UniProt.org/
| UPSOURCEDATE: Thu Apr 18 21:39:39 2024</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
Please see: help('select') for usage information</code></pre>
</div>
<p>That same object is the one that needs to be supplied to <em><a href="https://bioconductor.org/packages/3.19/AnnotationDbi" class="external-link">AnnotationDbi</a></em>
functions for running queries and retrieving annotations.</p>
</div>
<div class="section level3">
<h3 id="listing-information-available-in-an-annotation-package">Listing information available in an annotation package<a class="anchor" aria-label="anchor" href="#listing-information-available-in-an-annotation-package"></a>
</h3>
<p>The function <code>columns()</code> can be used to display the
annotations available in the object.</p>
<p>Here, the word ‘column’ refers to columns of tables used to store
information in database, the very same concept as ‘attributes’ in
BioMart. In other words, columns represent all the types of annotations
that may be retrieved from the object.</p>
<p>This is convenient as users do not need to memorise the names of the
columns of annotations available in the package.</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">columns</span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] "ACCNUM"       "ALIAS"        "ENSEMBL"      "ENSEMBLPROT"  "ENSEMBLTRANS"
 [6] "ENTREZID"     "ENZYME"       "EVIDENCE"     "EVIDENCEALL"  "GENENAME"
[11] "GENETYPE"     "GO"           "GOALL"        "IPI"          "MAP"
[16] "OMIM"         "ONTOLOGY"     "ONTOLOGYALL"  "PATH"         "PFAM"
[21] "PMID"         "PROSITE"      "REFSEQ"       "SYMBOL"       "UCSCKG"
[26] "UNIPROT"     </code></pre>
</div>
</div>
<div class="section level3">
<h3 id="listing-keys-and-key-types">Listing keys and key types<a class="anchor" aria-label="anchor" href="#listing-keys-and-key-types"></a>
</h3>
<p>In database terminology, <em>keys</em> are the values by which
information may be queried from a database table.</p>
<p>Information being organised in columns, <em>key types</em> are the
names of the columns in which the key values are stored.</p>
<p>Given the variable number of columns in database tables, some tables
may allow information to be queried by more than one key. As a result,
it is crucial to specify both the keys and the type of key as part of
the query.</p>
<p>The function <code>keytypes()</code> can be used to display the names
of the columns that may be used to query information from the
object.</p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">keytypes</span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] "ACCNUM"       "ALIAS"        "ENSEMBL"      "ENSEMBLPROT"  "ENSEMBLTRANS"
 [6] "ENTREZID"     "ENZYME"       "EVIDENCE"     "EVIDENCEALL"  "GENENAME"
[11] "GENETYPE"     "GO"           "GOALL"        "IPI"          "MAP"
[16] "OMIM"         "ONTOLOGY"     "ONTOLOGYALL"  "PATH"         "PFAM"
[21] "PMID"         "PROSITE"      "REFSEQ"       "SYMBOL"       "UCSCKG"
[26] "UNIPROT"     </code></pre>
</div>
<p>The function <code>keys()</code> can be used to display all the
possible values for a given key type.</p>
<p>It is generally better practice to specify the type of key being
queried (to avoid ambiguity), although database tables typically have a
‘primary key’ used if users do not specify a type themselves.</p>
<p>In the example below, we restrict the list of gene symbol keys to the
first few values, as the full set comprises 193279 values.</p>
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="fu">keys</span><span class="op">(</span><span class="va">org.Hs.eg.db</span>, keytype <span class="op">=</span> <span class="st">"SYMBOL"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "A1BG"  "A2M"   "A2MP1" "NAT1"  "NAT2"  "NATP" </code></pre>
</div>
</div>
<div class="section level3">
<h3 id="querying-information-from-an-annotation-package">Querying information from an annotation package<a class="anchor" aria-label="anchor" href="#querying-information-from-an-annotation-package"></a>
</h3>
<p>The function <code>select()</code> is the main <em><a href="https://bioconductor.org/packages/3.19/AnnotationDbi" class="external-link">AnnotationDbi</a></em>
query function. Given an <code>AnnotationDb</code> object, key values,
and columns (and optionally the type of key supplied if not the primary
key), it retrieves the columns requested by the user from the annotation
object.</p>
<p>In the example below, we re-use the vector of arbitrary gene
identifiers used in the BioMart example a few sections above.</p>
<p>As you can see from the output of the <code>columns()</code>
function, the annotation object does not contain some of the attributes
that we queried in the Biomart example. In this case, let us query:</p>
<ul>
<li>the gene symbol</li>
<li>the gene name</li>
<li>the gene type</li>
</ul>
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">select</span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="va">org.Hs.eg.db</span>,</span>
<span>  keys <span class="op">=</span> <span class="va">query_gene_ids</span>,</span>
<span>  columns <span class="op">=</span> <span class="fu">c</span><span class="op">(</span></span>
<span>    <span class="st">"SYMBOL"</span>,</span>
<span>    <span class="st">"GENENAME"</span>,</span>
<span>    <span class="st">"GENETYPE"</span></span>
<span>  <span class="op">)</span>,</span>
<span>  keytype <span class="op">=</span> <span class="st">"ENSEMBL"</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>'select()' returned 1:1 mapping between keys and columns</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>          ENSEMBL SYMBOL  GENENAME       GENETYPE
1 ENSG00000133101  CCNA1 cyclin A1 protein-coding
2 ENSG00000145386  CCNA2 cyclin A2 protein-coding
3 ENSG00000134057  CCNB1 cyclin B1 protein-coding
4 ENSG00000157456  CCNB2 cyclin B2 protein-coding
5 ENSG00000147082  CCNB3 cyclin B3 protein-coding</code></pre>
</div>
<p>One small but notable difference with <em><a href="https://bioconductor.org/packages/3.19/biomaRt" class="external-link">biomaRt</a></em>
is that the output of <code>select()</code> automatically contains the
column that correspond to the key type used in the query. In other
words, there is no need to specify the key type(s) again in the
column(s) to retrieve.</p>
</div>
<div class="section level3">
<h3 id="vectorized-11-mapping">Vectorized 1:1 mapping<a class="anchor" aria-label="anchor" href="#vectorized-11-mapping"></a>
</h3>
<p>It is sometimes possible for annotations to display 1-to-many
relationships. For instance, individual genes typically have a unique
Ensembl gene identifier, while they may be known under multiple gene
name aliases.</p>
<p>The <code>select()</code> function demonstrated in the previous
section automatically returns <em>all</em> values in the columns
requested, for the key specified. This is possible thanks to the tabular
format in which annotations are returned; rows are added, repeating
values as necessary to display them on the same row as every other
values they are associated with.</p>
<p>In some cases, that behaviour is not desirable. Instead, users may
wish to retrieve a single value for each key that they input. One common
scenario arises during differential gene expression (DGE), where gene
identifiers are used to uniquely identify genes throughout the analysis,
while gene symbols are added to the final table of DGE statistics, to
provide more readable human-friendly gene identifiers. However, it is
not desirable to duplicate rows of DGE statistics, and thus only a
single gene symbol is required to annotate each gene.</p>
<p>The function <code>mapIds()</code> can be used for this purpose. A
major difference between the functions <code>mapIds()</code> and
<code>select()</code> are their arguments <code>column</code> (singular)
and <code>columns</code> (plural), respectively. The function
<code>mapIds()</code> accepts a single column name and returns a named
character vector where names are the input query values, and values are
the corresponding values in the requested column.</p>
<p>To deal with 1-to-many relationships, the function
<code>mapIds()</code> has an argument <code>multiVals</code> which can
be used to specify how the function should handle multiple values. The
default is to take the first value and ignore any other value.</p>
<p>In the example below, we query the gene symbol for a set of Ensembl
gene identifiers.</p>
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">mapIds</span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="va">org.Hs.eg.db</span>,</span>
<span>  keys <span class="op">=</span> <span class="va">query_gene_ids</span>,</span>
<span>  column <span class="op">=</span> <span class="st">"SYMBOL"</span>,</span>
<span>  keytype <span class="op">=</span> <span class="st">"ENSEMBL"</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>'select()' returned 1:1 mapping between keys and columns</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>ENSG00000133101 ENSG00000145386 ENSG00000134057 ENSG00000157456 ENSG00000147082
        "CCNA1"         "CCNA2"         "CCNB1"         "CCNB2"         "CCNB3" </code></pre>
</div>
<div id="challenge" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Load the packages <em><a href="https://bioconductor.org/packages/3.19/EnsDb.Hsapiens.v86" class="external-link">EnsDb.Hsapiens.v86</a></em>
and <em><a href="https://bioconductor.org/packages/3.19/TxDb.Hsapiens.UCSC.hg38.knownGene" class="external-link">TxDb.Hsapiens.UCSC.hg38.knownGene</a></em>.
Then, display the columns of annotations available in those
packages.</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb35">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">EnsDb.Hsapiens.v86</span><span class="op">)</span></span>
<span><span class="fu">columns</span><span class="op">(</span><span class="va">EnsDb.Hsapiens.v86</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] "ENTREZID"            "EXONID"              "EXONIDX"
 [4] "EXONSEQEND"          "EXONSEQSTART"        "GENEBIOTYPE"
 [7] "GENEID"              "GENENAME"            "GENESEQEND"
[10] "GENESEQSTART"        "INTERPROACCESSION"   "ISCIRCULAR"
[13] "PROTDOMEND"          "PROTDOMSTART"        "PROTEINDOMAINID"
[16] "PROTEINDOMAINSOURCE" "PROTEINID"           "PROTEINSEQUENCE"
[19] "SEQCOORDSYSTEM"      "SEQLENGTH"           "SEQNAME"
[22] "SEQSTRAND"           "SYMBOL"              "TXBIOTYPE"
[25] "TXCDSSEQEND"         "TXCDSSEQSTART"       "TXID"
[28] "TXNAME"              "TXSEQEND"            "TXSEQSTART"
[31] "UNIPROTDB"           "UNIPROTID"           "UNIPROTMAPPINGTYPE" </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb37">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">TxDb.Hsapiens.UCSC.hg38.knownGene</span><span class="op">)</span></span>
<span><span class="fu">columns</span><span class="op">(</span><span class="va">TxDb.Hsapiens.UCSC.hg38.knownGene</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] "CDSCHROM"   "CDSEND"     "CDSID"      "CDSNAME"    "CDSPHASE"
 [6] "CDSSTART"   "CDSSTRAND"  "EXONCHROM"  "EXONEND"    "EXONID"
[11] "EXONNAME"   "EXONRANK"   "EXONSTART"  "EXONSTRAND" "GENEID"
[16] "TXCHROM"    "TXEND"      "TXID"       "TXNAME"     "TXSTART"
[21] "TXSTRAND"   "TXTYPE"    </code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>Bioconductor provides a wide range annotation packages.</li>
<li>Some Bioconductor software packages can be used to programmatically
access online resources.</li>
<li>Users should carefully choose their source of annotations based on
their needs and expectations.</li>
</ul>
</div>
</div>
</div>
</div>
</section></section><section id="aio-09-summarizedexperiment"><p>Content from <a href="09-summarizedexperiment.html">SummarizedExperimentクラス</a></p>
<hr>
<p>Last updated on 2025-07-22 |

        <a href="https://github.com/carpentries-incubator/bioc-project/edit/main/episodes/09-summarizedexperiment.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> NA minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<hr>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>SummarizedExperimentオブジェクトの情報はどのように整理されていますか？</li>
<li>その情報はどのように追加、編集、アクセスできますか？</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>実験データとメタデータがどのように1つのオブジェクトに保存されるかを説明してください。</li>
<li>データとメタデータを分析の過程全体で同期させることが重要な理由を説明してください。</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="パッケージのインストール">パッケージのインストール<a class="anchor" aria-label="anchor" href="#%E3%83%91%E3%83%83%E3%82%B1%E3%83%BC%E3%82%B8%E3%81%AE%E3%82%A4%E3%83%B3%E3%82%B9%E3%83%88%E3%83%BC%E3%83%AB"></a>
</h2>
<hr class="half-width">
<p>次のセクションに進む前に、必要なBioconductorのパッケージをいくつかインストールします。
最初に、<em><a href="https://bioconductor.org/packages/3.19/BiocManager" class="external-link">BiocManager</a></em>パッケージがインストールされているか確認します。それがない場合はインストールします。
その後、<code>BiocManager::install()</code>関数を使用して必要なパッケージをインストールします。</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu">requireNamespace</span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu">install.packages</span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="st">"SummarizedExperiment"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
</section><section><h2 class="section-heading" id="モチベーション">モチベーション<a class="anchor" aria-label="anchor" href="#%E3%83%A2%E3%83%81%E3%83%99%E3%83%BC%E3%82%B7%E3%83%A7%E3%83%B3"></a>
</h2>
<hr class="half-width">
<p>実験は多面的なデータセットで、通常、分析に必要な少なくとも2つの重要な情報を含んでいます：</p>
<ul>
<li>アッセイデータは、通常、サンプルのセット内の特徴のセットの測定を表す行列です
(例：RNAシーケンシング)。</li>
<li>サンプルメタデータは、通常、サンプルに関する情報を表す<code>data.frame</code>です
(例：治療群)。</li>
</ul>
<p>これらすべての情報は、下流の分析で情報を正確に処理し、信頼できる結果を生成するために、
同じサンプルで同じ順序に保たれる必要があります。</p>
<p>分析ワークフローで、サンプルのサブセットを分析したり、より正確な下流の分析を可能にするために削除する必要のある外れ値を特定したりすることは非常に一般的です。
そのような場合、実験のすべての側面は同じサンプルのセットにサブセットされる必要があります
-- 同じ順序で – データセットの一貫性と正確な結果を保つために。</p>
<p><code>SummarizedExperiment</code> – <em><a href="https://bioconductor.org/packages/3.19/SummarizedExperiment" class="external-link">SummarizedExperiment</a></em>パッケージで実装されている
--
は、個々の実験の重要な側面を単一のオブジェクトに統合するコンテナを提供し、サブセットや再配置の操作中にデータとメタデータを調整します。
多くの生物学的データ型と包括的な機能セットを受け入れる柔軟性があり、
Bioconductor全体で再利用される人気のあるデータ構造です。
例えば、<code>SummarizedExperiment</code>に慣れていることは、 <em><a href="https://bioconductor.org/packages/3.19/DESeq2" class="external-link">DESeq2</a></em>パッケージや、シングルセル解析のための<em><a href="https://bioconductor.org/packages/3.19/SingleCellExperiment" class="external-link">SingleCellExperiment</a></em>拡張クラスで作業するために必要です。</p>
</section><section><h2 class="section-heading" id="クラス構造">クラス構造<a class="anchor" aria-label="anchor" href="#%E3%82%AF%E3%83%A9%E3%82%B9%E6%A7%8B%E9%80%A0"></a>
</h2>
<hr class="half-width">
<p><code>SummarizedExperiment</code>は、行が対象の特徴（例：遺伝子、転写、エクソンなど）を表す行列のようなコンテナです。
列はサンプルを表します。</p>
<p>オブジェクトは1つ以上のアッセイを含むことができ、各アッセイは同じ次元の行列のようなオブジェクトで表されます。</p>
<p>特徴に関する情報は、<code>SummarizedExperiment</code>オブジェクト内にネストされた<code>DataFrame</code>オブジェクトに保存され、
関数<code>rowData()</code>を使用してアクセスできます。
<code>DataFrame</code>の各行は、SummarizedExperimentオブジェクトの対応する行の特徴に関する情報を提供します。
その情報には、実験とは無関係な注釈（例：遺伝子識別子）や、ワークフロー中にアッセイデータから計算された品質管理メトリックが含まれる場合があります。</p>
<p>同様に、サンプルに関する情報は、別の<code>DataFrame</code>オブジェクトに保存され、これも<code>SummarizedExperiment</code>オブジェクトにネストされ、
関数<code>colData()</code>を使用してアクセスできます。</p>
<p>次のグラフィックは、クラスのジオメトリを表示し、縦（列）および横（行）関係を強調します。
これは、<em><a href="https://bioconductor.org/packages/3.19/SummarizedExperiment" class="external-link">SummarizedExperiment</a></em>パッケージのビネットから得られました。</p>
<figure><img src="../fig/summarizedexperiment.svg" alt="SummarizedExperimentクラスの模式図です。" class="figure mx-auto d-block"></figure></section><section><h2 class="section-heading" id="summarizedexperimentオブジェクトの作成">SummarizedExperimentオブジェクトの作成<a class="anchor" aria-label="anchor" href="#summarizedexperiment%E3%82%AA%E3%83%96%E3%82%B8%E3%82%A7%E3%82%AF%E3%83%88%E3%81%AE%E4%BD%9C%E6%88%90"></a>
</h2>
<hr class="half-width">
<p>はじめにパッケージをロードしましょう：</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">SummarizedExperiment</span><span class="op">)</span></span></code></pre>
</div>
<p>それでは、レッスンセッティング中にダウンロードしたファイルからアッセイデータをインポートします。</p>
<p>そのファイルは単純なテキストファイルで、
最初の列には作られた特徴識別子が含まれ、
他のすべての列には作られたサンプルのシミュレーションデータが含まれています。
そのため、基本R関数<code>read.csv</code>を使用してファイルを解析し、
<code>data.frame</code>オブジェクトに変換できます。</p>
<p>以下の例では、行名は最初の列にあることを示しており、
関数は出力オブジェクト内ですぐに行名を設定します。
指定しなければ、関数は通常の列として解析し、デフォルトの整数インデックスで行名をそのままに
設定します。</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">count_data</span> <span class="op">&lt;-</span> <span class="fu">read.csv</span><span class="op">(</span><span class="st">"data/counts.csv"</span>, row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">count_data</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>        sample_1 sample_2 sample_3 sample_4
gene_1       109       84       91      105
gene_2       111       97       98      108
gene_3        89      121      105       99
gene_4       105      109      122      101
gene_5        82       97      112       83
gene_6        89       96       90      116
gene_7       121       95       88      106
gene_8       101      101       86      103
gene_9        91      119       89       87
gene_10       81      111       81      118
gene_11       93      118       93       99
gene_12      103      111      116      103
gene_13       89      126      103      100
gene_14      101      107      111       79
gene_15       96       91      103      108
gene_16      110      102      128      103
gene_17       95      106      118      100
gene_18       99      115      114      102
gene_19      114      105       94      118
gene_20      110       88       99      102
gene_21      116       95       94      105
gene_22      114       96      107       91
gene_23       97      120       93       90
gene_24       91       84      118       97
gene_25       99      106       97      110</code></pre>
</div>
<p>1つのアッセイデータ行列があれば、<code>SummarizedExperiment</code>オブジェクトを作成できますが、
サンプルメタデータがなければ、サンプルに関する情報を必要としない無監督分析のみが可能です。</p>
<p>以下の例では、カウントデータの行列を’counts’という名前で保存する<code>SummarizedExperiment</code>オブジェクトを作成します。
引数’Без’（複数形）は1つ以上のアッセイを受け入れることができ、 --
上記で説明されているように
–、そのため、唯一のアッセイ行列を名前付きの<code>list</code>にカプセル化します。
アッセイに名前を付けることは、複数のアッセイを含むワークフローで、各アッセイを明確に識別し、取得するために重要です。</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu">SummarizedExperiment</span><span class="op">(</span></span>
<span>  assays <span class="op">=</span> <span class="fu">list</span><span class="op">(</span>counts <span class="op">=</span> <span class="va">count_data</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">se</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>class: SummarizedExperiment
dim: 25 4
metadata(0):
assays(1): counts
rownames(25): gene_1 gene_2 ... gene_24 gene_25
rowData names(0):
colnames(4): sample_1 sample_2 sample_3 sample_4
colData names(0):</code></pre>
</div>
<p>上の出力では、オブジェクトの要約ビューがアッセイを確認でき、
そのため、全体的な<code>SummarizedExperiment</code>オブジェクトが4つのサンプル内の25の特徴に関する情報を含んでいることを思い出させます。
オブジェクトには、行メタデータや列メタデータは含まれていません。</p>
<p>より包括的な<code>SummarizedExperiment</code>オブジェクトを作成するために、他の2つのファイルから遺伝子メタデータとサンプルメタデータをインポートします。</p>
<p>ファイルはカウントデータと同様にフォーマットされているため、
再び基本R関数<code>read.csv()</code>を使用して、<code>data.frame</code>オブジェクトに解析します。</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sample_metadata</span> <span class="op">&lt;-</span> <span class="fu">read.csv</span><span class="op">(</span><span class="st">"data/sample_metadata.csv"</span>, row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">sample_metadata</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>         condition batch
sample_1         A     1
sample_2         A     2
sample_3         B     1
sample_4         B     2</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene_metadata</span> <span class="op">&lt;-</span> <span class="fu">read.csv</span><span class="op">(</span><span class="st">"data/gene_metadata.csv"</span>, row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">gene_metadata</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>        chromosome
gene_1           4
gene_2           4
gene_3           5
gene_4           4
gene_5           5
gene_6           1
gene_7           2
gene_8           1
gene_9           3
gene_10          1
gene_11          1
gene_12          5
gene_13          5
gene_14          1
gene_15          3
gene_16          4
gene_17          2
gene_18          5
gene_19          1
gene_20          3
gene_21          5
gene_22          5
gene_23          1
gene_24          4
gene_25          5</code></pre>
</div>
<p>これで、遺伝子とサンプルメタデータを含めた<code>SummarizedExperiment</code>オブジェクトを再作成できます：</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu">SummarizedExperiment</span><span class="op">(</span></span>
<span>  assays <span class="op">=</span> <span class="fu">list</span><span class="op">(</span>counts <span class="op">=</span> <span class="va">count_data</span><span class="op">)</span>,</span>
<span>  colData <span class="op">=</span> <span class="va">sample_metadata</span>,</span>
<span>  rowData <span class="op">=</span> <span class="va">gene_metadata</span></span>
<span><span class="op">)</span></span>
<span><span class="va">se</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>class: SummarizedExperiment
dim: 25 4
metadata(0):
assays(1): counts
rownames(25): gene_1 gene_2 ... gene_24 gene_25
rowData names(1): chromosome
colnames(4): sample_1 sample_2 sample_3 sample_4
colData names(2): condition batch</code></pre>
</div>
<p>この出力を前の’アッセイのみ’バージョンと比較すると、<code>rowData</code>と<code>colData</code>
のコンポーネントがそれぞれ1つと4つのメタデータを含むことがわかります。</p>
</section><section><h2 class="section-heading" id="情報にアクセスする">情報にアクセスする<a class="anchor" aria-label="anchor" href="#%E6%83%85%E5%A0%B1%E3%81%AB%E3%82%A2%E3%82%AF%E3%82%BB%E3%82%B9%E3%81%99%E3%82%8B"></a>
</h2>
<hr class="half-width">
<p>いくつかの関数が<code>SummarizedExperiment</code>オブジェクトのさまざまなコンポーネントにアクセスできます。</p>
<p><code>assays()</code>関数は、オブジェクトに保存されたアッセイのリストを返します。
出力は常に<code>List</code>であり、オブジェクトに単一のアッセイが含まれている場合でもです。</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">assays</span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>List of length 1
names(1): counts</code></pre>
</div>
<p><code>assayNames()</code>関数は、アッセイ名の文字ベクトルを返します。
これは、オブジェクトに大量のアッセイが含まれる場合に最も便利です。
<code>assays()</code>関数（上記参照）は、すべてを表示しないかもしれません。
さまざまなアッセイの名前を知ることは、個々のアッセイにアクセスするための鍵です。</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">assayNames</span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "counts"</code></pre>
</div>
<p><code>assay()</code>関数を使用して、オブジェクトから単一のアッセイを取得できます。
これに対して、関数には目的のアッセイの名前または整数位置を指定する必要があります。
未指定の場合、関数は自動的にオブジェクト内の最初のアッセイを返します。</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">se</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>       sample_1 sample_2 sample_3 sample_4
gene_1      109       84       91      105
gene_2      111       97       98      108
gene_3       89      121      105       99
gene_4      105      109      122      101
gene_5       82       97      112       83
gene_6       89       96       90      116</code></pre>
</div>
<p><code>colData()</code>および<code>rowData()</code>関数を使用して、サンプルメタデータおよび行メタデータを取得できます。</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">colData</span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DataFrame with 4 rows and 2 columns
           condition     batch
         &lt;character&gt; &lt;integer&gt;
sample_1           A         1
sample_2           A         2
sample_3           B         1
sample_4           B         2</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">rowData</span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DataFrame with 25 rows and 1 column
        chromosome
         &lt;integer&gt;
gene_1           4
gene_2           4
gene_3           5
gene_4           4
gene_5           5
...            ...
gene_21          5
gene_22          5
gene_23          1
gene_24          4
gene_25          5</code></pre>
</div>
<p>また、
<code>$</code>演算子を使用して、サンプルメタデータの単一列にアクセスできます。
この演算子の便利な機能は、RStudioで自動的にトリガーされるオートコンプリートと、ターミナルアプリケーションでのタブキーを使用して自動的にトリガーされることです。</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span><span class="op">$</span><span class="va">batch</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 1 2 1 2</code></pre>
</div>
<p>特に、特徴メタデータの単一列にアクセスするための演算子はありません。
そのため、ユーザーはまず<code>rowData()</code>によって返される完全な<code>DataFrame</code>にアクセスし、
その後、標準の$または[[演算子を使用して列にアクセスする必要があります。</p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">rowData</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">[[</span><span class="st">"chromosome"</span><span class="op">]</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] 4 4 5 4 5 1 2 1 3 1 1 5 5 1 3 4 2 5 1 3 5 5 1 4 5</code></pre>
</div>
<div class="section level3">
<h3 id="情報の追加と編集">情報の追加と編集<a class="anchor" aria-label="anchor" href="#%E6%83%85%E5%A0%B1%E3%81%AE%E8%BF%BD%E5%8A%A0%E3%81%A8%E7%B7%A8%E9%9B%86"></a>
</h3>
<p>情報は、<code>SummarizedExperiment</code>作成後に追加することができます。
実際、これは、正規化されたアッセイ値を計算し、
それをアッセイのリストに追加し、
特徴またはサンプルに対する品質管理メトリックを計算し、
それを<code>rowData</code>と<code>colData</code>のコンポーネントに適切に追加し、
全体のオブジェクト内に保存された情報の量を徐々に増やすプロセスの基礎となります。</p>
<p>情報にアクセスするための多くの関数が、上記のセクションで説明されていますが、
新しい値を追加したり、既存の値を編集したりするための対応関数があります。
編集は、既に使用されている名前の下に値を追加するだけの結果であり、
既存の値を置き換える効果があります。</p>
<p>以下の例では、’counts’アッセイに対して1の擬似カウントを追加した後に、
その結果としてアッセイ名’logcounts’を追加します。</p>
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">assay</span><span class="op">(</span><span class="va">se</span>, <span class="st">"logcounts"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">log1p</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">se</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">se</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>class: SummarizedExperiment
dim: 25 4
metadata(0):
assays(2): counts logcounts
rownames(25): gene_1 gene_2 ... gene_24 gene_25
rowData names(1): chromosome
colnames(4): sample_1 sample_2 sample_3 sample_4
colData names(2): condition batch</code></pre>
</div>
<p>上の出力では、オブジェクトが現在2つのアッセイを持っていることがわかります：
最初に作成されたオブジェクトに含まれている’counts’アッセイと、
今追加した’logcounts’アッセイです。</p>
<p>同様に、<code>colData()</code>および<code>rowData()</code>関数、
ならびに<code>$</code>演算子を使用して、対応するコンポーネントの値を追加および編集することができます。</p>
<p>以下の例では、各サンプルのカウントの合計を計算し、
結果を新しい名前’sum_counts’のサンプルメタデータテーブルに保存します。</p>
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">colData</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">[[</span><span class="st">"sum_counts"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">colSums</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">se</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">colData</span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DataFrame with 4 rows and 3 columns
           condition     batch sum_counts
         &lt;character&gt; &lt;integer&gt;  &lt;numeric&gt;
sample_1           A         1       2506
sample_2           A         2       2600
sample_3           B         1       2550
sample_4           B         2       2533</code></pre>
</div>
<p>次の例では、各特徴の平均カウントを計算し、
新しい名前’mean_counts’の特徴メタデータテーブルに保存します。</p>
<div class="codewrapper sourceCode" id="cb31">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">rowData</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">[[</span><span class="st">"mean_counts"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">rowSums</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">se</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">rowData</span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DataFrame with 25 rows and 2 columns
        chromosome mean_counts
         &lt;integer&gt;   &lt;numeric&gt;
gene_1           4         389
gene_2           4         414
gene_3           5         414
gene_4           4         437
gene_5           5         374
...            ...         ...
gene_21          5         410
gene_22          5         408
gene_23          1         400
gene_24          4         390
gene_25          5         412</code></pre>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>
<code>SummarizedExperiment</code>クラスは、アッセイデータとメタデータの両方を保存するための単一のコンテナを提供します。</li>
<li>アッセイデータとメタデータは、サブセット化と再配置の操作を通じて同期が保たれます。</li>
<li>
<code>SummarizedExperiment</code>オブジェクトのさまざまなコンポーネントに保存された情報にアクセスしたり、追加したり、編集したりするための包括的な関数セットが利用可能です。</li>
</ul>
</div>
</div>
</div>
</div>
</section></section>
</div>
    </main>
</div>
<!-- END  : inst/pkgdown/templates/content-extra.html -->

      </div>
<!--/div.row-->
      		<footer class="row footer mx-md-3"><hr>
<div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/carpentries-incubator/bioc-project/edit/main/README.md" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/carpentries-incubator/bioc-project/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/carpentries-incubator/bioc-project/" class="external-link">Source</a></p>
				<p><a href="https://github.com/carpentries-incubator/bioc-project/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:team@carpentries.org">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/main" class="external-link">sandpaper (0.16.13.9000)</a>, <a href="https://github.com/carpentries/pegboard/tree/main" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.6" class="external-link">varnish (1.0.6)</a></p>
			</div>
		</footer>
</div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "LearningResource",
  "@id": "https://carpentries-incubator.github.io/bioc-project/instructor/aio.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/LearningResource/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "All in One View",
  "creativeWorkStatus": "active",
  "url": "https://carpentries-incubator.github.io/bioc-project/instructor/aio.html",
  "identifier": "https://carpentries-incubator.github.io/bioc-project/instructor/aio.html",
  "dateCreated": "2020-09-14",
  "dateModified": "2025-07-22",
  "datePublished": "2025-07-22"
}

  </script><script>
		feather.replace();
	</script>
</body>
</html><!-- END:   inst/pkgdown/templates/layout.html-->

